---
title: "Updating PICRUSt2 database paper"
output:
  html_document: 
    toc: yes
    toc_float: yes
---

```{R, results='hide', fig.keep='all', message=FALSE}
library(reticulate)
use_python('/Users/robynwright/anaconda3/envs/r-environment/bin/python')
```

```{python}
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from scipy.stats import mannwhitneyu, ttest_ind, pearsonr, spearmanr, randint
from matplotlib.lines import Line2D
from scipy.spatial import distance
import matplotlib as mpl
import matplotlib.colors as colors

colours = ['#AF6182', '#F7D171', '#546D84', '#C57F3B', '#898B76']
folder = '/Users/robynwright/Dropbox/Papers_writing/Revision_PICRUSt2_database_application_note/'
overall_folder = '/Users/robynwright/Dropbox/Papers_writing/Revision_PICRUSt2_database_application_note/compile_output/'
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
dataset_names = ['Blueberry', 'Cameroon', 'HMP', 'Indian', 'Mammal', 'Ocean', 'Primate']
```

# Taxa counts

Get new counts:
```{python, eval=FALSE}
#gunzip GTDB_r214/bac120_metadata_r214.tsv.gz
#gunzip GTDB_r214/ar53_metadata_r214.tsv.gz

import pandas as pd
import numpy as np

#bac = 'GTDB_r214/bac120_metadata_r214.tsv'
#arc = 'GTDB_r214/ar53_metadata_r214.tsv'
bac = 'gtdb_picrust_ref/bac_annotations/metadata.csv'
arc = 'gtdb_picrust_ref/arc_annotations/metadata.csv'

domains = [arc, bac]
all_comp_cont = []
all_numbers = []

for d in range(len(domains)):
  domain = domains[d]
  #df = pd.read_csv(domain, index_col=0, header=0, sep='\t')
  df = pd.read_csv(domain, index_col=0, header=0)
  df = df[df['checkm_contamination'] <= 10]
  print(domain, 'genomes <= 10% contamination: ', df.shape[0])
  comp_cont = [[], []]
  numbers = [[], [], [], [], [], [], []]
  for comp in [90]:
    df = df[df['checkm_completeness'] >= comp]
    print(domain, 'genomes >= '+str(comp)+'% completeness:', df.shape[0])
    completeness = df['checkm_completeness'].values
    contamination = df['checkm_contamination'].values
    print('Completeness: mean = ', np.mean(completeness), 'median = ', np.median(completeness), '(range '+str(min(completeness))+'-'+str(max(completeness))+'%)')
    comp_cont[0].append([np.mean(completeness), np.median(completeness), min(completeness), max(completeness)])
    print('Contamination: mean = ', np.mean(contamination), 'median = ', np.median(contamination), '(range '+str(min(contamination))+'-'+str(max(contamination))+'%)')
    comp_cont[1].append([np.mean(contamination), np.median(contamination), min(contamination), max(contamination)])
    tax = df['gtdb_taxonomy'].values
    levels = [[], [], [], [], [], [], []]
    level_names = ['domain', 'phylum', 'class', 'order', 'family', 'genus', 'species']
    for t in range(len(tax)):
      this_tax = tax[t].split(';')
      for l in range(len(this_tax)):
        levels[l].append(this_tax[l])
    for l in range(len(levels)):
      numbers[l].append(len(set(levels[l])))
      print(level_names[l], len(set(levels[l])))
    print('\n')
  all_numbers.append(numbers)
  all_comp_cont.append(comp_cont)
  
print(all_comp_cont)
print(all_numbers)
```

```{python}
previous_taxonomy = pd.read_csv('/Users/robynwright/Dropbox/Papers_writing/PICRUSt2 database application note/16S_PICRUSt2_taxa_levels.tsv', index_col=0, header=0, sep='\t')
levels = ['Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'Genomes']
numbers = []
domains = ['Bacteria', 'Archaea']
for d in range(len(domains)):
  domain = domains[d]
  previous_taxonomy_reduced = previous_taxonomy[previous_taxonomy['Superkingdom'] == domain]
  domain_numbers = []
  for level in levels:
    if level != 'Genomes':
      this_level = previous_taxonomy_reduced[level].values
      domain_numbers.append(len(set(this_level)))
    else:
      domain_numbers.append(previous_taxonomy_reduced.shape[0])
  numbers.append(domain_numbers)

print(numbers)

finished = True
```

Figure:
```{python}
data = pd.read_csv('/Users/robynwright/Dropbox/Papers_writing/PICRUSt2 database application note/Number_of_genomes_old_and_new.csv', index_col=0, header=0)

plt.close()
ax1 = plt.subplot(111)
levels = ['Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'Genomes']
x2 = [l-0.22 for l in range(len(levels))]
x1 = [l+0.22 for l in range(len(levels))]
ax1.bar(x1, data.loc['PICRUSt2 updated Bacteria', :].values, bottom=[0 for l in range(len(levels))], color=colours[0], edgecolor='k', width=0.4)
ax1.bar(x1, data.loc['PICRUSt2 updated Archaea', :].values, bottom=data.loc['PICRUSt2 updated Bacteria', :].values, color=colours[1], edgecolor='k', width=0.4)
ax1.bar(x2, data.loc['PICRUSt2 default Bacteria', :].values, bottom=[0 for l in range(len(levels))], color=colours[0], edgecolor='k', width=0.4)
ax1.bar(x2, data.loc['PICRUSt2 default Archaea', :].values, bottom=data.loc['PICRUSt2 default Bacteria', :].values, color=colours[1], edgecolor='k', width=0.4)

plt.xticks([l for l in range(len(levels))], levels, rotation=45)
plt.ylabel('Number of taxa')
plt.close()

plt.figure(figsize=(15,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
levels = ['Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'Genomes']
x2 = [l-0.22 for l in range(len(levels))]
x1 = [l+0.22 for l in range(len(levels))]
ax1.bar(x1, data.loc['PICRUSt2 updated Bacteria', :].values, bottom=[0 for l in range(len(levels))], color=colours[1], edgecolor='k', width=0.4)
ax2.bar(x1, data.loc['PICRUSt2 updated Archaea', :].values, bottom=[0 for l in range(len(levels))], color=colours[1], edgecolor='k', width=0.4)
ax1.bar(x2, data.loc['PICRUSt2 default Bacteria', :].values, bottom=[0 for l in range(len(levels))], color=colours[0], edgecolor='k', width=0.4)
ax2.bar(x2, data.loc['PICRUSt2 default Archaea', :].values, bottom=[0 for l in range(len(levels))], color=colours[0], edgecolor='k', width=0.4)

adding = [10, 10, 10, 100, 100, 1000, 1000]
multiply=0.1
for l in range(len(levels)):
  tx = ax1.text(x1[l], data.loc['PICRUSt2 updated Bacteria', levels[l]]+data.loc['PICRUSt2 updated Bacteria', levels[l]]*multiply, str(data.loc['PICRUSt2 updated Bacteria', levels[l]]), ha='left', va='bottom', rotation=45)
  tx = ax1.text(x2[l], data.loc['PICRUSt2 default Bacteria', levels[l]]+data.loc['PICRUSt2 default Bacteria', levels[l]]*multiply, str(data.loc['PICRUSt2 default Bacteria', levels[l]]), ha='left', va='bottom', rotation=45)
  tx = ax2.text(x1[l], data.loc['PICRUSt2 updated Archaea', levels[l]]+data.loc['PICRUSt2 updated Archaea', levels[l]]*multiply, str(data.loc['PICRUSt2 updated Archaea', levels[l]]), ha='left', va='bottom', rotation=45)
  tx = ax2.text(x2[l], data.loc['PICRUSt2 default Archaea', levels[l]]+data.loc['PICRUSt2 default Archaea', levels[l]]*multiply, str(data.loc['PICRUSt2 default Archaea', levels[l]]), ha='left', va='bottom', rotation=45)

plt.sca(ax1)
plt.xticks([l for l in range(len(levels))], levels, rotation=45)
plt.ylabel('Number of taxa')
plt.title('Bacteria', fontweight='bold')
#plt.ylim([-500, 34500])
plt.semilogy()

plt.sca(ax2)
plt.xticks([l for l in range(len(levels))], levels, rotation=45)
plt.title('Archaea', fontweight='bold')
#plt.ylim([-20, 1300])
plt.semilogy()

plt.savefig(folder+'figures/number_of_taxa.png', dpi=600, bbox_inches='tight')

```

```{python}
data = pd.read_csv(folder+'Number_of_genomes_old_and_new.csv', index_col=0, header=0)

plt.figure(figsize=(10,4))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
levels = ['Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'Genomes']
locs = [0, 1.5, 3, 4.5, 6, 7.5, 9]
x2 = [locs[l]+0.3 for l in range(len(levels))]
x1 = [locs[l]-0.3 for l in range(len(levels))]
x1.reverse()
x2.reverse()
b = ax1.barh(x1, data.loc['PICRUSt2 updated Bacteria', :].values, color=colours[1], edgecolor='k', height=0.5)
b = ax2.barh(x1, data.loc['PICRUSt2 updated Archaea', :].values, color=colours[1], edgecolor='k', height=0.5)
b = ax1.barh(x2, data.loc['PICRUSt2 default Bacteria', :].values, color=colours[0], edgecolor='k', height=0.5)
b = ax2.barh(x2, data.loc['PICRUSt2 default Archaea', :].values, color=colours[0], edgecolor='k', height=0.5)

multiply = 0.1
adding = [10, 10, 100, 100, 100, 1000, 1000]
for l in range(len(levels)):
  upd_bac = data.loc['PICRUSt2 updated Bacteria', levels[l]]
  def_bac = data.loc['PICRUSt2 default Bacteria', levels[l]]
  upd_arc = data.loc['PICRUSt2 updated Archaea', levels[l]]
  def_arc = data.loc['PICRUSt2 default Archaea', levels[l]]
  # tx = ax1.text(upd_bac+upd_bac*multiply, x1[l], str(upd_bac), ha='left', va='center')
  # tx = ax1.text(def_bac+def_bac*multiply, x2[l], str(def_bac), ha='left', va='center')
  # tx = ax2.text(upd_arc+upd_arc*multiply, x1[l], str(upd_arc), ha='left', va='center')
  # tx = ax2.text(def_arc+def_arc*multiply, x2[l], str(def_arc), ha='left', va='center')
  tx = ax1.text(upd_bac+200, x1[l], str(upd_bac), ha='left', va='center')
  tx = ax1.text(def_bac+200, x2[l], str(def_bac), ha='left', va='center')
  tx = ax2.text(upd_arc+10, x1[l], str(upd_arc), ha='left', va='center')
  tx = ax2.text(def_arc+10, x2[l], str(def_arc), ha='left', va='center')


databases = ['Default PICRUSt2', 'Updated PICRUSt2']
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
#legend = ax1.legend(handles=handles, loc='upper right')
legend = ax2.legend(handles=handles, loc='upper right')

levels.reverse()
plt.sca(ax1)
yt = plt.yticks([locs[l] for l in range(len(levels))], levels)
xl = plt.xlabel('Number of taxa', fontweight='bold')
ti = plt.title('Bacteria', fontweight='bold')
#plt.semilogx()
#xt = plt.xlim([1e-1, 1e5])
xl = plt.xlim([-400, 31000])
# 
plt.sca(ax2)
yt = plt.yticks([locs[l] for l in range(len(levels))], [], rotation=45)
xl = plt.xlabel('Number of taxa', fontweight='bold')
ti = plt.title('Archaea', fontweight='bold')
#plt.semilogx()
#xt = plt.xlim([1e-1, 1e4])
xl = plt.xlim([-15, 1150])
# 
plt.subplots_adjust(wspace=0.1)
plt.savefig(folder+'figures/number_of_taxa_horizontal.png', dpi=600, bbox_inches='tight')
```

## Comparison of NSTI

```{python}
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
for ds in datasets:
  correct_domain, incorrect_domain = 0, 0
  df = pd.read_csv('/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/nsti_comparison/'+ds+'_bac_arc.csv', index_col=0, header=0)
  tax = pd.read_csv('/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/qiime2_classification/'+ds+'/taxonomy.tsv', index_col=0, header=0, sep='\t')
  df = df.loc[:, ['Placed in bacteria tree', 'Placed in archaea tree', 'Placed in default tree', 'bac_mp_nsti', 'arc_mp_nsti', 'def_mp_nsti']]
  df.columns = ['Placed in bacteria tree', 'Placed in archaea tree', 'Placed in default tree', 'bacteria_nsti', 'archaea_nsti', 'default_nsti']
  df['Lowest'] = ''
  df['Lowest bacteria archaea'] = ''
  df['Taxonomy'] = ''
  df['Tax confidence'] = ''
  for row in df.index.values:
    lowest, name = 100, ''
    for col in ['bacteria_nsti', 'archaea_nsti', 'default_nsti']:
      if df.loc[row, col] < lowest:
        lowest, name = df.loc[row, col], col
    df.loc[row, 'Lowest'] = name
    lowest, name = 100, ''
    for col in ['bacteria_nsti', 'archaea_nsti']:
      if df.loc[row, col] < lowest:
        lowest, name = df.loc[row, col], col
    df.loc[row, 'Lowest bacteria archaea'] = name
    df.loc[row, 'Taxonomy'] = tax.loc[row, 'Taxon']
    df.loc[row, 'Tax confidence'] = tax.loc[row, 'Confidence']
    if name == '':
      do_nothing = True
    elif name == 'bacteria_nsti' and 'Bacteria' in tax.loc[row, 'Taxon']:
      correct_domain += 1
    elif name == 'archaea_nsti' and 'Archaea' in tax.loc[row, 'Taxon']:
      correct_domain += 1
    else:
      incorrect_domain += 1
      print(row, df.loc[[row], ['bacteria_nsti', 'archaea_nsti', 'Taxonomy']].values)
  print(ds, correct_domain, incorrect_domain)
  df.to_csv('/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/nsti_comparison/'+ds+'_bac_arc_lowest.csv')
    
```

### Plot

Bacteria and archaea for new:
```{python}
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
x = [[0, 1, 2], [4, 5, 6], [8, 9, 10], [12, 13, 14], [16, 17, 18], [20, 21, 22], [24, 25, 26]]
fig = plt.figure(figsize=(15,5))
ax = plt.subplot(111)

c1 = 0
for ds in datasets:
  #if ds != 'cameroon': continue
  df = pd.read_csv('/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/nsti_comparison/'+ds+'_bac_arc_lowest.csv', index_col=0, header=0)
  nsti_bac, nsti_arc, nsti_def = df['bacteria_nsti'].values, df['archaea_nsti'].values, df['default_nsti'].values
  nsti_bac, nsti_arc, nsti_def = [n for n in nsti_bac if not math.isnan(n)], [n for n in nsti_arc if not math.isnan(n)], [n for n in nsti_def if not math.isnan(n)]
  nsti_bac, nsti_arc, nsti_def = [n for n in nsti_bac if not math.isnan(n)], [n for n in nsti_arc if not math.isnan(n)], [n for n in nsti_def if not math.isnan(n)]
  vals = [nsti_bac, nsti_arc, nsti_def]
  c2 = 0
  for v in vals:
    sc = ax.scatter(np.random.normal(x[c1][c2], scale=0.1, size=len(v)), v, marker='o', color=colours[c2+2], alpha=0.1)
    c2 += 1
  box = plt.boxplot(vals, positions=x[c1], widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  c2 = 0
  for v in vals:
    sc = ax.scatter(x[c1][c2], np.mean(v), marker='*', color='k')
    c2 += 1
  c1 += 1

yl = plt.ylabel('Nearest Sequenced Taxon Index', fontweight='bold')
xt = plt.xticks([xv[1] for xv in x], datasets)
yl = plt.ylim([-0.2, 10])

databases = ['Updated bacteria', 'Updated archaea', 'Default']
cols = [colours[2], colours[3], colours[4]]
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=cols[s], markersize=12) for s in range(len(databases))]
lg = plt.legend(handles=handles, loc='upper left')
  
plt.savefig(folder+'figures/nsti_bacteria_and_archaea.png', dpi=600, bbox_inches='tight')
```

Best of bacteria and archaea for new:
```{python}
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
x = [[0, 1], [3, 4], [6, 7], [9, 10], [12, 13], [15, 16], [18, 19]]
fig = plt.figure(figsize=(10,5))
ax = plt.subplot(111)

c1 = 0
for ds in datasets:
  #if ds != 'cameroon': continue
  df = pd.read_csv('/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/nsti_comparison/'+ds+'_bac_arc_lowest.csv', index_col=0, header=0)
  df['min_new'] = ''
  for row in df.index.values:
    nsti_bac, nsti_arc = df.loc[row, 'bacteria_nsti'], df.loc[row, 'archaea_nsti']
    if math.isnan(nsti_bac):
      if not math.isnan(nsti_arc):
        df.loc[row, 'min_new'] = nsti_arc
    elif math.isnan(nsti_arc):
      if not math.isnan(nsti_bac):
        df.loc[row, 'min_new'] = nsti_bac
    if nsti_bac < nsti_arc: df.loc[row, 'min_new'] = nsti_bac
    elif nsti_arc < nsti_bac: df.loc[row, 'min_new'] = nsti_arc
  nsti_bac = df['bacteria_nsti'].values
  nsti_new, nsti_def = df['min_new'].values, df['default_nsti'].values
  nsti_bac, nsti_new, nsti_def = [n for n in nsti_bac if not isinstance(n, str)], [n for n in nsti_new if not isinstance(n, str)], [n for n in nsti_def if not isinstance(n, str)]
  nsti_bac, nsti_new, nsti_def = [n for n in nsti_bac if not math.isnan(n)], [n for n in nsti_new if not math.isnan(n)], [n for n in nsti_def if not math.isnan(n)]
  vals = [nsti_def, nsti_new]
  c2 = 0
  for v in vals:
    sc = ax.scatter(np.random.normal(x[c1][c2], scale=0.1, size=len(v)), v, marker='o', color=colours[c2], alpha=0.1)
    c2 += 1
  box = plt.boxplot(vals, positions=x[c1], widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  c2 = 0
  for v in vals:
    sc = ax.scatter(x[c1][c2], np.mean(v), marker='*', color='k')
    c2 += 1
  c1 += 1

yl = plt.ylabel('Nearest Sequenced Taxon Index', fontweight='bold')
xt = plt.xticks([np.mean(xv) for xv in x], datasets)
yl = plt.ylim([-0.1, 2])

databases = ['Default PICRUSt2', 'Updated PICRUSt2']
cols = [colours[0], colours[1]]
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=cols[s], markersize=12) for s in range(len(databases))]
lg = plt.legend(handles=handles, loc='upper left')
  
#plt.show()
plt.savefig(folder+'figures/nsti_best.png', dpi=600, bbox_inches='tight')
```

## Correlation between old and new

Separate:
```{python}
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/'
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']

for trait in ['EC', 'KO', 'path']:
  #if trait != 'path': continue
  all_correlations = []
  for ds in datasets:
    #if ds != 'blueberry': break
    if trait != 'path':
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      rename = {}
      for row in new.index.values:
        if trait == 'EC': rename[row] = 'EC:'+row
        else: rename[row] = row.split(':')[1]
      new = new.rename(index=rename)
    else:
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/pathways_out/path_abun_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
    intersection = [trait for trait in old.index.values if trait in new.index.values]
    #print(new.shape[0], old.shape[0], len(intersection))
    new = new.loc[intersection, :]
    old = old.loc[intersection, :]
    correlations = []
    for sample in new.columns:
      new_vals = new.loc[:, sample].values
      old_vals = old.loc[:, sample].values
      stat, p = spearmanr(new_vals, old_vals)
      correlations.append(stat)
    all_correlations.append(correlations)
  plt.close()
  fig = plt.figure(figsize=(7,5))
  ax = plt.subplot(111)
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(d, scale=0.1, size=len(all_correlations[d])), all_correlations[d], color=colours[0], alpha=0.5)
  box = ax.boxplot(all_correlations, positions=[d for d in range(len(datasets))], widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xt = plt.xticks([d for d in range(len(datasets))], datasets)
  yl = plt.ylabel('Spearman correlation coefficient')
  ti = plt.title(trait.replace('path', 'Pathways').replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  plt.savefig(folder+'figures/correlation_'+trait+'.png', dpi=600, bbox_inches='tight')
        

```

Together:
```{python}
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/'
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']

plt.close()
fig = plt.figure(figsize=(10,5))
ax1 = plt.subplot(131)
ax2, ax3 = plt.subplot(132), plt.subplot(133)
axes = {'EC':ax1, 'KO':ax2, 'path':ax3}

for trait in ['EC', 'KO', 'path']:
  #if trait != 'path': continue
  all_correlations = []
  for ds in datasets:
    #if ds != 'blueberry': break
    if trait != 'path':
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      rename = {}
      for row in new.index.values:
        if trait == 'EC': rename[row] = 'EC:'+row
        else: rename[row] = row.split(':')[1]
      new = new.rename(index=rename)
    else:
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/pathways_out/path_abun_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
    intersection = [trait for trait in old.index.values if trait in new.index.values]
    #print(new.shape[0], old.shape[0], len(intersection))
    new = new.loc[intersection, :]
    old = old.loc[intersection, :]
    correlations = []
    for sample in new.columns:
      new_vals = new.loc[:, sample].values
      old_vals = old.loc[:, sample].values
      stat, p = spearmanr(new_vals, old_vals)
      correlations.append(stat)
      if p >= 0.001: print('This wasnt significant!!: '+ds+', '+sample)
    print(ds, trait, np.median(correlations))
    all_correlations.append(correlations)
  for d in range(len(datasets)):
    #if d > 4: continue
    sc = axes[trait].scatter(np.random.normal(d, scale=0.1, size=len(all_correlations[d])), all_correlations[d], color=colours[2], alpha=0.5)
  box = axes[trait].boxplot(all_correlations, positions=[d for d in range(len(datasets))], widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  plt.sca(axes[trait])
  xt = plt.xticks([d for d in range(len(datasets))], datasets, rotation=45)
  if trait == 'EC': yl = plt.ylabel('Spearman correlation coefficient')
  ti = plt.title(trait.replace('path', 'Pathways').replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  yl = plt.ylim([0.745, 0.955])
  if trait != 'EC': yt = axes[trait].set_yticklabels([])
  print('\n\n')

plt.savefig(folder+'figures/correlation_all.png', dpi=600, bbox_inches='tight')
        

```

## Correlation with MGS

Separate:
```{python}
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/'
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']

for trait in ['EC', 'KO', 'path']:
  #if trait != 'path': continue
  all_correlations = []
  for ds in datasets:
    #if ds != 'blueberry': break
    if trait != 'path':
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      rename = {}
      for row in new.index.values:
        if trait == 'EC': rename[row] = 'EC:'+row
        else: rename[row] = row.split(':')[1]
      new = new.rename(index=rename)
      mgs = pd.read_csv(folder+'/mgs_validation/'+ds+'/humann2_'+trait.lower()+'_unstrat.tsv', sep='\t', index_col=0, header=0)
    else:
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/pathways_out/path_abun_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_path_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      mgs = pd.read_csv(folder+'/mgs_validation/'+ds+'/humann2_pathabun_unstrat.tsv', sep='\t', index_col=0, header=0)
    samples = [s for s in new.columns if s in old.columns and s in mgs.columns]
    # print(len(samples), new.shape[1])
    comparisons = [old, new]
    names = ['Default', 'Updated']
    ds_correlations = []
    for c in range(len(comparisons)):
      comp = comparisons[c]
      intersection = [trait for trait in comp.index.values if trait in mgs.index.values]
      comp = comp.loc[intersection, :]
      mgs_comp = mgs.copy(deep=True)
      mgs_comp = mgs_comp.loc[intersection, :]
      correlations = []
      for sample in samples:
        comp_vals = comp.loc[:, sample].values
        mgs_vals = mgs_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mgs_vals)
        if p >= 0.01: print("This comparison wasn't significant!!: "+trait+", "+ds+", "+sample+", "+names[c])
        correlations.append(stat)
      ds_correlations.append(correlations)
    all_correlations.append(ds_correlations)
  plt.close()
  fig = plt.figure(figsize=(10,5))
  ax = plt.subplot(111)
  x = [0, 1]
  xlocs = []
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5)
    sc = ax.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5)
    box = ax.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  xt = plt.xticks(xlocs, datasets)
  yl = plt.ylabel('Spearman correlation coefficient')
  ti = plt.title(trait.replace('path', 'Pathways').replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  legend = ax.legend(handles=handles, loc='lower right')
  plt.savefig(folder+'figures/correlation_'+trait+'_mgs.png', dpi=600, bbox_inches='tight')
        

```

Combined:
```{python}
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/'
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']

fig = plt.figure(figsize=(12,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
axes = {'EC':ax1, 'KO':ax2}

for trait in ['EC', 'KO']:
  #if trait != 'path': continue
  all_correlations = []
  for ds in datasets:
    #if ds != 'blueberry': break
    if trait != 'path':
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      rename = {}
      for row in new.index.values:
        if trait == 'EC': rename[row] = 'EC:'+row
        else: rename[row] = row.split(':')[1]
      new = new.rename(index=rename)
      mgs = pd.read_csv(folder+'/mgs_validation/'+ds+'/humann2_'+trait.lower()+'_unstrat.tsv', sep='\t', index_col=0, header=0)
    else:
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/pathways_out/path_abun_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_path_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      mgs = pd.read_csv(folder+'/mgs_validation/'+ds+'/humann2_pathabun_unstrat.tsv', sep='\t', index_col=0, header=0)
    samples = [s for s in new.columns if s in old.columns and s in mgs.columns]
    # print(len(samples), new.shape[1])
    comparisons = [old, new]
    names = ['Default', 'Updated']
    ds_correlations = []
    for c in range(len(comparisons)):
      comp = comparisons[c]
      intersection = [trait for trait in comp.index.values if trait in mgs.index.values]
      comp = comp.loc[intersection, :]
      mgs_comp = mgs.copy(deep=True)
      mgs_comp = mgs_comp.loc[intersection, :]
      correlations = []
      for sample in samples:
        comp_vals = comp.loc[:, sample].values
        mgs_vals = mgs_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mgs_vals)
        if p >= 0.01: print("This comparison wasn't significant!!: "+trait+", "+ds+", "+sample+", "+names[c])
        correlations.append(stat)
      ds_correlations.append(correlations)
    all_correlations.append(ds_correlations)
  ax = axes[trait]
  plt.sca(ax)
  x = [0, 1]
  xlocs = []
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5)
    sc = ax.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5)
    box = ax.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  xt = plt.xticks(xlocs, datasets)
  if trait == 'EC': yl = plt.ylabel('Spearman correlation coefficient', fontweight='bold')
  ti = plt.title(trait.replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  if trait == 'KO': legend = ax.legend(handles=handles, loc='lower right')

ti = ax1.set_title('A', loc='left', fontweight='bold'), ax2.set_title('B', loc='left', fontweight='bold')

plt.savefig(folder+'figures/correlation_combined_mgs.png', dpi=600, bbox_inches='tight')
```


## Distances between old and new

Separate:
```{python}
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/'
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']

for trait in ['EC', 'KO', 'path']:
  #if trait != 'path': continue
  all_correlations = []
  for ds in datasets:
    #if ds != 'blueberry': break
    if trait != 'path':
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      rename = {}
      for row in new.index.values:
        if trait == 'EC': rename[row] = 'EC:'+row
        else: rename[row] = row.split(':')[1]
      new = new.rename(index=rename)
    else:
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/pathways_out/path_abun_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
    intersection = [trait for trait in old.index.values if trait in new.index.values]
    #print(new.shape[0], old.shape[0], len(intersection))
    new = new.loc[intersection, :]
    old = old.loc[intersection, :]
    correlations = []
    for sample in new.columns:
      combined = pd.concat([new.loc[:, [sample]].rename(columns={sample:'New'}), old.loc[:, [sample]].rename(columns={sample:'Old'})]).fillna(value=0)
      combined = combined.groupby(by=combined.index, axis=0).sum()
      X = combined.transpose().iloc[0:].values
      bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
      # new_vals = new.loc[:, .values
      # old_vals = old.loc[:, sample].values
      # stat, p = spearmanr(new_vals, old_vals)
      correlations.append(bc_dist)
    all_correlations.append(correlations)
  plt.close()
  fig = plt.figure(figsize=(7,5))
  ax = plt.subplot(111)
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(d, scale=0.1, size=len(all_correlations[d])), all_correlations[d], color=colours[0], alpha=0.5)
  box = ax.boxplot(all_correlations, positions=[d for d in range(len(datasets))], widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xt = plt.xticks([d for d in range(len(datasets))], datasets)
  yl = plt.ylabel('Bray-Curtis distance')
  ti = plt.title(trait.replace('path', 'Pathways').replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  plt.savefig(folder+'figures/distance_braycurtis_'+trait+'.png', dpi=600, bbox_inches='tight')
        

```

Together:
```{python}
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/'
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']

plt.close()
fig = plt.figure(figsize=(10,5))
ax1 = plt.subplot(131)
ax2, ax3 = plt.subplot(132), plt.subplot(133)
axes = {'EC':ax1, 'KO':ax2, 'path':ax3}

for trait in ['EC', 'KO', 'path']:
  #if trait != 'path': continue
  all_correlations = []
  for ds in datasets:
    #if ds != 'blueberry': break
    if trait != 'path':
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      rename = {}
      for row in new.index.values:
        if trait == 'EC': rename[row] = 'EC:'+row
        else: rename[row] = row.split(':')[1]
      new = new.rename(index=rename)
    else:
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/pathways_out/path_abun_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
    intersection = [trait for trait in old.index.values if trait in new.index.values]
    #print(new.shape[0], old.shape[0], len(intersection))
    new = new.loc[intersection, :]
    old = old.loc[intersection, :]
    correlations = []
    for sample in new.columns:
      # new_vals = new.loc[:, sample].values
      # old_vals = old.loc[:, sample].values
      # stat, p = spearmanr(new_vals, old_vals)
      combined = pd.concat([new.loc[:, [sample]].rename(columns={sample:'New'}), old.loc[:, [sample]].rename(columns={sample:'Old'})]).fillna(value=0)
      combined = combined.groupby(by=combined.index.values).sum()
      X = combined.transpose().iloc[0:].values
      bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
      correlations.append(bc_dist)
    #print(ds, trait, np.median(correlations))
    all_correlations.append(correlations)
  for d in range(len(datasets)):
    sc = axes[trait].scatter(np.random.normal(d, scale=0.1, size=len(all_correlations[d])), all_correlations[d], color=colours[2], alpha=0.5)
  box = axes[trait].boxplot(all_correlations, positions=[d for d in range(len(datasets))], widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  plt.sca(axes[trait])
  xt = plt.xticks([d for d in range(len(datasets))], datasets, rotation=45)
  if trait == 'EC': yl = plt.ylabel('Bray-Curtis distance')
  ti = plt.title(trait.replace('path', 'Pathways').replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  #yl = plt.ylim([0.745, 0.955])
  if trait != 'EC': yt = axes[trait].set_yticklabels([])
  print('\n\n')

plt.savefig(folder+'figures/distance_braycurtis_all.png', dpi=600, bbox_inches='tight')
        

```

Combined:
```{python}
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/'
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']

fig = plt.figure(figsize=(12,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
axes = {'EC':ax1, 'KO':ax2}

for trait in ['EC', 'KO']:
  #if trait != 'path': continue
  all_correlations = []
  for ds in datasets:
    #if ds != 'blueberry': break
    if trait != 'path':
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      rename = {}
      for row in new.index.values:
        if trait == 'EC': rename[row] = 'EC:'+row
        else: rename[row] = row.split(':')[1]
      new = new.rename(index=rename)
      mgs = pd.read_csv(folder+'/mgs_validation/'+ds+'/humann2_'+trait.lower()+'_unstrat.tsv', sep='\t', index_col=0, header=0)
    else:
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/pathways_out/path_abun_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_path_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      mgs = pd.read_csv(folder+'/mgs_validation/'+ds+'/humann2_pathabun_unstrat.tsv', sep='\t', index_col=0, header=0)
    samples = [s for s in new.columns if s in old.columns and s in mgs.columns]
    # print(len(samples), new.shape[1])
    comparisons = [old, new]
    names = ['Default', 'Updated']
    ds_correlations = []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      comp = comp.divide(comp.sum(axis=0), axis=1).multiply(100)
      intersection = [trait for trait in comp.index.values if trait in mgs.index.values]
      #comp = comp.loc[intersection, :]
      mgs_comp = mgs.copy(deep=True)
      mgs_comp = mgs_comp.divide(mgs_comp.sum(axis=0), axis=1).multiply(100)
      #mgs_comp = mgs_comp.loc[intersection, :]
      correlations = []
      for sample in samples:
        # comp_vals = comp.loc[:, sample].values
        # mgs_vals = mgs_comp.loc[:, sample].values
        # stat, p = spearmanr(comp_vals, mgs_vals)
        # if p >= 0.01: print("This comparison wasn't significant!!: "+trait+", "+ds+", "+sample+", "+names[c])
        combined = pd.concat([comp.loc[:, [sample]].rename(columns={sample:'Comp'}), mgs_comp.loc[:, [sample]].rename(columns={sample:'MGS'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        correlations.append(bc_dist)
      ds_correlations.append(correlations)
    all_correlations.append(ds_correlations)
  ax = axes[trait]
  plt.sca(ax)
  x = [0, 1]
  xlocs = []
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5)
    sc = ax.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5)
    box = ax.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  xt = plt.xticks(xlocs, datasets)
  if trait == 'EC': yl = plt.ylabel('Bray-Curtis distance', fontweight='bold')
  #ti = plt.title(trait.replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  #if trait == 'KO': legend = ax.legend(handles=handles, loc='lower right')

ti = ax1.set_title('C', loc='left', fontweight='bold'), ax2.set_title('D', loc='left', fontweight='bold')

plt.savefig(folder+'figures/distance_braycurtis_combined_mgs.png', dpi=600, bbox_inches='tight')
        

```

## Distance with MGS

Separate:
```{python}
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/'
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']

for trait in ['EC', 'KO', 'path']:
  #if trait != 'path': continue
  all_correlations = []
  for ds in datasets:
    #if ds != 'blueberry': break
    if trait != 'path':
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_'+trait.lower()+'_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      rename = {}
      for row in new.index.values:
        if trait == 'EC': rename[row] = 'EC:'+row
        else: rename[row] = row.split(':')[1]
      new = new.rename(index=rename)
      mgs = pd.read_csv(folder+'/mgs_validation/'+ds+'/humann2_'+trait.lower()+'_unstrat.tsv', sep='\t', index_col=0, header=0)
    else:
      new = pd.read_csv(folder+'picrust2_out_split_'+ds+'/pathways_out/path_abun_unstrat.tsv', sep='\t', index_col=0, header=0)
      old = pd.read_csv(folder+'picrust2_default/'+ds+'_picrust2_path_nsti2.0.tsv', sep='\t', index_col=0, header=0)
      mgs = pd.read_csv(folder+'/mgs_validation/'+ds+'/humann2_pathabun_unstrat.tsv', sep='\t', index_col=0, header=0)
    samples = [s for s in new.columns if s in old.columns and s in mgs.columns]
    # print(len(samples), new.shape[1])
    comparisons = [old, new]
    names = ['Default', 'Updated']
    ds_correlations = []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      comp = comp.divide(comp.sum(axis=0), axis=1).multiply(100)
      intersection = [trait for trait in comp.index.values if trait in mgs.index.values]
      #comp = comp.loc[intersection, :]
      mgs_comp = mgs.copy(deep=True)
      mgs_comp = mgs_comp.divide(mgs_comp.sum(axis=0), axis=1).multiply(100)
      #mgs_comp = mgs_comp.loc[intersection, :]
      correlations = []
      for sample in samples:
        combined = pd.concat([comp.loc[:, [sample]].rename(columns={sample:'Comp'}), mgs_comp.loc[:, [sample]].rename(columns={sample:'MGS'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        # comp_vals = comp.loc[:, sample].values
        # mgs_vals = mgs_comp.loc[:, sample].values
        # stat, p = spearmanr(comp_vals, mgs_vals)
        # if p >= 0.01: print("This comparison wasn't significant!!: "+trait+", "+ds+", "+sample+", "+names[c])
        correlations.append(bc_dist)
      ds_correlations.append(correlations)
    all_correlations.append(ds_correlations)
  plt.close()
  fig = plt.figure(figsize=(10,5))
  ax = plt.subplot(111)
  x = [0, 1]
  xlocs = []
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5)
    sc = ax.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5)
    box = ax.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  xt = plt.xticks(xlocs, datasets)
  yl = plt.ylabel('Bray-Curtis distance')
  ti = plt.title(trait.replace('path', 'Pathways').replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  legend = ax.legend(handles=handles, loc='lower right')
  plt.savefig(folder+'figures/distance_braycurtis_'+trait+'_mgs.png', dpi=600, bbox_inches='tight')

```

## Comparison with mock samples

Unzip files:
```{bash, eval=FALSE}
gunzip picrust2_out/*/EC_metagenome_out/pred_metagenome_unstrat.tsv.gz
gunzip picrust2_out/*/EC_metagenome_out/weighted_nsti.tsv.gz
gunzip picrust2_out/*/KO_metagenome_out/pred_metagenome_unstrat.tsv.gz
gunzip picrust2_out/*/KO_metagenome_out/weighted_nsti.tsv.gz
#rename all KEGG_ko to KO in feature_tables
gunzip picrust2_out_trim/*/EC_metagenome_out/pred_metagenome_unstrat.tsv.gz
gunzip picrust2_out_trim/*/EC_metagenome_out/weighted_nsti.tsv.gz
gunzip picrust2_out_trim/*/KO_metagenome_out/pred_metagenome_unstrat.tsv.gz
gunzip picrust2_out_trim/*/KO_metagenome_out/weighted_nsti.tsv.gz

gunzip mgs_test/*/EC_metagenome_out/pred_metagenome_unstrat.tsv.gz
gunzip mgs_test/*/EC_metagenome_out/weighted_nsti.tsv.gz
gunzip mgs_test/*/KO_metagenome_out/pred_metagenome_unstrat.tsv.gz
gunzip mgs_test/*/KO_metagenome_out/weighted_nsti.tsv.gz
gunzip mgs_test/*default*/marker_predicted_and_nsti.tsv.gz
```

### Taxa in mock samples

```{python}
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/'
tax = pd.read_csv(folder+'genome_full_tax.csv', index_col=0, header=0)

tax_genus, tax_class, tax_family = {}, {}, {}
for row in tax.index.values:
  this_tax = tax.loc[row, 'Taxonomy'].split(';')
  tax_genus[row] = this_tax[-2]
  tax_family[row] = this_tax[-3]
  tax_class[row] = this_tax[2]

combined = []
for ds in datasets:
  ft = pd.read_csv(folder+'mock_test/feature_tables/'+ds+'_16S.csv', index_col=0, header=0).drop('Tax', axis=1)
  ft = ft.rename(index=tax_class)
  #ft = ft.rename(index=tax_genus)
  #ft = ft.rename(index=tax_family)
  ft = ft.groupby(by=ft.index).sum()
  if 'Unassigned' in ft.index.values: ft = ft.drop('Unassigned')
  ft = ft.divide(ft.sum(axis=0), axis=1).multiply(100)
  ft[ds] = ft.mean(axis=1)
  combined.append(ft.loc[:, [ds]])

combined_df = pd.concat(combined).fillna(value=0)
combined_df = combined_df.groupby(by=combined_df.index).sum()
combined_df['Mean'] = combined_df.mean(axis=1)
combined_df = combined_df.sort_values(by=['Mean'], ascending=False)
combined_df = combined_df.iloc[:19, :]
combined_df = combined_df.drop('Mean', axis=1)
combined_df = combined_df.loc[sorted(combined_df.index.values), :]
combined_df.loc['Other', :] = 100-combined_df.sum(axis=0).values
combined_df.transpose().plot(kind='bar', stacked=True, figsize=(8,5), width=0.9, cmap='tab20', edgecolor='k')
plt.legend(loc='upper left', bbox_to_anchor=(1.01, 1.02))
yl = plt.ylabel('Relative abundance (%)')
yl = plt.ylim([0, 100])
plt.savefig(folder+'figures/relative_abundance_class.png', dpi=600, bbox_inches='tight')
#plt.savefig('/Users/robynwright/Dropbox/Papers_writing/PICRUSt2 database application note/figures/relative_abundance_genus.png', dpi=600, bbox_inches='tight')
#plt.savefig('/Users/robynwright/Dropbox/Papers_writing/PICRUSt2 database application note/figures/relative_abundance_family.png', dpi=600, bbox_inches='tight')
```

### Distances

```{python}
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/mock_test/'

for trait in ['EC', 'KO']:
  all_correlations = []
  for ds in datasets:
    mock = pd.read_csv(folder+'feature_tables/'+ds+'_mock_'+trait+'_not_16S_norm_metagenome.csv', index_col=0, header=0)
    if '-' in mock.index.values: mock = mock.drop('-', axis=0)
    old = pd.read_csv(folder+'picrust2_out/picrust2_out_default_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    rename = {}
    for row in old.index.values:
      if trait == 'EC': rename[row] = row.split(':')[1]
      else: rename[row] = 'ko:'+row
    old = old.rename(index=rename)
    new = pd.read_csv(folder+'picrust2_out/picrust2_out_updated_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    samples = [s for s in new.columns if s in old.columns and s in mock.columns]
    comparisons = [old, new]
    names = ['Default', 'Updated']
    ds_correlations = []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      comp = comp.divide(comp.sum(axis=0), axis=1).multiply(100)
      #intersection = [trait for trait in comp.index.values if trait in mgs.index.values]
      #comp = comp.loc[intersection, :]
      mock_comp = mock.copy(deep=True)
      mock_comp = mock_comp.divide(mock_comp.sum(axis=0), axis=1).multiply(100)
      #mgs_comp = mgs_comp.loc[intersection, :]
      correlations = []
      for sample in samples:
        combined = pd.concat([comp.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        # comp_vals = comp.loc[:, sample].values
        # mgs_vals = mgs_comp.loc[:, sample].values
        # stat, p = spearmanr(comp_vals, mgs_vals)
        # if p >= 0.01: print("This comparison wasn't significant!!: "+trait+", "+ds+", "+sample+", "+names[c])
        correlations.append(bc_dist)
      ds_correlations.append(correlations)
    all_correlations.append(ds_correlations)
  plt.close()
  fig = plt.figure(figsize=(10,5))
  ax = plt.subplot(111)
  x = [0, 1]
  xlocs = []
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5)
    sc = ax.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5)
    box = ax.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  xt = plt.xticks(xlocs, datasets)
  yl = plt.ylabel('Bray-Curtis distance')
  ti = plt.title(trait.replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  legend = ax.legend(handles=handles, loc='lower right')
  plt.savefig(folder+'figures/distance_braycurtis_'+trait+'_mock.png', dpi=600, bbox_inches='tight')
```

Combined figure:
```{python}
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/mock_test/'

fig = plt.figure(figsize=(12,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
axes = {'EC':ax1, 'KO':ax2}

for trait in ['EC', 'KO']:
  all_correlations = []
  for ds in datasets:
    mock = pd.read_csv(folder+'feature_tables/'+ds+'_mock_'+trait+'_not_16S_norm_metagenome.csv', index_col=0, header=0)
    if '-' in mock.index.values: mock = mock.drop('-', axis=0)
    old = pd.read_csv(folder+'picrust2_out/picrust2_out_default_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    rename = {}
    for row in old.index.values:
      if trait == 'EC': rename[row] = row.split(':')[1]
      else: rename[row] = 'ko:'+row
    old = old.rename(index=rename)
    new = pd.read_csv(folder+'picrust2_out/picrust2_out_updated_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    samples = [s for s in new.columns if s in old.columns and s in mock.columns]
    comparisons = [old, new]
    names = ['Default', 'Updated']
    ds_correlations = []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      comp = comp.divide(comp.sum(axis=0), axis=1).multiply(100)
      #intersection = [trait for trait in comp.index.values if trait in mgs.index.values]
      #comp = comp.loc[intersection, :]
      mock_comp = mock.copy(deep=True)
      mock_comp = mock_comp.divide(mock_comp.sum(axis=0), axis=1).multiply(100)
      #mgs_comp = mgs_comp.loc[intersection, :]
      correlations = []
      for sample in samples:
        combined = pd.concat([comp.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        # comp_vals = comp.loc[:, sample].values
        # mgs_vals = mgs_comp.loc[:, sample].values
        # stat, p = spearmanr(comp_vals, mgs_vals)
        # if p >= 0.01: print("This comparison wasn't significant!!: "+trait+", "+ds+", "+sample+", "+names[c])
        correlations.append(bc_dist)
      ds_correlations.append(correlations)
    all_correlations.append(ds_correlations)
  #plt.close()
  ax = axes[trait]
  plt.sca(ax)
  x = [0, 1]
  xlocs = []
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5)
    sc = ax.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5)
    box = ax.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  xt = plt.xticks(xlocs, datasets)
  if trait == 'EC': yl = plt.ylabel('Bray-Curtis distance', fontweight='bold')
  #ti = plt.title(trait.replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  #if trait == 'KO': legend = ax.legend(handles=handles, loc='upper right')

ti = ax1.set_title('C', loc='left', fontweight='bold'), ax2.set_title('D', loc='left', fontweight='bold')

plt.savefig(folder+'figures/distance_braycurtis_combined_mock.png', dpi=600, bbox_inches='tight')
```

#### V4-V5

```{python}
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/mock_test/'

fig = plt.figure(figsize=(12,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
axes = {'EC':ax1, 'KO':ax2}

for trait in ['EC', 'KO']:
  all_correlations = []
  for ds in datasets:
    mock = pd.read_csv(folder+'feature_tables/'+ds+'_mock_'+trait+'_not_16S_norm_metagenome.csv', index_col=0, header=0)
    if '-' in mock.index.values: mock = mock.drop('-', axis=0)
    old = pd.read_csv(folder+'picrust2_out_trim/picrust2_out_default_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    rename = {}
    for row in old.index.values:
      if trait == 'EC': rename[row] = row.split(':')[1]
      else: rename[row] = 'ko:'+row
    old = old.rename(index=rename)
    new = pd.read_csv(folder+'picrust2_out_trim/picrust2_out_updated_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    samples = [s for s in new.columns if s in old.columns and s in mock.columns]
    comparisons = [old, new]
    names = ['Default', 'Updated']
    ds_correlations = []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      comp = comp.divide(comp.sum(axis=0), axis=1).multiply(100)
      #intersection = [trait for trait in comp.index.values if trait in mgs.index.values]
      #comp = comp.loc[intersection, :]
      mock_comp = mock.copy(deep=True)
      mock_comp = mock_comp.divide(mock_comp.sum(axis=0), axis=1).multiply(100)
      #mgs_comp = mgs_comp.loc[intersection, :]
      correlations = []
      for sample in samples:
        combined = pd.concat([comp.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        # comp_vals = comp.loc[:, sample].values
        # mgs_vals = mgs_comp.loc[:, sample].values
        # stat, p = spearmanr(comp_vals, mgs_vals)
        # if p >= 0.01: print("This comparison wasn't significant!!: "+trait+", "+ds+", "+sample+", "+names[c])
        correlations.append(bc_dist)
      ds_correlations.append(correlations)
    all_correlations.append(ds_correlations)
  #plt.close()
  ax = axes[trait]
  plt.sca(ax)
  x = [0, 1]
  xlocs = []
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5)
    sc = ax.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5)
    box = ax.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  xt = plt.xticks(xlocs, datasets)
  if trait == 'EC': yl = plt.ylabel('Bray-Curtis distance', fontweight='bold')
  #ti = plt.title(trait.replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  #if trait == 'KO': legend = ax.legend(handles=handles, loc='upper right')

ti = ax1.set_title('C', loc='left', fontweight='bold'), ax2.set_title('D', loc='left', fontweight='bold')

plt.savefig(folder+'figures/distance_braycurtis_combined_mock_trimmed.png', dpi=600, bbox_inches='tight')
```

### Correlations

```{python}
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/mock_test/'

for trait in ['EC', 'KO']:
  all_correlations = []
  for ds in datasets:
    mock = pd.read_csv(folder+'feature_tables/'+ds+'_mock_'+trait+'_not_16S_norm_metagenome.csv', index_col=0, header=0)
    if '-' in mock.index.values: mock = mock.drop('-', axis=0)
    old = pd.read_csv(folder+'picrust2_out/picrust2_out_default_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    rename = {}
    for row in old.index.values:
      if trait == 'EC': rename[row] = row.split(':')[1]
      else: rename[row] = 'ko:'+row
    old = old.rename(index=rename)
    new = pd.read_csv(folder+'picrust2_out/picrust2_out_updated_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    samples = [s for s in new.columns if s in old.columns and s in mock.columns]
    comparisons = [old, new]
    names = ['Default', 'Updated']
    ds_correlations = []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      #comp = comp.divide(comp.sum(axis=0), axis=1).multiply(100)
      intersection = [trait for trait in comp.index.values if trait in mock.index.values]
      comp = comp.loc[intersection, :]
      mock_comp = mock.copy(deep=True)
      #mock_comp = mock_comp.divide(mock_comp.sum(axis=0), axis=1).multiply(100)
      mock_comp = mock_comp.loc[intersection, :]
      correlations = []
      for sample in samples:
        # combined = pd.concat([comp.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        # combined = combined.groupby(by=combined.index.values).sum()
        # X = combined.transpose().iloc[0:].values
        # bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mock_vals)
        if p >= 0.01: print("This comparison wasn't significant!!: "+trait+", "+ds+", "+sample+", "+names[c])
        correlations.append(stat)
      ds_correlations.append(correlations)
    all_correlations.append(ds_correlations)
  plt.close()
  fig = plt.figure(figsize=(10,5))
  ax = plt.subplot(111)
  x = [0, 1]
  xlocs = []
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5)
    sc = ax.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5)
    box = ax.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  xt = plt.xticks(xlocs, datasets)
  yl = plt.ylabel('Spearman correlation coefficient')
  ti = plt.title(trait.replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  legend = ax.legend(handles=handles, loc='lower right')
  plt.savefig(folder+'figures/correlation_'+trait+'_mock.png', dpi=600, bbox_inches='tight')
```

Combined figure:
```{python}
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/mock_test/'

fig = plt.figure(figsize=(12,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
axes = {'EC':ax1, 'KO':ax2}

for trait in ['EC', 'KO']:
  all_correlations = []
  for ds in datasets:
    mock = pd.read_csv(folder+'feature_tables/'+ds+'_mock_'+trait+'_not_16S_norm_metagenome.csv', index_col=0, header=0)
    if '-' in mock.index.values: mock = mock.drop('-', axis=0)
    old = pd.read_csv(folder+'picrust2_out/picrust2_out_default_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    rename = {}
    for row in old.index.values:
      if trait == 'EC': rename[row] = row.split(':')[1]
      else: rename[row] = 'ko:'+row
    old = old.rename(index=rename)
    new = pd.read_csv(folder+'picrust2_out/picrust2_out_updated_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    samples = [s for s in new.columns if s in old.columns and s in mock.columns]
    comparisons = [old, new]
    names = ['Default', 'Updated']
    ds_correlations = []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      #comp = comp.divide(comp.sum(axis=0), axis=1).multiply(100)
      intersection = [trait for trait in comp.index.values if trait in mock.index.values]
      comp = comp.loc[intersection, :]
      mock_comp = mock.copy(deep=True)
      #mock_comp = mock_comp.divide(mock_comp.sum(axis=0), axis=1).multiply(100)
      mock_comp = mock_comp.loc[intersection, :]
      correlations = []
      for sample in samples:
        # combined = pd.concat([comp.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        # combined = combined.groupby(by=combined.index.values).sum()
        # X = combined.transpose().iloc[0:].values
        # bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mock_vals)
        if p >= 0.01: print("This comparison wasn't significant!!: "+trait+", "+ds+", "+sample+", "+names[c])
        correlations.append(stat)
      ds_correlations.append(correlations)
    all_correlations.append(ds_correlations)
  #plt.close()
  ax = axes[trait]
  plt.sca(ax)
  x = [0, 1]
  xlocs = []
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5)
    sc = ax.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5)
    box = ax.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  xt = plt.xticks(xlocs, datasets)
  if trait == 'EC': yl = plt.ylabel('Spearman correlation coefficient', fontweight='bold')
  ti = plt.title(trait.replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  if trait == 'KO': legend = ax.legend(handles=handles, loc='upper right')

ti = ax1.set_title('A', loc='left', fontweight='bold'), ax2.set_title('B', loc='left', fontweight='bold')

plt.savefig(folder+'figures/correlation_combined_mock.png', dpi=600, bbox_inches='tight')
```

Combined figure:
```{python}
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/mock_test/'

fig = plt.figure(figsize=(12,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
axes = {'EC':ax1, 'KO':ax2}

for trait in ['EC', 'KO']:
  all_correlations = []
  for ds in datasets:
    mock = pd.read_csv(folder+'feature_tables/'+ds+'_mock_'+trait+'_not_16S_norm_metagenome.csv', index_col=0, header=0)
    if '-' in mock.index.values: mock = mock.drop('-', axis=0)
    old = pd.read_csv(folder+'picrust2_out_trim/picrust2_out_default_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    rename = {}
    for row in old.index.values:
      if trait == 'EC': rename[row] = row.split(':')[1]
      else: rename[row] = 'ko:'+row
    old = old.rename(index=rename)
    new = pd.read_csv(folder+'picrust2_out_trim/picrust2_out_updated_'+ds+'/'+trait+'_metagenome_out/pred_metagenome_unstrat.tsv', sep='\t', index_col=0, header=0)
    samples = [s for s in new.columns if s in old.columns and s in mock.columns]
    comparisons = [old, new]
    names = ['Default', 'Updated']
    ds_correlations = []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      #comp = comp.divide(comp.sum(axis=0), axis=1).multiply(100)
      intersection = [trait for trait in comp.index.values if trait in mock.index.values]
      comp = comp.loc[intersection, :]
      mock_comp = mock.copy(deep=True)
      #mock_comp = mock_comp.divide(mock_comp.sum(axis=0), axis=1).multiply(100)
      mock_comp = mock_comp.loc[intersection, :]
      correlations = []
      for sample in samples:
        # combined = pd.concat([comp.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        # combined = combined.groupby(by=combined.index.values).sum()
        # X = combined.transpose().iloc[0:].values
        # bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mock_vals)
        if p >= 0.01: print("This comparison wasn't significant!!: "+trait+", "+ds+", "+sample+", "+names[c])
        correlations.append(stat)
      ds_correlations.append(correlations)
    all_correlations.append(ds_correlations)
  #plt.close()
  ax = axes[trait]
  plt.sca(ax)
  x = [0, 1]
  xlocs = []
  for d in range(len(datasets)):
    sc = ax.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5)
    sc = ax.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5)
    box = ax.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  xt = plt.xticks(xlocs, datasets)
  if trait == 'EC': yl = plt.ylabel('Spearman correlation coefficient', fontweight='bold')
  ti = plt.title(trait.replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  if trait == 'KO': legend = ax.legend(handles=handles, loc='upper right')

ti = ax1.set_title('A', loc='left', fontweight='bold'), ax2.set_title('B', loc='left', fontweight='bold')

plt.savefig(folder+'figures/correlation_combined_mock_trimmed.png', dpi=600, bbox_inches='tight')
```

## Overall figure

```{python}
fig = plt.figure(figsize=(15,15))
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']

titles = ['Number of taxa', 'Database sizes', 'Number of genes', 'ASVs not placed in trees (%)', 'Nearest Sequenced Taxon Index', 'Performance']
ax1_bacteria = plt.subplot2grid((20,40),(0,0), rowspan=4, colspan=17)
ax1_archaea = plt.subplot2grid((20,40),(4,0), rowspan=4, colspan=17)
ax2_size_db = plt.subplot2grid((20,40),(0,21), rowspan=2, colspan=17)
ax3_num_genes = plt.subplot2grid((20,40),(3,21), rowspan=3, colspan=17)
ax4_asvs_removed = plt.subplot2grid((20,40),(9,21), rowspan=4, colspan=17)
ax6_nsti = plt.subplot2grid((20,40),(9,0), rowspan=4, colspan=17)
#ax7_legend = plt.subplot2grid((20,30),(12,21), rowspan=1, colspan=17)
#ax8_performance = plt.subplot2grid((20,40),(15,0), rowspan=5, colspan=30)
ax_performance = []
for a in range(len(datasets)):
  ax = plt.subplot2grid((20,7), (15,a), rowspan=5)
  ti = plt.title(datasets[a], fontweight='bold')
  ax_performance.append(ax)
  if a == 3:
    tx = plt.text(0.5, 1.15, 'Performance', fontweight='bold', fontsize=14, ha='center', va='center', transform=ax.transAxes)


axes = [ax1_bacteria, ax2_size_db, ax3_num_genes, ax4_asvs_removed, ax6_nsti]
for a in range(len(axes)):
  ti = axes[a].set_title(titles[a], fontweight='bold', fontsize=14)

#plot number of taxa
data = pd.read_csv(folder+'Number_of_genomes_old_and_new.csv', index_col=0, header=0)
data = data.loc[:, ['Phylum', 'Family', 'Genus', 'Genomes']]

levels = data.columns
#locs = [[20, 18.5, 17, 15.5, 14, 12.5, 11], [9, 7.5, 6, 4.5, 3, 1.5, 0]]
locs = [[20, 18.5, 17, 15.5]]

b = ax1_bacteria.barh([l-0.3 for l in locs[0]], data.loc['PICRUSt2 updated Bacteria'].values, color=colours[1], edgecolor='k', height=0.5)
b = ax1_bacteria.barh([l+0.3 for l in locs[0]], data.loc['PICRUSt2 default Bacteria'].values, color=colours[0], edgecolor='k', height=0.5)
b = ax1_archaea.barh([l-0.3 for l in locs[0]], data.loc['PICRUSt2 updated Archaea'].values, color=colours[1], edgecolor='k', height=0.5)
b = ax1_archaea.barh([l+0.3 for l in locs[0]], data.loc['PICRUSt2 default Archaea'].values, color=colours[0], edgecolor='k', height=0.5)

#adding = [10, 10, 100, 100, 100, 1000, 1000]
adding = [10, 100, 100, 1000]
for l in range(len(levels)):
  upd_bac = data.loc['PICRUSt2 updated Bacteria', levels[l]]
  def_bac = data.loc['PICRUSt2 default Bacteria', levels[l]]
  upd_arc = data.loc['PICRUSt2 updated Archaea', levels[l]]
  def_arc = data.loc['PICRUSt2 default Archaea', levels[l]]
  tx = ax1_bacteria.text(upd_bac+200, locs[0][l]-0.3, str(upd_bac), ha='left', va='center', fontsize=10)
  tx = ax1_bacteria.text(def_bac+200, locs[0][l]+0.3, str(def_bac), ha='left', va='center', fontsize=10)
  tx = ax1_archaea.text(upd_arc+10, locs[0][l]-0.3, str(upd_arc), ha='left', va='center', fontsize=10)
  tx = ax1_archaea.text(def_arc+10, locs[0][l]+0.3, str(def_arc), ha='left', va='center', fontsize=10)
  
plt.sca(ax1_bacteria)
yt = plt.yticks(locs[0], levels)
yl = plt.ylabel('Bacteria', fontweight='bold')
xt = plt.xticks([])
xl = plt.xlim([-400, 31000])

plt.sca(ax1_archaea)
yt = plt.yticks(locs[0], levels)
yl = plt.ylabel('Archaea', fontweight='bold')
xt = plt.xticks([])
xl = plt.xlim([-15, 1150])

#plot nsti
x = [[0, 1], [3, 4], [6, 7], [9, 10], [12, 13], [15, 16], [18, 19]]
fig = plt.figure(figsize=(15,5))

c1 = 0
for ds in datasets:
  #if ds != 'cameroon': continue
  df = pd.read_csv('/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/nsti_comparison/'+ds+'_bac_arc_lowest.csv', index_col=0, header=0)
  df['min_new'] = ''
  for row in df.index.values:
    nsti_bac, nsti_arc = df.loc[row, 'bacteria_nsti'], df.loc[row, 'archaea_nsti']
    if math.isnan(nsti_bac):
      if not math.isnan(nsti_arc):
        df.loc[row, 'min_new'] = nsti_arc
    elif math.isnan(nsti_arc):
      if not math.isnan(nsti_bac):
        df.loc[row, 'min_new'] = nsti_bac
    if nsti_bac < nsti_arc: df.loc[row, 'min_new'] = nsti_bac
    elif nsti_arc < nsti_bac: df.loc[row, 'min_new'] = nsti_arc
  nsti_bac = df['bacteria_nsti'].values
  nsti_new, nsti_def = df['min_new'].values, df['default_nsti'].values
  nsti_bac, nsti_new, nsti_def = [n for n in nsti_bac if not isinstance(n, str)], [n for n in nsti_new if not isinstance(n, str)], [n for n in nsti_def if not isinstance(n, str)]
  nsti_bac, nsti_new, nsti_def = [n for n in nsti_bac if not math.isnan(n)], [n for n in nsti_new if not math.isnan(n)], [n for n in nsti_def if not math.isnan(n)]
  vals = [nsti_def, nsti_new]
  c2 = 0
  for v in vals:
    sc = ax6_nsti.scatter(np.random.normal(x[c1][c2], scale=0.1, size=len(v)), v, marker='o', color=colours[c2], alpha=0.05, s=5)
    c2 += 1
  box = ax6_nsti.boxplot(vals, positions=x[c1], widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  c2 = 0
  for v in vals:
    sc = ax6_nsti.scatter(x[c1][c2], np.mean(v), marker='*', color='k')
    q75, q25 = np.percentile(v, 75), np.percentile(v, 25)
    iqr = q75-q25
    tx = ax6_nsti.text(x[c1][c2], (q75+iqr*1.5)+0.05, str(round(np.median(v), 2)), rotation=90, ha='center', va='bottom')
    print(ds, np.mean(v))
    c2 += 1
  c1 += 1

plt.sca(ax6_nsti)
yl = plt.ylabel('Nearest Sequenced Taxon Index', fontweight='bold')
xt = plt.xticks([np.mean(xv) for xv in x], datasets, rotation=45)
yl = plt.ylim([-0.1, 2])

databases = ['Default PICRUSt2', 'Updated PICRUSt2']
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
legend = ax5_num_genes.legend(handles=handles, loc='upper left', bbox_to_anchor=(-0.2,-0.4))

plt.savefig(folder+'figures/Overall_plot.png', dpi=600, bbox_inches='tight')
```

## Copy all output to put together

```{python}
import os

# copy mock full length output
# folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/mock_test/picrust2_out_trim/'
# prefix = 'V4V5'
# folders = os.listdir(folder)
# out_folder = '/Users/robynwright/Dropbox/Papers_writing/PICRUSt2_database_application_note/compile_output/mock_picrust2/'

# for fol in folders:
#   if fol == '.DS_Store': continue
#   if 'split' not in fol: continue
#   if 'pipeline' in fol: continue
#   ds = fol.split('_', 2)[2].split('_')
#   ds = ds[1]+'_'+ds[0].replace('split', 'updated')
#   os.system('cp '+folder+fol+'/EC_metagenome_out/pred_metagenome_unstrat.tsv '+out_folder+ds+'_'+prefix+'_EC_metagenome.tsv')
#   os.system('cp '+folder+fol+'/EC_metagenome_out/pred_metagenome_unstrat.tsv '+out_folder+ds+'_'+prefix+'_EC_metagenome.tsv')
#   os.system('cp '+folder+fol+'/KO_metagenome_out/pred_metagenome_unstrat.tsv '+out_folder+ds+'_'+prefix+'_KO_metagenome.tsv')
#   os.system('cp '+folder+fol+'/KO_metagenome_out/weighted_nsti.tsv '+out_folder+ds+'_'+prefix+'_weighted_nsti.tsv')
#   if 'default' in fol:
#     os.system('cp '+folder+fol+'/marker_predicted_and_nsti.tsv.gz '+out_folder+ds+'_'+prefix+'_nsti.tsv.gz')
#   else:
#     os.system('cp '+folder+fol+'/combined_marker_predicted_and_nsti.tsv.gz '+out_folder+ds+'_'+prefix+'_nsti.tsv.gz')
  
# copy mock V4-V5 output
# folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/'
# prefix = 'V4V5'
# folders = os.listdir(folder)
# out_folder = '/Users/robynwright/Dropbox/Papers_writing/PICRUSt2_database_application_note/compile_output/mgs_picrust2/'

# for fol in folders:
#   if fol == '.DS_Store': continue
#   if 'split' not in fol: continue
#   if 'pipeline' in fol: continue
#   ds = fol.split('_', 2)[2].split('_')
#   ds = ds[1]+'_'+ds[0].replace('split', 'updated')
#   os.system('cp '+folder+fol+'/EC_metagenome_out/pred_metagenome_unstrat.tsv '+out_folder+ds+'_EC_metagenome.tsv')
#   os.system('cp '+folder+fol+'/EC_metagenome_out/pred_metagenome_unstrat.tsv '+out_folder+ds+'_EC_metagenome.tsv')
#   os.system('cp '+folder+fol+'/KO_metagenome_out/pred_metagenome_unstrat.tsv '+out_folder+ds+'_KO_metagenome.tsv')
#   os.system('cp '+folder+fol+'/KO_metagenome_out/weighted_nsti.tsv.gz '+out_folder+ds+'_weighted_nsti.tsv.gz')
#   if 'default' in fol:
#     os.system('cp '+folder+fol+'/marker_predicted_and_nsti.tsv.gz '+out_folder+ds+'_nsti.tsv.gz')
#   else:
#     os.system('cp '+folder+fol+'/combined_marker_predicted_and_nsti.tsv '+out_folder+ds+'_nsti.tsv')

# copy mock truth
# datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
# folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/mock_test/feature_tables/'
# out_folder = '/Users/robynwright/Dropbox/Papers_writing/PICRUSt2_database_application_note/compile_output/mock/'
# for ds in datasets:
#   os.system('cp '+folder+ds+'_16S.csv '+out_folder+ds+'_genome_abundance.csv')
#   os.system('cp '+folder+ds+'_mock_EC_metagenome.csv '+out_folder+ds+'_EC_metagenome_norm.csv')
#   os.system('cp '+folder+ds+'_mock_KO_metagenome.csv '+out_folder+ds+'_KO_metagenome_norm.csv')
# 
# os.system('cp /Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/genome_full_tax.csv '+out_folder)

# copy mock picrust2
datasets = ['blueberry', 'cameroon', 'hmp', 'indian', 'mammal', 'ocean', 'primate']
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/mock_test/feature_tables/'
out_folder = '/Users/robynwright/Dropbox/Papers_writing/PICRUSt2_database_application_note/compile_output/mock/'
for ds in datasets:
  os.system('cp '+folder+ds+'_mock_EC_metagenome_picrust2_genomes.csv '+out_folder+ds+'_EC_metagenome_picrust2_genomes_norm.csv')
  os.system('cp '+folder+ds+'_mock_KEGG_ko_metagenome_picrust2_genomes.csv '+out_folder+ds+'_KO_metagenome_picrust2_genomes_norm.csv')

# copy picrust2 dataset output
# folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/mgs_test/'
# folders = os.listdir(folder)
# out_folder = '/Users/robynwright/Dropbox/Papers_writing/PICRUSt2_database_application_note/compile_output/mgs_picrust2/'
# 
# for fol in folders:
#   if fol == '.DS_Store': continue
#   ds = fol.replace('picrust2_out_', '').split('_')
#   ds = ds[1]+'_'+ds[0].replace('split', 'updated')
#   os.system('cp '+folder+fol+'/EC_metagenome_out/pred_metagenome_unstrat.tsv '+out_folder+ds+'_EC_metagenome.tsv')
#   os.system('cp '+folder+fol+'/KO_metagenome_out/pred_metagenome_unstrat.tsv '+out_folder+ds+'_KO_metagenome.tsv')
#   os.system('cp '+folder+fol+'/KO_metagenome_out/weighted_nsti.tsv '+out_folder+ds+'_weighted_nsti.tsv')
#   if 'default' in ds:
#     os.system('cp '+folder+fol+'/marker_predicted_and_nsti.tsv '+out_folder+ds+'_nsti.tsv')
#   else:
#     os.system('cp '+folder+fol+'/combined_marker_predicted_and_nsti.tsv '+out_folder+ds+'_nsti.tsv')

# copy mgs "gold standard"
folder = '/Users/robynwright/Dropbox/Langille_Lab_postdoc/microbiome_helper/PICRUSt2/new_database/picrust_test_v2.6.0/mgs_validation/'
folders = os.listdir(folder)
out_folder = '/Users/robynwright/Dropbox/Papers_writing/PICRUSt2_database_application_note/compile_output/mgs/'

for fol in folders:
  if fol == '.DS_Store': continue
  if fol not in datasets: continue
  os.system('cp '+folder+fol+'/humann2_ec_unstrat.tsv '+out_folder+fol+'_EC_metagenome.tsv')
  os.system('cp '+folder+fol+'/humann2_ko_unstrat.tsv '+out_folder+fol+'_KO_metagenome.tsv')
  

```

## Figure 2 NSTI and weighted NSTI

```{python}
fig = plt.figure(figsize=(5,5))
ax1 = plt.subplot2grid((2,1),(0,0), colspan=4)
ax2 = plt.subplot2grid((2,1),(1,0), colspan=4)

plt.sca(ax1)
x = [0, 1]
xlocs = []
all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
  asvs = list(set(list(default.index.values)+list(updated.index.values)))
  for asv in asvs:
    if asv in default.index.values:
      def_nsti = default.loc[asv, 'metadata_NSTI']
      all_nsti_default.append(def_nsti)
    else: def_nsti = 'NA'
    if asv in updated.index.values:
      upd_nsti = updated.loc[asv, 'metadata_NSTI']
      all_nsti_updated.append(upd_nsti)
    else: upd_nsti = 'NA'
    if def_nsti == 'NA' or upd_nsti == 'NA':
      continue
    else:
      nsti_difference.append(upd_nsti-def_nsti)
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  sc = ax1.scatter(np.random.normal(x[0], scale=0.1, size=len(default)), default, color=colours[0], alpha=0.3, s=5)
  sc = ax1.scatter(np.random.normal(x[1], scale=0.1, size=len(updated)), updated, color=colours[1], alpha=0.3, s=5)
  mo = max([max(default), max(updated)])
  if mo > 2: mo = 2
  t, p = ttest_ind(default, updated)
  t = round(t, 2)
  if p < 0.001: p = '***'
  elif p < 0.01: p = '**'
  elif p <= 0.05: p = '*'
  else: p = ''
  lp = ax1.plot([x[0], x[0], x[1], x[1]], [mo*1.05, mo*1.07, mo*1.07, mo*1.05], 'k-')
  lp = ax1.text(np.mean(x), mo*1.09, p, ha='center', va='bottom', fontweight='bold')
  box = ax1.boxplot([default, updated], positions=x, widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] += 2.5
  x[1] += 2.5

print('Default database: mean =', round(np.mean(all_nsti_default), 3), ', median =', round(np.median(all_nsti_default), 3), 'range ', round(min(all_nsti_default), 3), '-', round(max(all_nsti_default), 3))
print('Updated database: mean =', round(np.mean(all_nsti_updated), 3), ', median =', round(np.median(all_nsti_updated), 3), 'range ', round(min(all_nsti_updated), 3), '-', round(max(all_nsti_updated), 3))
print('Difference: mean =', round(np.mean(nsti_difference), 3), ', median =', round(np.median(nsti_difference), 3), 'range ', round(min(nsti_difference), 3), '-', round(max(nsti_difference), 3))

yl = plt.ylim([-0.05, 2.5])
xt = plt.xticks(xlocs, [])
yl = plt.ylabel('NSTI (per sequence)', fontweight='bold')
databases = ['PICRUSt2-oldIMG', 'PICRUSt2-SC']
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=10) for s in range(len(databases))]
lg = ax1.legend(handles=handles, loc='upper left', bbox_to_anchor=(1.01, 1.01), fontsize=8)

plt.sca(ax2)
x = [0, 1]
xlocs = []
all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_weighted_nsti.tsv', index_col=0, header=0, sep='\t')#['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_weighted_nsti.tsv', index_col=0, header=0, sep='\t')#['weighted_NSTI'].values
  samples = list(set(list(default.index.values)+list(updated.index.values)))
  for sample in samples:
    if sample in default.index.values:
      def_nsti = default.loc[sample, 'weighted_NSTI']
      all_nsti_default.append(def_nsti)
    else: def_nsti = 'NA'
    if sample in updated.index.values:
      upd_nsti = updated.loc[sample, 'weighted_NSTI']
      all_nsti_updated.append(upd_nsti)
    else: upd_nsti = 'NA'
    if def_nsti == 'NA' or upd_nsti == 'NA':
      continue
    else:
      nsti_difference.append(upd_nsti-def_nsti)
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  sc = ax2.scatter(np.random.normal(x[0], scale=0.1, size=len(default)), default, color=colours[0], alpha=0.5, s=10)
  sc = ax2.scatter(np.random.normal(x[1], scale=0.1, size=len(updated)), updated, color=colours[1], alpha=0.5, s=10)
  mo = max([max(default), max(updated)])
  if mo > 2: mo = 2
  t, p = ttest_ind(default, updated)
  t = round(t, 2)
  if p < 0.001: p = '***'
  elif p < 0.01: p = '**'
  elif p <= 0.05: p = '*'
  else: p = 'NS'
  lp = ax2.plot([x[0], x[0], x[1], x[1]], [mo*1.05, mo*1.07, mo*1.07, mo*1.05], 'k-')
  lp = ax2.text(np.mean(x), mo*1.09, p, ha='center', va='bottom', fontweight='bold')
  print(ds, round(np.median(updated)-np.median(default), 2))
  box = ax2.boxplot([default, updated], positions=x, widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] += 2.5
  x[1] += 2.5

print('Default database: mean =', round(np.mean(all_nsti_default), 3), ', median =', round(np.median(all_nsti_default), 3), 'range ', round(min(all_nsti_default), 3), '-', round(max(all_nsti_default), 3))
print('Updated database: mean =', round(np.mean(all_nsti_updated), 3), ', median =', round(np.median(all_nsti_updated), 3), 'range ', round(min(all_nsti_updated), 3), '-', round(max(all_nsti_updated), 3))
print('Difference: mean =', round(np.mean(nsti_difference), 3), ', median =', round(np.median(nsti_difference), 3), 'range ', round(min(nsti_difference), 3), '-', round(max(nsti_difference), 3))

xt = plt.xticks(xlocs, dataset_names, rotation=90)
yl = plt.ylim([-0.05, 0.68])
yl = plt.ylabel('Weighted NSTI (per sample)', fontweight='bold')

ti = ax1.set_title('A', fontweight='bold', loc='left')
ti = ax2.set_title('B', fontweight='bold', loc='left')

plt.savefig(overall_folder+'figures/figS1_nsti_datasets.png', dpi=600, bbox_inches='tight')
```

## Figure 3 Correlations between default and updated PICRUSt2 databases

```{python}
plt.close()
fig = plt.figure(figsize=(10,5))
ax1 = plt.subplot(121)
ax2 = plt.subplot(122)
axes = {'KO':ax1, 'EC':ax2}
label = {'KO':'A', 'EC':'B'}

for trait in ['KO', 'EC']:
  all_correlations = []
  for ds in datasets:
    default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_'+trait+'_metagenome.tsv', sep='\t', index_col=0, header=0)
    updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_'+trait+'_metagenome.tsv', sep='\t', index_col=0, header=0)
    rename = {}
    for row in updated.index.values:
      if trait == 'EC': rename[row] = 'EC:'+row
      else: rename[row] = row.split(':')[1]
    updated = updated.rename(index=rename)
    intersection = [trait for trait in default.index.values if trait in updated.index.values]
    updated = updated.loc[intersection, :]
    default = default.loc[intersection, :]
    correlations = []
    for sample in updated.columns:
      new_vals = updated.loc[:, sample].values
      old_vals = default.loc[:, sample].values
      stat, p = spearmanr(new_vals, old_vals)
      correlations.append(stat)
      if p >= 0.001: print('This wasnt significant!!: '+ds+', '+sample)
    #print(ds, trait, np.median(correlations))
    all_correlations.append(correlations)
  for d in range(len(datasets)):
    #if d > 4: continue
    sc = axes[trait].scatter(np.random.normal(d, scale=0.1, size=len(all_correlations[d])), all_correlations[d], color=colours[2], alpha=0.5)
  box = axes[trait].boxplot(all_correlations, positions=[d for d in range(len(datasets))], widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  plt.sca(axes[trait])
  xt = plt.xticks([d for d in range(len(datasets))], dataset_names, rotation=90)
  if trait == 'KO': yl = plt.ylabel('Spearman correlation coefficient', fontweight='bold')
  ti = plt.title(trait.replace('path', 'Pathways').replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  ti = plt.title(label[trait], fontweight='bold', loc='left')
  yl = plt.ylim([0.745, 0.955])
  if trait != 'KO': yt = axes[trait].set_yticklabels([])
  print('\n\n')

plt.savefig(overall_folder+'figures/figS2_correlation_default_updated.png', dpi=600, bbox_inches='tight')
```

## Figure 4 Correlation and distance between MGS and default/updated databases

```{python}
plt.close()
fig = plt.figure(figsize=(10,8))

traits = ['KO', 'EC']
labels = [['A', 'C'], ['B', 'D']]
for t in range(len(traits)):
  ax1 = plt.subplot2grid((2,2),(0,t))
  ax2 = plt.subplot2grid((2,2),(1,t))
  x = [0, 1]
  xlocs = []
  all_correlations, all_distances = [], []
  for ds in datasets:
    mock = pd.read_csv(overall_folder+'mgs/'+ds+'_'+traits[t]+'_metagenome.tsv', index_col=0, header=0, sep='\t')
    default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_'+traits[t]+'_metagenome.tsv', index_col=0, header=0, sep='\t')
    updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_'+traits[t]+'_metagenome.tsv', index_col=0, header=0, sep='\t')
    rename = {}
    for row in updated.index.values:
      if traits[t] == 'KO': rename[row] = row.split(':')[1]
      else: rename[row] = 'EC:'+row
    updated = updated.rename(index=rename)
    samples = [s for s in mock.columns if s in default.columns and s in updated.columns]
    comparisons = [default, updated]
    names = ['Default', 'Updated']
    ds_correlations, ds_distances = [], []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      mock_comp = mock.copy(deep=True)
      comp_relabun = comp.copy(deep=True)
      comp_relabun = comp_relabun.divide(comp_relabun.sum(axis=0), axis=1).multiply(100)
      mock_comp_relabun = mock_comp.copy(deep=True)
      mock_comp_relabun = mock_comp_relabun.divide(mock_comp_relabun.sum(axis=0), axis=1).multiply(100)
      intersection = [trait for trait in comp.index.values if trait in mock_comp.index.values]
      comp = comp.loc[intersection, :]
      mock_comp = mock_comp.loc[intersection, :]
      correlations, distances = [], []
      for sample in samples:
        combined = pd.concat([comp_relabun.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp_relabun.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mock_vals)
        ap = distances.append(bc_dist), correlations.append(stat)
      ap = ds_correlations.append(correlations), ds_distances.append(distances)
    ap = all_correlations.append(ds_correlations), all_distances.append(ds_distances)
  for d in range(len(datasets)):
    sc = ax1.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5, s=10)
    sc = ax1.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5, s=10)
    mo = max([max(all_correlations[d][0]), max(all_correlations[d][1])])
    if mo > 2: mo = 2
    tstat, p = ttest_ind(all_correlations[d][0], all_correlations[d][1])
    tstat = round(tstat, 2)
    if p < 0.001: p = '***'
    elif p < 0.01: p = '**'
    elif p <= 0.05: p = '*'
    else: p = 'NS'
    lp = ax1.plot([x[0], x[0], x[1], x[1]], [mo*1.05, mo*1.06, mo*1.06, mo*1.05], 'k-')
    lp = ax1.text(np.mean(x), mo*1.07, p, ha='center', va='bottom', fontweight='bold')
    box = ax1.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    sc = ax2.scatter(np.random.normal(x[0], scale=0.1, size=len(all_distances[d][0])), all_distances[d][0], color=colours[0], alpha=0.5, s=10)
    sc = ax2.scatter(np.random.normal(x[1], scale=0.1, size=len(all_distances[d][1])), all_distances[d][1], color=colours[1], alpha=0.5, s=10)
    mo = max([max(all_distances[d][0]), max(all_distances[d][1])])
    if mo > 2: mo = 2
    tstat, p = ttest_ind(all_distances[d][0], all_distances[d][1])
    tstat = round(tstat, 2)
    if p < 0.001: p = '***'
    elif p < 0.01: p = '**'
    elif p <= 0.05: p = '*'
    else: p = 'NS'
    lp = ax2.plot([x[0], x[0], x[1], x[1]], [mo*1.05, mo*1.06, mo*1.06, mo*1.05], 'k-')
    lp = ax2.text(np.mean(x), mo*1.07, p, ha='center', va='bottom', fontweight='bold')
    box = ax2.boxplot(all_distances[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xl = xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  plt.sca(ax1)
  if traits[t] == 'KO':
    yl = plt.ylabel('Spearman correlation coefficient', fontweight='bold')
  if t == 0:
    yl = plt.ylim([0.65, 1.0])
  else:
    yl = plt.ylim([0.65, 0.95])
  xt = plt.xticks(xlocs, [])
  ti = plt.title(traits[t].replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  ti = plt.title(labels[t][0], loc='left', fontweight='bold')
  #yl = plt.ylim([0.32, 0.84])
  plt.sca(ax2)
  xt = plt.xticks(xlocs, dataset_names, rotation=90)
  if traits[t] == 'KO': yl = plt.ylabel('Bray-Curtis dissimilarity', fontweight='bold')
  ti = plt.title(labels[t][1], loc='left', fontweight='bold')
  databases = ['PICRUSt2-oldIMG', 'PICRUSt2-SC']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  if t == 1: 
    legend = ax2.legend(handles=handles, loc='lower right')
    yl = plt.ylim([0.85, 1.07])
  else:
    yl = plt.ylim([0.65, 1.1])

plt.savefig(overall_folder+'figures/figS3_mgs_correlation_distance.png', dpi=600, bbox_inches='tight')
```

## Figure 5 Taxa in the mock samples and NSTI/weighted NSTI

```{python}
plt.close()
fig = plt.figure(figsize=(10,5))
ax1 = plt.subplot2grid((1,9),(0,0), colspan=2)
ax2 = plt.subplot2grid((2,9),(0,5), colspan=4)
ax3 = plt.subplot2grid((2,9),(1,5), colspan=4)

tax = pd.read_csv(overall_folder+'mock/genome_full_tax.csv', index_col=0, header=0)

tax_genus, tax_class, tax_family = {}, {}, {}
for row in tax.index.values:
  this_tax = tax.loc[row, 'Taxonomy'].split(';')
  tax_genus[row] = this_tax[-2]
  tax_family[row] = this_tax[-3]
  tax_class[row] = this_tax[2].replace('c__', '')
  
combined = []
for ds in datasets:
  ft = pd.read_csv(overall_folder+'mock/'+ds+'_genome_abundance.csv', index_col=0, header=0).drop('Tax', axis=1)
  ft = ft.rename(index=tax_class)
  ft = ft.groupby(by=ft.index).sum()
  if 'Unassigned' in ft.index.values: ft = ft.drop('Unassigned')
  ft = ft.divide(ft.sum(axis=0), axis=1).multiply(100)
  ft[ds] = ft.mean(axis=1)
  combined.append(ft.loc[:, [ds]])

combined_df = pd.concat(combined).fillna(value=0)
combined_df = combined_df.groupby(by=combined_df.index).sum()
combined_df['Mean'] = combined_df.mean(axis=1)
combined_df = combined_df.sort_values(by=['Mean'], ascending=False)
combined_df = combined_df.iloc[:19, :]
combined_df = combined_df.drop('Mean', axis=1)
combined_df = combined_df.loc[sorted(combined_df.index.values), :]
combined_df.loc['Other', :] = 100-combined_df.sum(axis=0).values
plt.sca(ax1)
a1 = combined_df.transpose().plot(ax=ax1, kind='bar', stacked=True, width=0.9, cmap='tab20', edgecolor='k')
lg = plt.legend(loc='upper left', bbox_to_anchor=(1.01, 1.02), fontsize=8)
yl = plt.ylabel('Relative abundance (%)')
yl = plt.ylim([0, 100])
xt = plt.xticks([d for d in range(len(dataset_names))], dataset_names, rotation=90)

plt.sca(ax2)
x = [0, 1]
xlocs = []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  sc = ax2.scatter(np.random.normal(x[0], scale=0.1, size=len(default)), default, color=colours[0], alpha=0.5, s=10)
  sc = ax2.scatter(np.random.normal(x[1], scale=0.1, size=len(updated)), updated, color=colours[1], alpha=0.5, s=10)
  print('Per sequence', ds, np.median(default), np.median(updated))
  box = ax2.boxplot([default, updated], positions=x, widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] += 2.5
  x[1] += 2.5

yl = plt.ylim([-0.05, 1])
xt = plt.xticks(xlocs, [])
yl = plt.ylabel('NSTI (per sequence)', fontweight='bold')

print('\n\n')

plt.sca(ax3)
x = [0, 1]
xlocs = []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  sc = ax3.scatter(np.random.normal(x[0], scale=0.1, size=len(default)), default, color=colours[0], alpha=0.5, s=10)
  sc = ax3.scatter(np.random.normal(x[1], scale=0.1, size=len(updated)), updated, color=colours[1], alpha=0.5, s=10)
  box = ax3.boxplot([default, updated], positions=x, widths=0.8, showfliers=False)
  print('Per sample', ds, np.median(default), np.median(updated))
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] += 2.5
  x[1] += 2.5
  
#yl = plt.ylim([-0.05, 1])
xt = plt.xticks(xlocs, dataset_names, rotation=90)
yl = plt.ylabel('Weighted NSTI (per sample)', fontweight='bold')
databases = ['Default PICRUSt2', 'Updated PICRUSt2']
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=10) for s in range(len(databases))]
lg = ax3.legend(handles=handles, loc='lower left', bbox_to_anchor=(-0.82, 0), fontsize=8)

ti = ax1.set_title('A', fontweight='bold', loc='left')
ti = ax2.set_title('B', fontweight='bold', loc='left')
ti = ax3.set_title('C', fontweight='bold', loc='left')

plt.savefig(overall_folder+'figures/fig5_taxa_in_mock.png', dpi=600, bbox_inches='tight')
```

```{python}
plt.close()
fig = plt.figure(figsize=(15,20))
ax9 = plt.subplot2grid((60,40),(20,17), rowspan=14, colspan=11)
ax10 = plt.subplot2grid((60,40),(34,17), rowspan=2, colspan=11)
ax11 = plt.subplot2grid((60,40),(20,29), rowspan=14, colspan=11)
ax12 = plt.subplot2grid((60,40),(34,29), rowspan=2, colspan=11)


plt.sca(ax9)
x = [18.5, 17.5]
xlocs = []
all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
  asvs = list(set(list(default.index.values)+list(updated.index.values)))
  for asv in asvs:
    if asv in default.index.values:
      def_nsti = default.loc[asv, 'metadata_NSTI']
      all_nsti_default.append(def_nsti)
    else: def_nsti = 'NA'
    if asv in updated.index.values:
      upd_nsti = updated.loc[asv, 'metadata_NSTI']
      all_nsti_updated.append(upd_nsti)
    else: upd_nsti = 'NA'
    if def_nsti == 'NA' or upd_nsti == 'NA':
      continue
    else:
      nsti_difference.append(upd_nsti-def_nsti)
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  sc = ax9.scatter(default, np.random.normal(x[0], scale=0.1, size=len(default)), color=colours[0], alpha=0.5, s=10)
  sc = ax9.scatter(updated, np.random.normal(x[1], scale=0.1, size=len(updated)), color=colours[1], alpha=0.5, s=10)
  tstat, p = ttest_ind(default, updated)
  fc, alp = 'w', 0.7
  if p <= 0.05: fc, alp ='gray', 0.6
  fontcol = 'k'
  if np.median(default) < np.median(updated): fontcol = '#a93226'
  tx = ax9.text(0.97, x[0], round(np.median(default), 3), ha='right', va='center', color=fontcol, bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax9.text(0.97, x[1], round(np.median(updated), 3), ha='right', va='center', color=fontcol, bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  box = ax9.boxplot([default, updated], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] -= 2.5
  x[1] -= 2.5

yl = plt.xlim([-0.05, 1])
xt = plt.yticks(xlocs, dataset_names), plt.xticks([])

# 
plt.sca(ax10)
x = [1, 0]
sc = ax10.scatter(all_nsti_default, np.random.normal(x[0], scale=0.1, size=len(all_nsti_default)), color=colours[0], alpha=0.5, s=10)
sc = ax10.scatter(all_nsti_updated, np.random.normal(x[1], scale=0.1, size=len(all_nsti_updated)), color=colours[1], alpha=0.5, s=10)
yl = plt.xlim([-0.05, 1])
box = ax10.boxplot([all_nsti_default, all_nsti_updated], positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([0.5], ['Overall'])
xl = plt.xlabel('NSTI per sequence', fontweight='bold')
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tstat, p = ttest_ind(all_nsti_default, all_nsti_updated)
fc, alp = 'w', 0.7
if p <= 0.05: fc, alp ='gray', 0.6
tx = ax10.text(0.97, x[0], round(np.median(all_nsti_default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
tx = ax10.text(0.97, x[1], round(np.median(all_nsti_updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))

plt.sca(ax11)
x = [18.5, 17.5]
xlocs = []
all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_weighted_nsti.tsv', index_col=0, header=0, sep='\t')#['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_weighted_nsti.tsv', index_col=0, header=0, sep='\t')#['weighted_NSTI'].values
  asvs = list(set(list(default.index.values)+list(updated.index.values)))
  for asv in asvs:
    if asv in default.index.values:
      def_nsti = default.loc[asv, 'weighted_NSTI']
      all_nsti_default.append(def_nsti)
    else: def_nsti = 'NA'
    if asv in updated.index.values:
      upd_nsti = updated.loc[asv, 'weighted_NSTI']
      all_nsti_updated.append(upd_nsti)
    else: upd_nsti = 'NA'
    if def_nsti == 'NA' or upd_nsti == 'NA':
      continue
    else:
      nsti_difference.append(upd_nsti-def_nsti)
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  sc = ax11.scatter(default, np.random.normal(x[0], scale=0.1, size=len(default)), color=colours[0], alpha=0.5, s=10)
  sc = ax11.scatter(updated, np.random.normal(x[1], scale=0.1, size=len(updated)), color=colours[1], alpha=0.5, s=10)
  tstat, p = ttest_ind(default, updated)
  fc, alp = 'w', 0.7
  if p <= 0.05: fc, alp ='gray', 0.6
  fontcol = 'k'
  if np.median(default) < np.median(updated): fontcol = '#a93226'
  tx = ax11.text(0.68, x[0], round(np.median(default), 3), ha='right', va='center', color=fontcol, bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax11.text(0.68, x[1], round(np.median(updated), 3), ha='right', va='center', color=fontcol, bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  box = ax11.boxplot([default, updated], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] -= 2.5
  x[1] -= 2.5

yl = plt.xlim([-0.02, 0.7])
xt = plt.yticks([]), plt.xticks([])

# 
plt.sca(ax12)
x = [1, 0]
sc = ax12.scatter(all_nsti_default, np.random.normal(x[0], scale=0.1, size=len(all_nsti_default)), color=colours[0], alpha=0.5, s=10)
sc = ax12.scatter(all_nsti_updated, np.random.normal(x[1], scale=0.1, size=len(all_nsti_updated)), color=colours[1], alpha=0.5, s=10)
yl = plt.xlim([-0.02, 0.7])
box = ax12.boxplot([all_nsti_default, all_nsti_updated], positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([])
xl = plt.xlabel('NSTI per sample', fontweight='bold')
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tstat, p = ttest_ind(all_nsti_default, all_nsti_updated)
fc, alp = 'w', 0.7
if p <= 0.05: fc, alp ='gray', 0.6
tx = ax12.text(0.68, x[0], round(np.median(all_nsti_default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
tx = ax12.text(0.68, x[1], round(np.median(all_nsti_updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))

ti = ax9.set_title('A', fontweight='bold', loc='left')
ti = ax9.set_title('NSTI per sequence', fontweight='bold')
ti = ax11.set_title('B', fontweight='bold', loc='left')
ti = ax11.set_title('NSTI per sample', fontweight='bold')

databases = ['PICRUSt2-oldIMG', 'PICRUSt2-SC']
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
legend = ax11.legend(handles=handles, loc='upper left', bbox_to_anchor=(1.01, 1.01))


plt.savefig(overall_folder+'figures/figS4_weighted_nsti_and_nsti.png', dpi=600, bbox_inches='tight')
```

## Figure 6 Correlation and distance between mock and default/updated databases for full-length 16S

```{python}
plt.close()
fig = plt.figure(figsize=(15,20))
ax9 = plt.subplot2grid((60,40),(20,17), rowspan=14, colspan=11)
ax10 = plt.subplot2grid((60,40),(34,17), rowspan=2, colspan=11)
ax11 = plt.subplot2grid((60,40),(20,29), rowspan=14, colspan=11)
ax12 = plt.subplot2grid((60,40),(34,29), rowspan=2, colspan=11)

x = [18.5, 17.5]
xlocs = []
all_correlations, all_distances = [], []
all_correlations_default, all_correlations_updated = [], []
all_distances_default, all_distances_updated = [], []
for ds in datasets:
  mock = pd.read_csv(overall_folder+'mock/'+ds+'_EC_metagenome_norm.csv', index_col=0, header=0)
  mock = mock.drop('-', axis=0)
  mock = mock[mock.max(axis=1) > 0]
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_EC_metagenome.tsv', index_col=0, header=0, sep='\t')
  rename = {}
  for row in default.index.values:
    if traits[t] == 'KO': rename[row] = 'ko:'+row
    else: rename[row] = row.split(':')[1]
  default = default.rename(index=rename)
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_EC_metagenome.tsv', index_col=0, header=0, sep='\t')
  samples = [s for s in mock.columns if s in default.columns and s in updated.columns]
  comparisons = [default, updated]
  names = ['Default', 'Updated']
  ds_correlations, ds_distances = [], []
  for c in range(len(comparisons)):
    comp = comparisons[c].copy(deep=True)
    mock_comp = mock.copy(deep=True)
    comp_relabun = comp.copy(deep=True)
    comp_relabun = comp_relabun.divide(comp_relabun.sum(axis=0), axis=1).multiply(100)
    mock_comp_relabun = mock_comp.copy(deep=True)
    mock_comp_relabun = mock_comp_relabun.divide(mock_comp_relabun.sum(axis=0), axis=1).multiply(100)
    intersection = [trait for trait in comp.index.values if trait in mock_comp.index.values]
    comp = comp.loc[intersection, :]
    mock_comp = mock_comp.loc[intersection, :]
    correlations, distances = [], []
    for sample in samples:
        combined = pd.concat([comp_relabun.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp_relabun.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mock_vals)
        ap = distances.append(bc_dist), correlations.append(stat)
        if c == 0:
          ap = all_distances_default.append(bc_dist), all_correlations_default.append(stat)
        else:
          ap = all_distances_updated.append(bc_dist), all_correlations_updated.append(stat)
    ap = ds_correlations.append(correlations), ds_distances.append(distances)
  ap = all_correlations.append(ds_correlations), all_distances.append(ds_distances)
for d in range(len(datasets)):
  sc = ax9.scatter(all_correlations[d][0], np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), color=colours[0], alpha=0.5, s=10)
  sc = ax9.scatter(all_correlations[d][1], np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), color=colours[1], alpha=0.5, s=10)
  tstat, p = ttest_ind(all_correlations[d][0], all_correlations[d][1])
  fc, alp = 'w', 0.7
  if p <= 0.05: fc, alp ='gray', 0.6
  tx = ax9.text(0.57, x[0], round(np.median(all_correlations[d][0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax9.text(0.57, x[1], round(np.median(all_correlations[d][1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  box = ax9.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  sc = ax11.scatter(all_distances[d][0], np.random.normal(x[0], scale=0.1, size=len(all_distances[d][0])), color=colours[0], alpha=0.5, s=10)
  sc = ax11.scatter(all_distances[d][1], np.random.normal(x[1], scale=0.1, size=len(all_distances[d][1])), color=colours[1], alpha=0.5, s=10)
  tstat, p = ttest_ind(all_distances[d][0], all_distances[d][1])
  fc, alp = 'w', 0.7
  if p <= 0.05: fc, alp ='gray', 0.6
  tx = ax11.text(0.74, x[0], round(np.median(all_distances[d][0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax11.text(0.74, x[1], round(np.median(all_distances[d][1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  box = ax11.boxplot(all_distances[d], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xl = xlocs.append(np.mean(x))
  x[0] += 2.5
  x[1] += 2.5
  
plt.sca(ax10)
x = [1, 0]
sc = ax10.scatter(all_correlations_default, np.random.normal(x[0], scale=0.1, size=len(all_correlations_default)), color=colours[0], alpha=0.5, s=10)
sc = ax10.scatter(all_correlations_updated, np.random.normal(x[1], scale=0.1, size=len(all_correlations_updated)), color=colours[1], alpha=0.5, s=10)
yl = plt.xlim([-0.05, 1])
box = ax10.boxplot([all_correlations_default, all_correlations_updated], positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([0.5], ['Overall'])
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tstat, p = ttest_ind(all_correlations_default, all_correlations_updated)
fc, alp = 'w', 0.7
if p <= 0.05: fc, alp ='gray', 0.6
tx = ax10.text(0.57, x[0], round(np.median(all_correlations_default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
tx = ax10.text(0.57, x[1], round(np.median(all_correlations_updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))

plt.sca(ax12)
x = [1, 0]
sc = ax12.scatter(all_distances_default, np.random.normal(x[0], scale=0.1, size=len(all_distances_default)), color=colours[0], alpha=0.5, s=10)
sc = ax12.scatter(all_distances_updated, np.random.normal(x[1], scale=0.1, size=len(all_distances_updated)), color=colours[1], alpha=0.5, s=10)
yl = plt.xlim([-0.05, 1])
box = ax12.boxplot([all_distances_default, all_distances_updated], positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([0.5], ['Overall'])
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tstat, p = ttest_ind(all_distances_default, all_distances_updated)
fc, alp = 'w', 0.7
if p <= 0.05: fc, alp ='gray', 0.6
tx = ax12.text(0.74, x[0], round(np.median(all_distances_default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
tx = ax12.text(0.74, x[1], round(np.median(all_distances_updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))

ti = ax9.set_title('A', fontweight='bold', loc='left')
ti = ax9.set_title("Spearman's correlation", fontweight='bold')
ti = ax11.set_title('B', fontweight='bold', loc='left')
ti = ax11.set_title('Bray-Curtis dissimilarity', fontweight='bold')

databases = ['PICRUSt2-oldIMG', 'PICRUSt2-SC']
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
legend = ax11.legend(handles=handles, loc='upper left', bbox_to_anchor=(1.01, 1.01))

plt.sca(ax9)
yt = plt.yticks(xlocs, dataset_names)
xl = plt.xlim([0.36, 0.58])
xt = plt.xticks([])

plt.sca(ax10)
xl = plt.xlabel('Spearman correlation coefficient', fontweight='bold')
xl = plt.xlim([0.36, 0.58])

plt.sca(ax11)
yt = plt.yticks([])
xl = plt.xlim([0.4, 0.75])
xt = plt.xticks([])

plt.sca(ax12)
xl = plt.xlabel('Bray-Curtis dissimilarity index', fontweight='bold')
xl = plt.xlim([0.4, 0.75])
yt = plt.yticks([])


plt.savefig(overall_folder+'figures/figS5_spearman_and_distance_ec_numbers.png', dpi=600, bbox_inches='tight')
```

```{python}
plt.close()
fig = plt.figure(figsize=(10,8))

traits = ['KO', 'EC']
labels = [['A', 'C'], ['B', 'D']]
for t in range(len(traits)):
  ax1 = plt.subplot2grid((2,2),(0,t))
  ax2 = plt.subplot2grid((2,2),(1,t))
  x = [0, 1]
  xlocs = []
  all_correlations, all_distances = [], []
  for ds in datasets:
    mock = pd.read_csv(overall_folder+'mock/'+ds+'_'+traits[t]+'_metagenome_norm.csv', index_col=0, header=0)
    mock = mock.drop('-', axis=0)
    mock = mock[mock.max(axis=1) > 0]
    default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_'+traits[t]+'_metagenome.tsv', index_col=0, header=0, sep='\t')
    rename = {}
    for row in default.index.values:
      if traits[t] == 'KO': rename[row] = 'ko:'+row
      else: rename[row] = row.split(':')[1]
    default = default.rename(index=rename)
    updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_'+traits[t]+'_metagenome.tsv', index_col=0, header=0, sep='\t')
    samples = [s for s in mock.columns if s in default.columns and s in updated.columns]
    comparisons = [default, updated]
    names = ['Default', 'Updated']
    ds_correlations, ds_distances = [], []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      mock_comp = mock.copy(deep=True)
      comp_relabun = comp.copy(deep=True)
      comp_relabun = comp_relabun.divide(comp_relabun.sum(axis=0), axis=1).multiply(100)
      mock_comp_relabun = mock_comp.copy(deep=True)
      mock_comp_relabun = mock_comp_relabun.divide(mock_comp_relabun.sum(axis=0), axis=1).multiply(100)
      intersection = [trait for trait in comp.index.values if trait in mock_comp.index.values]
      #if c == 1: print(len(comp.index.values), len(mock_comp.index.values), len(intersection))
      comp = comp.loc[intersection, :]
      mock_comp = mock_comp.loc[intersection, :]
      # comp_relabun_int = comp_relabun.loc[intersection, :]
      # mock_comp_relabun_int = mock_comp_relabun.loc[intersection, :]
      correlations, distances = [], []
      for sample in samples:
        combined = pd.concat([comp_relabun.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp_relabun.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        #combined = combined[combined.max(axis=1) > 0]
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        # print(combined.shape[0])
        # combined = combined[combined.min(axis=1) > 0]
        # print(combined.shape[0])
        # comp_vals = comp.loc[combined.index.values, sample].values
        # mock_vals = mock_comp.loc[combined.index.values, sample].values
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        # comp_vals = combined.loc[:, 'Comp'].values
        # mock_vals = combined.loc[:, 'Mock'].values
        stat, p = spearmanr(comp_vals, mock_vals)
        ap = distances.append(bc_dist), correlations.append(stat)
      ap = ds_correlations.append(correlations), ds_distances.append(distances)
    ap = all_correlations.append(ds_correlations), all_distances.append(ds_distances)
  for d in range(len(datasets)):
    sc = ax1.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5, s=10)
    sc = ax1.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5, s=10)
    print(traits[t], datasets[d], np.median(all_correlations[d][0]), np.median(all_correlations[d][1]))
    box = ax1.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    sc = ax2.scatter(np.random.normal(x[0], scale=0.1, size=len(all_distances[d][0])), all_distances[d][0], color=colours[0], alpha=0.5, s=10)
    sc = ax2.scatter(np.random.normal(x[1], scale=0.1, size=len(all_distances[d][1])), all_distances[d][1], color=colours[1], alpha=0.5, s=10)
    box = ax2.boxplot(all_distances[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xl = xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  plt.sca(ax1)
  if traits[t] == 'KO':
    yl = plt.ylabel('Spearman correlation coefficient', fontweight='bold')
  xt = plt.xticks(xlocs, [])
  ti = plt.title(traits[t].replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  ti = plt.title(labels[t][0], loc='left', fontweight='bold')
  #yl = plt.ylim([0.32, 0.84])
  plt.sca(ax2)
  xt = plt.xticks(xlocs, dataset_names, rotation=90)
  if traits[t] == 'KO': yl = plt.ylabel('Bray-Curtis dissimilarity', fontweight='bold')
  ti = plt.title(labels[t][1], loc='left', fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  if t == 1: legend = ax2.legend(handles=handles, loc='upper right')
  #yl = plt.ylim([0.17, 0.73])

plt.savefig(overall_folder+'figures/fig6_mock_correlation_distance_norm.png', dpi=600, bbox_inches='tight')
#plt.savefig(overall_folder+'figures/mock_correlation_distance_norm.png', dpi=600, bbox_inches='tight')
```

```{python}
plt.close()
fig = plt.figure(figsize=(10,8))

traits = ['KO', 'EC']
labels = [['A', 'C'], ['B', 'D']]
for t in range(len(traits)):
  ax1 = plt.subplot2grid((2,2),(0,t))
  ax2 = plt.subplot2grid((2,2),(1,t))
  x = [0, 1, 2]
  xlocs = []
  all_correlations, all_distances = [], []
  for ds in datasets:
    mock = pd.read_csv(overall_folder+'mock/'+ds+'_'+traits[t]+'_metagenome_norm.csv', index_col=0, header=0)
    mock = mock.drop('-', axis=0)
    mock = mock[mock.max(axis=1) > 0]
    mock_pic2 = pd.read_csv(overall_folder+'mock/'+ds+'_'+traits[t]+'_metagenome_picrust2_genomes_norm.csv', index_col=0, header=0)
    #mock_pic2 = mock_pic2.drop('-', axis=0)
    mock_pic2 = mock_pic2[mock_pic2.max(axis=1) > 0]
    default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_'+traits[t]+'_metagenome.tsv', index_col=0, header=0, sep='\t')
    rename = {}
    for row in default.index.values:
      if traits[t] == 'KO': rename[row] = 'ko:'+row
      else: rename[row] = row.split(':')[1]
    default = default.rename(index=rename)
    updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_'+traits[t]+'_metagenome.tsv', index_col=0, header=0, sep='\t')
    samples = [s for s in mock.columns if s in default.columns and s in updated.columns and s in mock_pic2.columns]
    comparisons = [default, updated, mock_pic2]
    names = ['Default', 'Updated', 'PICRUSt2 genomes']
    ds_correlations, ds_distances = [], []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      mock_comp = mock.copy(deep=True)
      comp_relabun = comp.copy(deep=True)
      comp_relabun = comp_relabun.divide(comp_relabun.sum(axis=0), axis=1).multiply(100)
      mock_comp_relabun = mock_comp.copy(deep=True)
      mock_comp_relabun = mock_comp_relabun.divide(mock_comp_relabun.sum(axis=0), axis=1).multiply(100)
      intersection = [trait for trait in comp.index.values if trait in mock_comp.index.values]
      #if c == 1: print(len(comp.index.values), len(mock_comp.index.values), len(intersection))
      comp = comp.loc[intersection, :]
      mock_comp = mock_comp.loc[intersection, :]
      # comp_relabun_int = comp_relabun.loc[intersection, :]
      # mock_comp_relabun_int = mock_comp_relabun.loc[intersection, :]
      correlations, distances = [], []
      for sample in samples:
        combined = pd.concat([comp_relabun.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp_relabun.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        #combined = combined[combined.max(axis=1) > 0]
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        # print(combined.shape[0])
        # combined = combined[combined.min(axis=1) > 0]
        # print(combined.shape[0])
        # comp_vals = comp.loc[combined.index.values, sample].values
        # mock_vals = mock_comp.loc[combined.index.values, sample].values
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        # comp_vals = combined.loc[:, 'Comp'].values
        # mock_vals = combined.loc[:, 'Mock'].values
        stat, p = spearmanr(comp_vals, mock_vals)
        ap = distances.append(bc_dist), correlations.append(stat)
      ap = ds_correlations.append(correlations), ds_distances.append(distances)
    ap = all_correlations.append(ds_correlations), all_distances.append(ds_distances)
  for d in range(len(datasets)):
    sc = ax1.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[0], alpha=0.5, s=10)
    sc = ax1.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[1], alpha=0.5, s=10)
    sc = ax1.scatter(np.random.normal(x[2], scale=0.1, size=len(all_correlations[d][2])), all_correlations[d][2], color=colours[2], alpha=0.5, s=10)
    box = ax1.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    sc = ax2.scatter(np.random.normal(x[0], scale=0.1, size=len(all_distances[d][0])), all_distances[d][0], color=colours[0], alpha=0.5, s=10)
    sc = ax2.scatter(np.random.normal(x[1], scale=0.1, size=len(all_distances[d][1])), all_distances[d][1], color=colours[1], alpha=0.5, s=10)
    sc = ax2.scatter(np.random.normal(x[2], scale=0.1, size=len(all_distances[d][2])), all_distances[d][2], color=colours[2], alpha=0.5, s=10)
    box = ax2.boxplot(all_distances[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xl = xlocs.append(np.mean(x))
    x[0] += 3.5
    x[1] += 3.5
    x[2] += 3.5
  plt.sca(ax1)
  if traits[t] == 'KO':
    yl = plt.ylabel('Spearman correlation coefficient', fontweight='bold')
  xt = plt.xticks(xlocs, [])
  ti = plt.title(traits[t].replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  ti = plt.title(labels[t][0], loc='left', fontweight='bold')
  #yl = plt.ylim([0.32, 0.84])
  plt.sca(ax2)
  xt = plt.xticks(xlocs, dataset_names, rotation=90)
  if traits[t] == 'KO': yl = plt.ylabel('Bray-Curtis dissimilarity', fontweight='bold')
  ti = plt.title(labels[t][1], loc='left', fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2', 'PICRUSt2 genomes']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  if t == 1: legend = ax2.legend(handles=handles, loc='upper right')
  #yl = plt.ylim([0.17, 0.73])

plt.savefig(overall_folder+'figures/mock_correlation_distance_norm_with_best.png', dpi=600, bbox_inches='tight')
#plt.savefig(overall_folder+'figures/mock_correlation_distance_norm.png', dpi=600, bbox_inches='tight')
```

## Figure 7 Correlation between Spearman/Bray-Curtis and NSTI

```{python}
plt.close()
fig = plt.figure(figsize=(10,8))

traits = ['KO', 'EC']
labels = [['A', 'C'], ['B', 'D']]
for t in range(len(traits)):
  ax1 = plt.subplot2grid((2,2),(0,t))
  ax2 = plt.subplot2grid((2,2),(1,t))
  # x = [0, 1]
  # xlocs = []
  # all_correlations, all_distances = [], []
  correlations, distances, nsti_ind = [[], []], [[], []], [[], []]
  for ds in datasets:
    mock = pd.read_csv(overall_folder+'mock/'+ds+'_'+traits[t]+'_metagenome.csv', index_col=0, header=0)
    default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_'+traits[t]+'_metagenome.tsv', index_col=0, header=0, sep='\t')
    rename = {}
    for row in default.index.values:
      if traits[t] == 'KO': rename[row] = 'ko:'+row
      else: rename[row] = row.split(':')[1]
    default = default.rename(index=rename)
    updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_'+traits[t]+'_metagenome.tsv', index_col=0, header=0, sep='\t')
    nsti_default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_weighted_nsti.tsv', index_col=0, header=0, sep='\t')
    nsti_updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_weighted_nsti.tsv', index_col=0, header=0, sep='\t')
    samples = [s for s in mock.columns if s in default.columns and s in updated.columns]
    comparisons = [default, updated]
    nsti_vals = [nsti_default, nsti_updated]
    names = ['Default', 'Updated']
    #ds_correlations, ds_distances = [], []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      mock_comp = mock.copy(deep=True)
      comp_relabun = comp.copy(deep=True)
      comp_relabun = comp_relabun.divide(comp_relabun.sum(axis=0), axis=1).multiply(100)
      mock_comp_relabun = mock_comp.copy(deep=True)
      mock_comp_relabun = mock_comp_relabun.divide(mock_comp_relabun.sum(axis=0), axis=1).multiply(100)
      intersection = [trait for trait in comp.index.values if trait in mock_comp.index.values]
      comp = comp.loc[intersection, :]
      mock_comp = mock_comp.loc[intersection, :]
      for sample in samples:
        combined = pd.concat([comp_relabun.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp_relabun.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        #sc = ax2.scatter(nsti_vals[c].loc[sample, 'weighted_NSTI'], bc_dist, color=colours[c], s=10, alpha=0.8)
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mock_vals)
        #sc = ax2.scatter(nsti_vals[c].loc[sample, 'weighted_NSTI'], stat, color=colours[c], s=10, alpha=0.8)
        nv = nsti_vals[c].loc[sample, 'weighted_NSTI']
        ap = correlations[c].append(stat), distances[c].append(bc_dist), nsti_ind[c].append(nv)
  for c in range(len(correlations)):
    for z in range(2):
      if z == 0: 
        vals = correlations[c]
        ax = ax1
      else: 
        vals = distances[c]
        ax = ax2
      par = np.polyfit(nsti_ind[c], vals, 1, full=True)
      slope, intercept = par[0][0], par[0][1]
      xl = [min(nsti_ind[c]), max(nsti_ind[c])]
      yl = [slope*xx + intercept for xx in xl]
      sc = ax.scatter(nsti_ind[c], vals, color=colours[c], s=10, alpha=0.5)
      pl = ax.plot(xl, yl, color=colours[c], linestyle='-')
      stat, p = spearmanr(nsti_ind[c], vals)
      if stat < 0: va='top'
      else: va='bottom'
      if p <= 0.05:
        tx = ax.text(xl[-1], yl[-1], 'R$^{2}$='+str(round(stat, 2))+'*', ha='center', va=va, bbox=dict(boxstyle='round', facecolor=colours[c], alpha=0.5))
      else:
        tx = ax.text(xl[-1], yl[-1], 'R$^{2}$='+str(round(stat, 2)), ha='center', va=va, bbox=dict(boxstyle='round', facecolor=colours[c], alpha=0.5))
  plt.sca(ax1)
  if traits[t] == 'KO':
    yl = plt.ylabel('Spearman correlation coefficient', fontweight='bold')
  ti = plt.title(traits[t].replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  ti = plt.title(labels[t][0], loc='left', fontweight='bold')
  plt.sca(ax2)
  xl = plt.xlabel('Weighted NSTI', fontweight='bold')
  if traits[t] == 'KO': yl = plt.ylabel('Bray-Curtis distance', fontweight='bold')
  ti = plt.title(labels[t][1], loc='left', fontweight='bold')
  databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  if t == 1: legend = ax2.legend(handles=handles, loc='upper right')
  #yl = plt.ylim([0.17, 0.73])

plt.savefig(overall_folder+'figures/fig7_weighted_nsti_correlation.png', dpi=600, bbox_inches='tight')
#plt.show()
```

## Figure 8 full-length 16S vs V4-V5 for the updated database

```{python}
plt.close()
fig = plt.figure(figsize=(10,8))

traits = ['KO', 'EC']
labels = [['A', 'C'], ['B', 'D']]
for t in range(len(traits)):
  ax1 = plt.subplot2grid((2,2),(0,t))
  ax2 = plt.subplot2grid((2,2),(1,t))
  all_correlations, all_distances = [], []
  for ds in datasets:
    mock = pd.read_csv(overall_folder+'mock/'+ds+'_'+traits[t]+'_metagenome.csv', index_col=0, header=0)
    trim = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_V4V5_'+traits[t]+'_metagenome.tsv', index_col=0, header=0, sep='\t')
    full = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_'+traits[t]+'_metagenome.tsv', index_col=0, header=0, sep='\t')
    samples = [s for s in mock.columns if s in full.columns and s in trim.columns]
    comparisons = [trim, full]
    names = ['V4-V5', 'Full-length']
    ds_correlations, ds_distances = [], []
    for c in range(len(comparisons)):
      comp = comparisons[c].copy(deep=True)
      mock_comp = mock.copy(deep=True)
      comp_relabun = comp.copy(deep=True)
      comp_relabun = comp_relabun.divide(comp_relabun.sum(axis=0), axis=1).multiply(100)
      mock_comp_relabun = mock_comp.copy(deep=True)
      mock_comp_relabun = mock_comp_relabun.divide(mock_comp_relabun.sum(axis=0), axis=1).multiply(100)
      intersection = [trait for trait in comp.index.values if trait in mock_comp.index.values]
      comp = comp.loc[intersection, :]
      mock_comp = mock_comp.loc[intersection, :]
      correlations, distances = [], []
      for sample in samples:
        combined = pd.concat([comp_relabun.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp_relabun.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mock_vals)
        ap = distances.append(bc_dist), correlations.append(stat)
      ap = ds_correlations.append(correlations), ds_distances.append(distances)
    ap = all_correlations.append(ds_correlations), all_distances.append(ds_distances)
  x = [0, 1]
  xlocs = []
  for d in range(len(datasets)):
    print(datasets[d], len(all_correlations[d][0]))
    sc = ax1.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), all_correlations[d][0], color=colours[3], alpha=0.5, s=10)
    sc = ax1.scatter(np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), all_correlations[d][1], color=colours[4], alpha=0.5, s=10)
    box = ax1.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    sc = ax2.scatter(np.random.normal(x[0], scale=0.1, size=len(all_distances[d][0])), all_distances[d][0], color=colours[3], alpha=0.5, s=10)
    sc = ax2.scatter(np.random.normal(x[1], scale=0.1, size=len(all_distances[d][1])), all_distances[d][1], color=colours[4], alpha=0.5, s=10)
    box = ax2.boxplot(all_distances[d], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xl = xlocs.append(np.mean(x))
    x[0] += 2.5
    x[1] += 2.5
  plt.sca(ax1)
  if traits[t] == 'KO':
    yl = plt.ylabel('Spearman correlation coefficient', fontweight='bold')
  xt = plt.xticks(xlocs, [])
  ti = plt.title(traits[t].replace('KO', 'KEGG orthologs').replace('EC', 'EC numbers'), fontweight='bold')
  ti = plt.title(labels[t][0], loc='left', fontweight='bold')
  #yl = plt.ylim([0.32, 0.84])
  plt.sca(ax2)
  xt = plt.xticks(xlocs, dataset_names, rotation=90)
  if traits[t] == 'KO': yl = plt.ylabel('Bray-Curtis distance', fontweight='bold')
  ti = plt.title(labels[t][1], loc='left', fontweight='bold')
  databases = ['V4-V5 region of 16S', 'Full-length 16S']
  handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s+3], markersize=12) for s in range(len(databases))]
  if t == 1: legend = ax2.legend(handles=handles, loc='upper right')
  #yl = plt.ylim([0.17, 0.73])

plt.savefig(overall_folder+'figures/fig8_mock_correlation_distance_trim_full.png', dpi=600, bbox_inches='tight')
```

## Figure 9 correlations between mock genome and closest PICRUSt2 genome

```{python, eval=FALSE}
plt.close()
fig = plt.figure(figsize=(10,16))

traits = ['ko', 'ec']
labels = [['A', 'C', 'E', 'G'], ['B', 'D', 'F', 'H']]
trait_names = ['KEGG orthologs', 'EC numbers']

for t in range(len(traits)):
  ax1 = plt.subplot2grid((4,2),(0,t))
  ax2 = plt.subplot2grid((4,2),(1,t))
  ax3 = plt.subplot2grid((4,2),(2,t))
  ax4 = plt.subplot2grid((4,2),(3,t))
  all_correlations, all_bc_distances, all_jac_distances, all_nsti = [], [], [], []
  for ds in datasets:
    comp_file = pd.read_csv(overall_folder+'correlations/'+traits[t]+'_'+ds+'.csv', index_col=1, header=0)
    all_correlations.append(list(comp_file['Spearman'].values))
    all_bc_distances.append(list(comp_file['Bray-Curtis'].values))
    all_jac_distances.append(list(comp_file['Jaccard'].values))
    all_nsti.append(list(comp_file['NSTI'].values))
  x = [0]
  xlocs = []
  for d in range(len(datasets)):
    print(datasets[d], len(all_correlations[d]))
    sc = ax1.scatter(np.random.normal(x[0], scale=0.1, size=len(all_correlations[d])), all_correlations[d], color=colours[2], alpha=0.5, s=10)
    box = ax1.boxplot([all_correlations[d]], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    sc = ax2.scatter(np.random.normal(x[0], scale=0.1, size=len(all_bc_distances[d])), all_bc_distances[d], color=colours[2], alpha=0.5, s=10)
    box = ax2.boxplot([all_bc_distances[d]], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    sc = ax3.scatter(np.random.normal(x[0], scale=0.1, size=len(all_jac_distances[d])), all_jac_distances[d], color=colours[2], alpha=0.5, s=10)
    box = ax3.boxplot([all_jac_distances[d]], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    sc = ax4.scatter(np.random.normal(x[0], scale=0.1, size=len(all_nsti[d])), all_nsti[d], color=colours[2], alpha=0.5, s=10)
    box = ax4.boxplot([all_nsti[d]], positions=x, widths=0.8, showfliers=False)
    for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
    xl = xlocs.append(np.mean(x))
    x[0] += 1.5
  plt.sca(ax1)
  if traits[t] == 'ko': yl = plt.ylabel('Spearman correlation coefficient', fontweight='bold')
  xt = plt.xticks(xlocs, [])
  ti = plt.title(trait_names[t], fontweight='bold')
  ti = plt.title(labels[t][0], loc='left', fontweight='bold')
  #yl = plt.ylim([0.32, 0.84])
  plt.sca(ax2)
  xt = plt.xticks(xlocs, [])
  if traits[t] == 'ko': yl = plt.ylabel('Bray-Curtis distance', fontweight='bold')
  ti = plt.title(labels[t][1], loc='left', fontweight='bold')
  plt.sca(ax3)
  xt = plt.xticks(xlocs, [])
  if traits[t] == 'ko': yl = plt.ylabel('Jaccard distance', fontweight='bold')
  ti = plt.title(labels[t][2], loc='left', fontweight='bold')
  plt.sca(ax4)
  xt = plt.xticks(xlocs, dataset_names, rotation=90)
  if traits[t] == 'ko': yl = plt.ylabel('NSTI', fontweight='bold')
  ti = plt.title(labels[t][3], loc='left', fontweight='bold')
  yl = plt.ylim([-0.05, 0.4])
  # databases = ['Default PICRUSt2', 'Updated PICRUSt2']
  # handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
  # if t == 1: legend = ax1.legend(handles=handles, loc='upper right')

plt.savefig(overall_folder+'figures/fig9_genome_correlations.png', dpi=600, bbox_inches='tight')
```

## Applications note overall figure

```{python}
fig = plt.figure(figsize=(15,20))

ax1 = plt.subplot2grid((60,20),(0,0),rowspan=6, colspan=17, frameon=False)
ax2 = plt.subplot2grid((60,20),(0,17),rowspan=6, colspan=3, frameon=False)
ax3 = plt.subplot2grid((60,20),(8,0), rowspan=1, colspan=10)
ax4 = plt.subplot2grid((60,20),(8,10), rowspan=1, colspan=10)
ax5 = plt.subplot2grid((60,20),(9,0), rowspan=8, colspan=10)
ax6 = plt.subplot2grid((60,20),(9,10), rowspan=8, colspan=10)
#NSTI
# ax7 = plt.subplot2grid((60,40),(19,0), rowspan=1, colspan=15)
# ax8 = plt.subplot2grid((60,40),(19,15), rowspan=1, colspan=15)
# ax9 = plt.subplot2grid((60,40),(20,0), rowspan=8, colspan=13)
# ax10 = plt.subplot2grid((60,40),(20,13), rowspan=8, colspan=2)
# ax11 = plt.subplot2grid((60,40),(20,15), rowspan=8, colspan=13)
# ax12 = plt.subplot2grid((60,40),(20,28), rowspan=8, colspan=2)
# ax13 = plt.subplot2grid((60,40),(19,30), rowspan=1, colspan=10)
# ax14 = plt.subplot2grid((60,40),(20,33), rowspan=8, colspan=7)

ax7 = plt.subplot2grid((60,40),(19,0), rowspan=1, colspan=19)
ax8 = plt.subplot2grid((60,40),(19,21), rowspan=1, colspan=19)
ax9 = plt.subplot2grid((60,40),(20,0), rowspan=8, colspan=16)
ax10 = plt.subplot2grid((60,40),(20,16), rowspan=8, colspan=3)
ax11 = plt.subplot2grid((60,40),(20,21), rowspan=8, colspan=16)
ax12 = plt.subplot2grid((60,40),(20,37), rowspan=8, colspan=3)

#Spearmans
# ax7 = plt.subplot2grid((60,40),(19,0), rowspan=1, colspan=15)
# ax8 = plt.subplot2grid((60,40),(19,15), rowspan=1, colspan=15)
# ax9 = plt.subplot2grid((60,40),(20,0), rowspan=8, colspan=13)
# ax10 = plt.subplot2grid((60,40),(20,13), rowspan=8, colspan=2)
# ax11 = plt.subplot2grid((60,40),(20,15), rowspan=8, colspan=13)
# ax12 = plt.subplot2grid((60,40),(20,28), rowspan=8, colspan=2)
# ax13 = plt.subplot2grid((60,40),(19,30), rowspan=1, colspan=10)
# ax14 = plt.subplot2grid((60,40),(20,33), rowspan=8, colspan=7)

ax_others = plt.subplot2grid((60,40),(39,0), rowspan=1, colspan=20)

ti = ax1.set_title('a) Database construction', fontweight='bold', loc='left', fontsize=16)
ti = ax3.set_title('b) Number of taxa included', fontweight='bold', loc='left', fontsize=16)
ti = ax7.set_title('c) Validation of updated PICRUSt2 database using simulated samples', fontweight='bold', loc='left', fontsize=16)

steps = ['GTDB genomes', 'GTDB reference\ngenomes', 'Present in GTDB\nphylogenetic trees', 'Have high quality\n16S rRNA gene', '10% contamination\n& 90% completion']
locs = [0, 0.9, 1.9, 2.95, 4.05]
loc_arrows = [0.5, 1.42, 2.5, 3.53, '']
genomes_at_each_step = [[394932, 7777], [80789, 4416], [77930, 4184], [31476, 1531], [26868, 1002]]
annotations_included = 'BiGG reactions\nCAZy\nEC numbers\nGO\nKO\nPfam\nGene names'

for i in range(len(steps)):
  if i != 4:
    tx = ax1.annotate(steps[i], xy=(loc_arrows[i], 1), xycoords='data', xytext=(locs[i], 1), textcoords=None, va='center', ha='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.5), fontsize=12, fontweight='bold', arrowprops=dict(facecolor='black', shrink=0.05, lw=1))
  else:
    tx = ax1.text(locs[i], 1, steps[i], ha='center', va='center', fontweight='bold', bbox=dict(boxstyle='round', facecolor='w', alpha=0.5), fontsize=12)
  tx = ax1.text(locs[i], 0.95, genomes_at_each_step[i][0], ha='center', va='center', bbox=dict(boxstyle='round', facecolor=colours[-1], alpha=0.5), fontsize=12)
  tx = ax1.text(locs[i], 0.92, genomes_at_each_step[i][1], ha='center', va='center', bbox=dict(boxstyle='round', facecolor=colours[-2], alpha=0.5), fontsize=12)

tx = ax1.text(-0.5, 0.95, 'Bacteria', ha='center', va='center', bbox=dict(boxstyle='round', facecolor=colours[-1], alpha=0.5), fontsize=12)
  tx = ax1.text(-0.5, 0.92, 'Archaea', ha='center', va='center', bbox=dict(boxstyle='round', facecolor=colours[-2], alpha=0.5), fontsize=12)

plt.sca(ax1)
li = plt.xlim([-0.4, 4.5]), plt.ylim([0.9, 1.04])
xt = plt.xticks([]), plt.yticks([])

tx = ax2.text(0.5, 0.5, annotations_included, ha='center', va='top', bbox=dict(boxstyle='round', facecolor=colours[-3], alpha=0.5), fontsize=12)
plt.sca(ax2)
li = plt.xlim([0, 1]), plt.ylim([0, 0.55])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax3)
bar = ax3.bar(0, 1, width=1, color=colours[-1], alpha=0.5)
tx = ax3.text(0, 0.5, 'Bacteria', fontweight='bold', fontsize=14, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax4)
bar = ax4.bar(0, 1, width=1, color=colours[-2], alpha=0.5)
tx = ax4.text(0, 0.5, 'Archaea', fontweight='bold', fontsize=14, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

#number of taxa
data = pd.read_csv('/Users/robynwright/Dropbox/Papers_writing/PICRUSt2_database_application_note/Number_of_genomes_old_and_new.csv', index_col=0, header=0)
levels = ['Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'Genomes']
locs = [0, 1.5, 3, 4.5, 6, 7.5, 9]
x2 = [locs[l]+0.3 for l in range(len(levels))]
x1 = [locs[l]-0.3 for l in range(len(levels))]
x1.reverse()
x2.reverse()
b = ax5.barh(x1, data.loc['PICRUSt2 updated Bacteria', :].values, color=colours[1], edgecolor='k', height=0.5)
b = ax6.barh(x1, data.loc['PICRUSt2 updated Archaea', :].values, color=colours[1], edgecolor='k', height=0.5)
b = ax5.barh(x2, data.loc['PICRUSt2 default Bacteria', :].values, color=colours[0], edgecolor='k', height=0.5)
b = ax6.barh(x2, data.loc['PICRUSt2 default Archaea', :].values, color=colours[0], edgecolor='k', height=0.5)

multiply = 0.1
adding = [10, 10, 100, 100, 100, 1000, 1000]
for l in range(len(levels)):
  upd_bac = data.loc['PICRUSt2 updated Bacteria', levels[l]]
  def_bac = data.loc['PICRUSt2 default Bacteria', levels[l]]
  upd_arc = data.loc['PICRUSt2 updated Archaea', levels[l]]
  def_arc = data.loc['PICRUSt2 default Archaea', levels[l]]
  tx = ax5.text(upd_bac+200, x1[l], str(upd_bac), ha='left', va='center', fontsize=10)
  tx = ax5.text(def_bac+200, x2[l], str(def_bac), ha='left', va='center', fontsize=10)
  tx = ax6.text(upd_arc+10, x1[l], str(upd_arc), ha='left', va='center', fontsize=10)
  tx = ax6.text(def_arc+10, x2[l], str(def_arc), ha='left', va='center', fontsize=10)


databases = ['Default PICRUSt2', 'Updated PICRUSt2']
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
#legend = ax1.legend(handles=handles, loc='upper right')
legend = ax6.legend(handles=handles, loc='upper right')

levels.reverse()
plt.sca(ax5)
yt = plt.yticks([locs[l] for l in range(len(levels))], levels), plt.xticks([])
xl = plt.xlim([-400, 31000])

plt.sca(ax6)
yt = plt.yticks([]), plt.xticks([])
xl = plt.xlim([-15, 1150])

plt.sca(ax7)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, 'NSTI (per sequence)', ha='center', va='center', fontsize=12, fontweight='bold')

plt.sca(ax8)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, 'Weighted NSTI (per sample)', ha='center', va='center', fontsize=12, fontweight='bold')

plt.sca(ax9)
x = [0, 1]
xlocs = []
all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
ds_average = {'blueberry':[[], []], 'cameroon':[[], []], 'hmp':[[], []], 'indian':[[], []], 'mammal':[[], []], 'ocean':[[], []], 'primate':[[], []], 'overall':[[], []]}
for ds in datasets:
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
  asvs = list(set(list(default.index.values)+list(updated.index.values)))
  for asv in asvs:
    if asv in default.index.values:
      def_nsti = default.loc[asv, 'metadata_NSTI']
      all_nsti_default.append(def_nsti)
    else: def_nsti = 'NA'
    if asv in updated.index.values:
      upd_nsti = updated.loc[asv, 'metadata_NSTI']
      all_nsti_updated.append(upd_nsti)
    else: upd_nsti = 'NA'
    if def_nsti == 'NA' or upd_nsti == 'NA':
      continue
    else:
      nsti_difference.append(upd_nsti-def_nsti)
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  sc = ax9.scatter(np.random.normal(x[0], scale=0.1, size=len(default)), default, color=colours[0], alpha=0.5, s=10)
  sc = ax9.scatter(np.random.normal(x[1], scale=0.1, size=len(updated)), updated, color=colours[1], alpha=0.5, s=10)
  ds_average[ds][0].append(np.median(default))
  ds_average[ds][1].append(np.median(updated))
  box = ax9.boxplot([default, updated], positions=x, widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] += 2.5
  x[1] += 2.5

yl = plt.ylim([-0.05, 1])
xt = plt.xticks(xlocs, dataset_names, rotation=45)
yl = plt.ylabel('Nearest Sequenced Taxon\nIndex (NSTI)', fontweight='bold')

plt.sca(ax10)
x = [0, 1]
sc = ax10.scatter(np.random.normal(x[0], scale=0.1, size=len(all_nsti_default)), all_nsti_default, color=colours[0], alpha=0.5, s=10)
sc = ax10.scatter(np.random.normal(x[1], scale=0.1, size=len(all_nsti_updated)), all_nsti_updated, color=colours[1], alpha=0.5, s=10)
yl = plt.ylim([-0.05, 1])
box = ax10.boxplot([all_nsti_default, all_nsti_updated], positions=x, widths=0.8, showfliers=False)
xt = plt.xticks([0.5], ['Overall'], rotation=45), plt.yticks([])
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
ds_average['overall'][0].append(np.median(default))
ds_average['overall'][1].append(np.median(updated))

plt.sca(ax11)
x = [0, 1]
xlocs = []
all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
means_default, medians_default, means_updated, medians_updated = [], [], [], []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_weighted_nsti.tsv', index_col=0, header=0, sep='\t')
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_weighted_nsti.tsv', index_col=0, header=0, sep='\t')
  asvs = list(set(list(default.index.values)+list(updated.index.values)))
  for asv in asvs:
    if asv in default.index.values:
      def_nsti = default.loc[asv, 'weighted_NSTI']
      all_nsti_default.append(def_nsti)
    else: def_nsti = 'NA'
    if asv in updated.index.values:
      upd_nsti = updated.loc[asv, 'weighted_NSTI']
      all_nsti_updated.append(upd_nsti)
    else: upd_nsti = 'NA'
    if def_nsti == 'NA' or upd_nsti == 'NA':
      continue
    else:
      nsti_difference.append(upd_nsti-def_nsti)
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  sc = ax11.scatter(np.random.normal(x[0], scale=0.1, size=len(default)), default, color=colours[0], alpha=0.5, s=10)
  sc = ax11.scatter(np.random.normal(x[1], scale=0.1, size=len(updated)), updated, color=colours[1], alpha=0.5, s=10)
  ds_average[ds][0].append(np.median(default))
  ds_average[ds][1].append(np.median(updated))
  box = ax11.boxplot([default, updated], positions=x, widths=0.8, showfliers=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] += 2.5
  x[1] += 2.5

yl = plt.ylim([-0.05, 0.5])
xt = plt.xticks(xlocs, dataset_names, rotation=45)
#yl = plt.ylabel('Nearest Sequenced Taxon\nIndex (NSTI)', fontweight='bold')

plt.sca(ax12)
x = [0, 1]
sc = ax12.scatter(np.random.normal(x[0], scale=0.1, size=len(all_nsti_default)), all_nsti_default, color=colours[0], alpha=0.5, s=10)
sc = ax12.scatter(np.random.normal(x[1], scale=0.1, size=len(all_nsti_updated)), all_nsti_updated, color=colours[1], alpha=0.5, s=10)
yl = plt.ylim([-0.05, 0.5])
box = ax12.boxplot([all_nsti_default, all_nsti_updated], positions=x, widths=0.8, showfliers=False)
xt = plt.xticks([0.5], ['Overall'], rotation=45), plt.yticks([])
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
ds_average['overall'][0].append(np.median(default))
ds_average['overall'][1].append(np.median(updated))

# plt.sca(ax14)
# y = [8, 7, 6, 5, 4, 3, 2, 1]
# names = datasets+['Overall']
# for a in range(len(y)):
#   ba = ax14.bar([0, 1, 2, 3], [1, 1, 1, 1], bottom=[y[a], y[a], y[a], y[a]], width=1, color='w', edgecolor='k')
#   tx = ax14.text(0, y[a]+0.5, round(medians_default[a], 3), ha='center', va='center', fontsize=10)
#   tx = ax14.text(1, y[a]+0.5, round(medians_updated[a], 3), ha='center', va='center', fontsize=10)
# 
# yt = plt.yticks([ya+0.5 for ya in y], dataset_names+['Overall'])
# li = plt.xlim([-0.5, 3.5]), plt.ylim([1, 9])
# xt = plt.xticks([0, 1, 2, 3], ['Default', 'Updated', 'Default\nweighted', 'Updated\nweighted'], rotation=90)
# 
# y = [8, 7, 6, 5, 4, 3, 2, 1]
# for a in range(len(y)):
#   tx = ax14.text(2, y[a]+0.5, round(medians_default[a], 3), ha='center', va='center', fontsize=10)
#   tx = ax14.text(3, y[a]+0.5, round(medians_updated[a], 3), ha='center', va='center', fontsize=10)

plt.savefig(overall_folder+'figures/overall_figure_old.png', dpi=600, bbox_inches='tight')
```

```{python}
fig = plt.figure(figsize=(15,20))

ax1 = plt.subplot2grid((60,20),(0,0),rowspan=6, colspan=17, frameon=False)
ax2 = plt.subplot2grid((60,20),(0,17),rowspan=6, colspan=3, frameon=False)
ax3 = plt.subplot2grid((60,20),(8,0), rowspan=1, colspan=10)
ax4 = plt.subplot2grid((60,20),(8,10), rowspan=1, colspan=10)
ax5 = plt.subplot2grid((60,20),(9,0), rowspan=8, colspan=10)
ax6 = plt.subplot2grid((60,20),(9,10), rowspan=8, colspan=10)
#NSTI
ax7 = plt.subplot2grid((60,40),(19,0), rowspan=1, colspan=14)
ax8 = plt.subplot2grid((60,40),(19,14), rowspan=1, colspan=26)
ax9 = plt.subplot2grid((60,40),(20,17), rowspan=14, colspan=11)
ax10 = plt.subplot2grid((60,40),(34,17), rowspan=2, colspan=11)
ax11 = plt.subplot2grid((60,40),(20,29), rowspan=14, colspan=11)
ax12 = plt.subplot2grid((60,40),(34,29), rowspan=2, colspan=11)
#tax
ax13 = plt.subplot2grid((60,40),(20,0), rowspan=16, colspan=6)

#Spearmans
ax14 = plt.subplot2grid((60,40),(40,0), rowspan=1, colspan=11)
ax15 = plt.subplot2grid((60,40),(40,12), rowspan=1, colspan=11)
ax16 = plt.subplot2grid((60,40),(41,0), rowspan=14, colspan=11)
#ax17 = plt.subplot2grid((60,40),(55,0), rowspan=2, colspan=11)
ax18 = plt.subplot2grid((60,40),(41,12), rowspan=14, colspan=11)
#ax19 = plt.subplot2grid((60,40),(55,12), rowspan=2, colspan=11)

ti = ax1.set_title('a) Database construction', fontweight='bold', loc='left', fontsize=16)
ti = ax3.set_title('b) Number of taxa included', fontweight='bold', loc='left', fontsize=16)
ti = ax7.set_title('c) Validation of updated PICRUSt2 database using simulated samples', fontweight='bold', loc='left', fontsize=16)

steps = ['GTDB genomes', 'GTDB reference\ngenomes', 'Present in GTDB\nphylogenetic trees', 'Have high quality\n16S rRNA gene', '10% contamination\n& 90% completion']
locs = [0, 0.9, 1.9, 2.95, 4.05]
loc_arrows = [0.5, 1.42, 2.5, 3.53, '']
genomes_at_each_step = [[394932, 7777], [80789, 4416], [77930, 4184], [31476, 1531], [26868, 1002]]
annotations_included = 'BiGG reactions\nCAZy\nEC numbers\nGO\nKO\nPfam\nGene names'

for i in range(len(steps)):
  if i != 4:
    tx = ax1.annotate(steps[i], xy=(loc_arrows[i], 1), xycoords='data', xytext=(locs[i], 1), textcoords=None, va='center', ha='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.5), fontsize=12, fontweight='bold', arrowprops=dict(facecolor='black', shrink=0.05, lw=1))
  else:
    tx = ax1.text(locs[i], 1, steps[i], ha='center', va='center', fontweight='bold', bbox=dict(boxstyle='round', facecolor='w', alpha=0.5), fontsize=12)
  tx = ax1.text(locs[i], 0.95, genomes_at_each_step[i][0], ha='center', va='center', bbox=dict(boxstyle='round', facecolor=colours[-1], alpha=0.5), fontsize=12)
  tx = ax1.text(locs[i], 0.92, genomes_at_each_step[i][1], ha='center', va='center', bbox=dict(boxstyle='round', facecolor=colours[-2], alpha=0.5), fontsize=12)

tx = ax1.text(-0.5, 0.95, 'Bacteria', ha='center', va='center', bbox=dict(boxstyle='round', facecolor=colours[-1], alpha=0.5), fontsize=12)
  tx = ax1.text(-0.5, 0.92, 'Archaea', ha='center', va='center', bbox=dict(boxstyle='round', facecolor=colours[-2], alpha=0.5), fontsize=12)

plt.sca(ax1)
li = plt.xlim([-0.4, 4.5]), plt.ylim([0.9, 1.04])
xt = plt.xticks([]), plt.yticks([])

tx = ax2.text(0.5, 0.5, annotations_included, ha='center', va='top', bbox=dict(boxstyle='round', facecolor=colours[-3], alpha=0.5), fontsize=12)
plt.sca(ax2)
li = plt.xlim([0, 1]), plt.ylim([0, 0.55])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax3)
bar = ax3.bar(0, 1, width=1, color=colours[-1], alpha=0.5)
tx = ax3.text(0, 0.5, 'Bacteria', fontweight='bold', fontsize=14, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax4)
bar = ax4.bar(0, 1, width=1, color=colours[-2], alpha=0.5)
tx = ax4.text(0, 0.5, 'Archaea', fontweight='bold', fontsize=14, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

#number of taxa
data = pd.read_csv('/Users/robynwright/Dropbox/Papers_writing/PICRUSt2_database_application_note/Number_of_genomes_old_and_new.csv', index_col=0, header=0)
levels = ['Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'Genomes']
locs = [0, 1.5, 3, 4.5, 6, 7.5, 9]
x2 = [locs[l]+0.3 for l in range(len(levels))]
x1 = [locs[l]-0.3 for l in range(len(levels))]
x1.reverse()
x2.reverse()
b = ax5.barh(x1, data.loc['PICRUSt2 updated Bacteria', :].values, color=colours[1], edgecolor='k', height=0.5)
b = ax6.barh(x1, data.loc['PICRUSt2 updated Archaea', :].values, color=colours[1], edgecolor='k', height=0.5)
b = ax5.barh(x2, data.loc['PICRUSt2 default Bacteria', :].values, color=colours[0], edgecolor='k', height=0.5)
b = ax6.barh(x2, data.loc['PICRUSt2 default Archaea', :].values, color=colours[0], edgecolor='k', height=0.5)

multiply = 0.1
adding = [10, 10, 100, 100, 100, 1000, 1000]
for l in range(len(levels)):
  upd_bac = data.loc['PICRUSt2 updated Bacteria', levels[l]]
  def_bac = data.loc['PICRUSt2 default Bacteria', levels[l]]
  upd_arc = data.loc['PICRUSt2 updated Archaea', levels[l]]
  def_arc = data.loc['PICRUSt2 default Archaea', levels[l]]
  tx = ax5.text(upd_bac+200, x1[l], str(upd_bac), ha='left', va='center', fontsize=10)
  tx = ax5.text(def_bac+200, x2[l], str(def_bac), ha='left', va='center', fontsize=10)
  tx = ax6.text(upd_arc+10, x1[l], str(upd_arc), ha='left', va='center', fontsize=10)
  tx = ax6.text(def_arc+10, x2[l], str(def_arc), ha='left', va='center', fontsize=10)


databases = ['Default PICRUSt2', 'Updated PICRUSt2']
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
#legend = ax1.legend(handles=handles, loc='upper right')
legend = ax6.legend(handles=handles, loc='upper right')

levels.reverse()
plt.sca(ax5)
yt = plt.yticks([locs[l] for l in range(len(levels))], levels), plt.xticks([])
xl = plt.xlim([-400, 31000])

plt.sca(ax6)
yt = plt.yticks([]), plt.xticks([])
xl = plt.xlim([-15, 1150])

plt.sca(ax7)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, 'Simulated sample composition', ha='center', va='center', fontsize=12, fontweight='bold')

plt.sca(ax8)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, 'Nearest Sequenced Taxon Indices (NSTI)', ha='center', va='center', fontsize=12, fontweight='bold')

plt.sca(ax9)
x = [18.5, 17.5]
xlocs = []
all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
  asvs = list(set(list(default.index.values)+list(updated.index.values)))
  for asv in asvs:
    if asv in default.index.values:
      def_nsti = default.loc[asv, 'metadata_NSTI']
      all_nsti_default.append(def_nsti)
    else: def_nsti = 'NA'
    if asv in updated.index.values:
      upd_nsti = updated.loc[asv, 'metadata_NSTI']
      all_nsti_updated.append(upd_nsti)
    else: upd_nsti = 'NA'
    if def_nsti == 'NA' or upd_nsti == 'NA':
      continue
    else:
      nsti_difference.append(upd_nsti-def_nsti)
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  sc = ax9.scatter(default, np.random.normal(x[0], scale=0.1, size=len(default)), color=colours[0], alpha=0.5, s=10)
  sc = ax9.scatter(updated, np.random.normal(x[1], scale=0.1, size=len(updated)), color=colours[1], alpha=0.5, s=10)
  fc, alp = 'w', 0.7
  if np.median(default) < np.median(updated): fc, alp = 'gray', 0.6
  tx = ax9.text(0.97, x[0], round(np.median(default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax9.text(0.97, x[1], round(np.median(updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  box = ax9.boxplot([default, updated], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] -= 2.5
  x[1] -= 2.5

yl = plt.xlim([-0.05, 1])
xt = plt.yticks(xlocs, dataset_names), plt.xticks([])

# 
plt.sca(ax10)
x = [1, 0]
sc = ax10.scatter(all_nsti_default, np.random.normal(x[0], scale=0.1, size=len(all_nsti_default)), color=colours[0], alpha=0.5, s=10)
sc = ax10.scatter(all_nsti_updated, np.random.normal(x[1], scale=0.1, size=len(all_nsti_updated)), color=colours[1], alpha=0.5, s=10)
yl = plt.xlim([-0.05, 1])
box = ax10.boxplot([all_nsti_default, all_nsti_updated], positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([0.5], ['Overall'])
xl = plt.xlabel('NSTI per sequence', fontweight='bold')
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tx = ax10.text(0.97, x[0], round(np.median(all_nsti_default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))
tx = ax10.text(0.97, x[1], round(np.median(all_nsti_updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))

plt.sca(ax11)
x = [18.5, 17.5]
xlocs = []
all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_weighted_nsti.tsv', index_col=0, header=0, sep='\t')#['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_weighted_nsti.tsv', index_col=0, header=0, sep='\t')#['weighted_NSTI'].values
  asvs = list(set(list(default.index.values)+list(updated.index.values)))
  for asv in asvs:
    if asv in default.index.values:
      def_nsti = default.loc[asv, 'weighted_NSTI']
      all_nsti_default.append(def_nsti)
    else: def_nsti = 'NA'
    if asv in updated.index.values:
      upd_nsti = updated.loc[asv, 'weighted_NSTI']
      all_nsti_updated.append(upd_nsti)
    else: upd_nsti = 'NA'
    if def_nsti == 'NA' or upd_nsti == 'NA':
      continue
    else:
      nsti_difference.append(upd_nsti-def_nsti)
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  sc = ax11.scatter(default, np.random.normal(x[0], scale=0.1, size=len(default)), color=colours[0], alpha=0.5, s=10)
  sc = ax11.scatter(updated, np.random.normal(x[1], scale=0.1, size=len(updated)), color=colours[1], alpha=0.5, s=10)
  fc, alp = 'w', 0.7
  if np.median(default) < np.median(updated): fc, alp = 'gray', 0.6
  tx = ax11.text(0.68, x[0], round(np.median(default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax11.text(0.68, x[1], round(np.median(updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  box = ax11.boxplot([default, updated], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] -= 2.5
  x[1] -= 2.5

yl = plt.xlim([-0.02, 0.7])
xt = plt.yticks([]), plt.xticks([])

# 
plt.sca(ax12)
x = [1, 0]
sc = ax12.scatter(all_nsti_default, np.random.normal(x[0], scale=0.1, size=len(all_nsti_default)), color=colours[0], alpha=0.5, s=10)
sc = ax12.scatter(all_nsti_updated, np.random.normal(x[1], scale=0.1, size=len(all_nsti_updated)), color=colours[1], alpha=0.5, s=10)
yl = plt.xlim([-0.02, 0.7])
box = ax12.boxplot([all_nsti_default, all_nsti_updated], positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([])
xl = plt.xlabel('NSTI per sample', fontweight='bold')
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tx = ax12.text(0.68, x[0], round(np.median(all_nsti_default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))
tx = ax12.text(0.68, x[1], round(np.median(all_nsti_updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))

#taxonomy

tax = pd.read_csv(overall_folder+'mock/genome_full_tax.csv', index_col=0, header=0)

tax_genus, tax_class, tax_family = {}, {}, {}
for row in tax.index.values:
  this_tax = tax.loc[row, 'Taxonomy'].split(';')
  tax_genus[row] = this_tax[-2]
  tax_family[row] = this_tax[-3]
  tax_class[row] = this_tax[2].replace('c__', '')

combined = []
for ds in datasets:
  ft = pd.read_csv(overall_folder+'mock/'+ds+'_genome_abundance.csv', index_col=0, header=0).drop('Tax', axis=1)
  ft = ft.rename(index=tax_class)
  ft = ft.groupby(by=ft.index).sum()
  if 'Unassigned' in ft.index.values: ft = ft.drop('Unassigned')
  ft = ft.divide(ft.sum(axis=0), axis=1).multiply(100)
  ft[ds] = ft.mean(axis=1)
  combined.append(ft.loc[:, [ds]])

combined_df = pd.concat(combined).fillna(value=0)
combined_df = combined_df.groupby(by=combined_df.index).sum()
combined_df['Mean'] = combined_df.mean(axis=1)
combined_df = combined_df.sort_values(by=['Mean'], ascending=False)
combined_df = combined_df.iloc[:19, :]
combined_df = combined_df.drop('Mean', axis=1)
combined_df = combined_df.loc[sorted(combined_df.index.values), :]
combined_df.loc['Other', :] = 100-combined_df.sum(axis=0).values
plt.sca(ax13)
a1 = combined_df.transpose().plot(ax=ax13, kind='bar', stacked=True, width=0.9, cmap='tab20', edgecolor='k')
lg = plt.legend(loc='upper left', bbox_to_anchor=(1.1, 1.02), fontsize=10)
yl = plt.ylabel('Relative abundance (%)')
yl = plt.ylim([0, 100])
xt = plt.xticks([d for d in range(len(dataset_names))], dataset_names, rotation=90)

plt.sca(ax14)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, "Spearman's correlation", ha='center', va='center', fontsize=12, fontweight='bold')

plt.sca(ax15)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, 'Bray-Curtis dissimilarity', ha='center', va='center', fontsize=12, fontweight='bold')

x = [18.5, 17.5]
xlocs = []
all_correlations, all_distances = [], []
for ds in datasets:
  mock = pd.read_csv(overall_folder+'mock/'+ds+'_EC_metagenome_norm.csv', index_col=0, header=0)
  mock = mock.drop('-', axis=0)
  mock = mock[mock.max(axis=1) > 0]
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_EC_metagenome.tsv', index_col=0, header=0, sep='\t')
  rename = {}
  for row in default.index.values:
    if traits[t] == 'KO': rename[row] = 'ko:'+row
    else: rename[row] = row.split(':')[1]
  default = default.rename(index=rename)
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_EC_metagenome.tsv', index_col=0, header=0, sep='\t')
  samples = [s for s in mock.columns if s in default.columns and s in updated.columns]
  comparisons = [default, updated]
  names = ['Default', 'Updated']
  ds_correlations, ds_distances = [], []
  for c in range(len(comparisons)):
    comp = comparisons[c].copy(deep=True)
    mock_comp = mock.copy(deep=True)
    comp_relabun = comp.copy(deep=True)
    comp_relabun = comp_relabun.divide(comp_relabun.sum(axis=0), axis=1).multiply(100)
    mock_comp_relabun = mock_comp.copy(deep=True)
    mock_comp_relabun = mock_comp_relabun.divide(mock_comp_relabun.sum(axis=0), axis=1).multiply(100)
    intersection = [trait for trait in comp.index.values if trait in mock_comp.index.values]
    comp = comp.loc[intersection, :]
    mock_comp = mock_comp.loc[intersection, :]
    correlations, distances = [], []
    for sample in samples:
        combined = pd.concat([comp_relabun.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp_relabun.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mock_vals)
        ap = distances.append(bc_dist), correlations.append(stat)
    ap = ds_correlations.append(correlations), ds_distances.append(distances)
  ap = all_correlations.append(ds_correlations), all_distances.append(ds_distances)
for d in range(len(datasets)):
  sc = ax16.scatter(all_correlations[d][0], np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), color=colours[0], alpha=0.5, s=10)
  sc = ax16.scatter(all_correlations[d][1], np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), color=colours[1], alpha=0.5, s=10)
  fc, alp = 'w', 0.7
  if np.median(all_correlations[d][0]) > np.median(all_correlations[d][1]): fc, alp = 'gray', 0.6
  tx = ax16.text(0.57, x[0], round(np.median(all_correlations[d][0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax16.text(0.57, x[1], round(np.median(all_correlations[d][1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  box = ax16.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  sc = ax18.scatter(all_distances[d][0], np.random.normal(x[0], scale=0.1, size=len(all_distances[d][0])), color=colours[0], alpha=0.5, s=10)
  sc = ax18.scatter(all_distances[d][1], np.random.normal(x[1], scale=0.1, size=len(all_distances[d][1])), color=colours[1], alpha=0.5, s=10)
  fc, alp = 'w', 0.7
  if np.median(all_distances[d][0]) < np.median(all_distances[d][1]): fc, alp = 'gray', 0.6
  tx = ax18.text(0.74, x[0], round(np.median(all_distances[d][0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax18.text(0.74, x[1], round(np.median(all_distances[d][1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  box = ax18.boxplot(all_distances[d], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xl = xlocs.append(np.mean(x))
  x[0] += 2.5
  x[1] += 2.5

plt.sca(ax16)
xl = plt.xlabel('Spearman correlation coefficient', fontweight='bold')
yt = plt.yticks(xlocs, dataset_names)
xl = plt.xlim([0.36, 0.58])

plt.sca(ax18)
xl = plt.xlabel('Bray-Curtis dissimilarity index', fontweight='bold')
yt = plt.yticks([])
xl = plt.xlim([0.4, 0.75])

plt.savefig(overall_folder+'figures/overall_figure_ec_long.png', dpi=600, bbox_inches='tight')
```

```{python}
fig = plt.figure(figsize=(15,20))

ax_db = plt.subplot2grid((60,20),(0,0),rowspan=5, colspan=5, frameon=False)
ax_annot_def_title = plt.subplot2grid((60,20),(7,0),rowspan=1, colspan=2)
ax_annot_upd_title = plt.subplot2grid((60,20),(7,2),rowspan=1, colspan=2)
ax_annot_def = plt.subplot2grid((60,20),(8,0),rowspan=8, colspan=2)
ax_annot_upd = plt.subplot2grid((60,20),(8,2),rowspan=8, colspan=2)
ax1 = plt.subplot2grid((60,20),(0,6),rowspan=1, colspan=4)
ax2 = plt.subplot2grid((60,20),(1,6),rowspan=5, colspan=4)
ax3 = plt.subplot2grid((60,20),(0,10), rowspan=1, colspan=4)
ax4 = plt.subplot2grid((60,20),(1,10), rowspan=5, colspan=4)
ax5 = plt.subplot2grid((60,20),(6,6), rowspan=9, colspan=4)
ax6 = plt.subplot2grid((60,20),(6,10), rowspan=9, colspan=4)
ax_leg = plt.subplot2grid((60,20),(15,6), rowspan=2, colspan=8, frameon=False)
#NSTI

ax8 = plt.subplot2grid((60,20),(19,0), rowspan=1, colspan=5)
ax8_2 = plt.subplot2grid((60,20),(19,5), rowspan=1, colspan=5)
ax9 = plt.subplot2grid((60,20),(20,0), rowspan=14, colspan=5)
ax10 = plt.subplot2grid((60,20),(34,0), rowspan=2, colspan=5)
ax11 = plt.subplot2grid((60,20),(20,5), rowspan=14, colspan=5)
ax12 = plt.subplot2grid((60,20),(34,5), rowspan=2, colspan=5)
#tax
ax13 = plt.subplot2grid((60,20),(0,15), rowspan=14, colspan=2)

#Spearmans
ax14 = plt.subplot2grid((60,20),(19,10), rowspan=1, colspan=5)
ax15 = plt.subplot2grid((60,20),(19,15), rowspan=1, colspan=5)
ax16 = plt.subplot2grid((60,20),(20,10), rowspan=14, colspan=5)
ax17 = plt.subplot2grid((60,20),(34,10), rowspan=2, colspan=5)
ax18 = plt.subplot2grid((60,20),(20,15), rowspan=14, colspan=5)
ax19 = plt.subplot2grid((60,20),(34,15), rowspan=2, colspan=5)

ti = ax_db.set_title('a) Database construction', fontweight='bold', loc='left', fontsize=16)
ti = ax_annot_def_title.set_title('b) Functions included', fontweight='bold', loc='left', fontsize=16)
ti = ax1.set_title('c) Number of taxa included', fontweight='bold', loc='left', fontsize=16)
ti = ax13.set_title('d) Simulated sample', fontweight='bold', loc='left', fontsize=16)
ti = ax8.set_title('e) Performance of PICRUSt2 databases on simulated samples', fontweight='bold', loc='left', fontsize=16)

steps = ['Step 1: Get GTDB genomes\n', 'Step 2: Filter to GTDB reference genomes\n', 'Step 3: Annotate genomes with Eggnog\n', 'Step 4: Identify genomes present in GTDB\n            phylogenetic trees\n', 'Step 5: Identify genomes with 16S\n', 'Step 6: Keep only genomes 10% cont-\n            amination & 90% completion']
string = ''
for step in steps: string += step
tx = ax_db.text(-0.7, 1.03, string, ha='left', va='top', fontsize=10)

plt.sca(ax_db)
li = plt.xlim([0, 1]), plt.ylim([0, 1])

plt.sca(ax_db)
li = plt.xlim([-0.4, 4.5]), plt.ylim([0.9, 1.04])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax1)
bar = ax1.bar(0, 1, width=1, color=colours[-1], alpha=0.5)
tx = ax1.text(0, 0.5, 'Bacteria', fontweight='bold', fontsize=12, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax_annot_def_title)
tx = ax_annot_def_title.text(0, 0.5, 'Default', fontweight='bold', fontsize=12, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax_annot_upd_title)
tx = ax_annot_upd_title.text(0, 0.5, 'Updated', fontweight='bold', fontsize=12, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax3)
bar = ax3.bar(0, 1, width=1, color=colours[-2], alpha=0.5)
tx = ax3.text(0, 0.5, 'Archaea', fontweight='bold', fontsize=12, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

# #number of taxa
data = pd.read_csv('/Users/robynwright/Dropbox/Papers_writing/PICRUSt2_database_application_note/Number_of_genomes_old_and_new.csv', index_col=0, header=0)
levels = ['Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'Genomes']
locs = [0, 1.5, 3, 4.5, 6, 7.5, 9]
x2 = [locs[l]+0.3 for l in range(len(levels))]
x1 = [locs[l]-0.3 for l in range(len(levels))]
b = ax5.barh(x1, data.loc['PICRUSt2 updated Bacteria', :].values, color=colours[1], edgecolor='k', height=0.5)
b = ax6.barh(x1, data.loc['PICRUSt2 updated Archaea', :].values, color=colours[1], edgecolor='k', height=0.5)
b = ax5.barh(x2, data.loc['PICRUSt2 default Bacteria', :].values, color=colours[0], edgecolor='k', height=0.5)
b = ax6.barh(x2, data.loc['PICRUSt2 default Archaea', :].values, color=colours[0], edgecolor='k', height=0.5)
# 
multiply = 0.1
adding = [10, 10, 100, 100, 100, 1000, 1000]
for l in range(len(levels)):
  upd_bac = data.loc['PICRUSt2 updated Bacteria', levels[l]]
  def_bac = data.loc['PICRUSt2 default Bacteria', levels[l]]
  upd_arc = data.loc['PICRUSt2 updated Archaea', levels[l]]
  def_arc = data.loc['PICRUSt2 default Archaea', levels[l]]
  tx = ax5.text(upd_bac+200, x1[l], str(upd_bac), ha='left', va='center', fontsize=10)
  tx = ax5.text(def_bac+200, x2[l], str(def_bac), ha='left', va='center', fontsize=10)
  tx = ax6.text(upd_arc+10, x1[l], str(upd_arc), ha='left', va='center', fontsize=10)
  tx = ax6.text(def_arc+10, x2[l], str(def_arc), ha='left', va='center', fontsize=10)

genomes_at_each_step = [[394932, 7777], [80789, 4416], [77930, 4184], [31476, 1531], [26868, 1002]]
#ax2, ax4
y = [5, 4, 3, 2, 1]
steps = ['Step 1', 'Step 2', 'Step 4', 'Step 5', 'Step 6']

for a in range(len(genomes_at_each_step)):
  ba = ax2.barh(y[a], genomes_at_each_step[a][0], color=colours[2], edgecolor='k')
  tx = ax2.text(genomes_at_each_step[a][0]+5000, y[a], genomes_at_each_step[a][0], ha='left', va='center')
  ba = ax4.barh(y[a], genomes_at_each_step[a][1], color=colours[2], edgecolor='k')
  tx = ax4.text(genomes_at_each_step[a][1]+100, y[a], genomes_at_each_step[a][1], ha='left', va='center')
  
plt.sca(ax2)
xl = plt.xlim([0, 530000])
yt = plt.yticks(y, steps), plt.xticks([])

plt.sca(ax4)
xl = plt.xlim([0, 9500])
yt = plt.yticks([]), plt.xticks([])

databases = ['Default PICRUSt2\ndatabase', 'Updated PICRUSt2\ndatabase']
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
legend = ax_leg.legend(handles=handles, loc='center', ncol=2)
plt.sca(ax_leg)
xt = plt.xticks([]), plt.yticks([])

functions = ['BiGG reactions', 'CAZy', 'COG', 'EC numbers', 'Gene names', 'GO', 'KO', 'Pfam', 'Phenotypes', 'TIGRFAM']
y = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
default_counts = ['-', '-', 4598, 2913, '-', '-', 10543, 11089, 41, 4287]
updated_counts = []
names = ['bigg_reaction', 'cazy', '', 'ec', 'preferred_name', 'go', 'ko', 'pfam', '', '']

# for name in names:
#   if name == '': 
#     updated_counts.append(0)
#     continue
#   bac = pd.read_csv(filepath_or_buffer='/Users/robynwright/Dropbox/Langille_Lab_postdoc/Github/picrust2/picrust2/default_files/bacteria/'+name+'.txt.gz', index_col=0, header=0, sep='\t')
#   arc = pd.read_csv(filepath_or_buffer='/Users/robynwright/Dropbox/Langille_Lab_postdoc/Github/picrust2/picrust2/default_files/archaea/'+name+'.txt.gz', index_col=0, header=0, sep='\t')
#   names = list(set(list(bac.columns)+list(arc.columns)))
#   updated_counts.append(len(names))
#
# print(updated_counts)

updated_counts = [10452, 151, '-', 3763, 31131, 23273, 14106, 11402, '-', '-']
for f in range(len(functions)):
  if default_counts[f] == '-':
    ba = ax_annot_def.bar(0, 1, bottom=y[f]-0.5, width=1, edgecolor='k', color='gray', alpha=0.2)
  else:
    ba = ax_annot_def.bar(0, 1, bottom=y[f]-0.5, width=1, edgecolor='k', color='w')
  tx = ax_annot_def.text(0, y[f], default_counts[f], ha='center', va='center')
  if updated_counts[f] == '-':
    ba = ax_annot_upd.bar(0, 1, bottom=y[f]-0.5, width=1, edgecolor='k', color='gray', alpha=0.2)
  else:
    ba = ax_annot_upd.bar(0, 1, bottom=y[f]-0.5, width=1, edgecolor='k', color='w')
  tx = ax_annot_upd.text(0, y[f], updated_counts[f], ha='center', va='center')

plt.sca(ax_annot_def)
yt = plt.yticks(y, functions), plt.xticks([])
li = plt.ylim([0.5, 10.5]), plt.xlim([-0.5, 0.5])

plt.sca(ax_annot_upd)
yt = plt.yticks([]), plt.xticks([])
li = plt.ylim([0.5, 10.5]), plt.xlim([-0.5, 0.5])

plt.sca(ax5)
yt = plt.yticks([locs[l] for l in range(len(levels))], [l.replace('Genomes', 'Genome') for l in levels]), plt.xticks([])
xl = plt.xlim([-400, 34500])
# 
plt.sca(ax6)
yt = plt.yticks([]), plt.xticks([])
xl = plt.xlim([-15, 1280])

plt.sca(ax7)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, 'Simulated sample composition', ha='center', va='center', fontsize=12, fontweight='bold')

plt.sca(ax8)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, 'NSTI per sequence', ha='center', va='center', fontsize=12, fontweight='bold')

plt.sca(ax8_2)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, 'NSTI per sample', ha='center', va='center', fontsize=12, fontweight='bold')

plt.sca(ax9)
x = [18.5, 17.5]
xlocs = []
all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
  asvs = list(set(list(default.index.values)+list(updated.index.values)))
  for asv in asvs:
    if asv in default.index.values:
      def_nsti = default.loc[asv, 'metadata_NSTI']
      all_nsti_default.append(def_nsti)
    else: def_nsti = 'NA'
    if asv in updated.index.values:
      upd_nsti = updated.loc[asv, 'metadata_NSTI']
      all_nsti_updated.append(upd_nsti)
    else: upd_nsti = 'NA'
    if def_nsti == 'NA' or upd_nsti == 'NA':
      continue
    else:
      nsti_difference.append(upd_nsti-def_nsti)
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
  sc = ax9.scatter(default, np.random.normal(x[0], scale=0.1, size=len(default)), color=colours[0], alpha=0.5, s=10)
  sc = ax9.scatter(updated, np.random.normal(x[1], scale=0.1, size=len(updated)), color=colours[1], alpha=0.5, s=10)
  fc, alp = 'w', 0.7
  if np.median(default) < np.median(updated): fc, alp = 'gray', 0.6
  tx = ax9.text(0.97, x[0], round(np.median(default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax9.text(0.97, x[1], round(np.median(updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  box = ax9.boxplot([default, updated], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] -= 2.5
  x[1] -= 2.5

yl = plt.xlim([-0.05, 1])
xt = plt.yticks(xlocs, dataset_names), plt.xticks([])

#
plt.sca(ax10)
x = [1, 0]
sc = ax10.scatter(all_nsti_default, np.random.normal(x[0], scale=0.1, size=len(all_nsti_default)), color=colours[0], alpha=0.5, s=10)
sc = ax10.scatter(all_nsti_updated, np.random.normal(x[1], scale=0.1, size=len(all_nsti_updated)), color=colours[1], alpha=0.5, s=10)
yl = plt.xlim([-0.05, 1])
box = ax10.boxplot([all_nsti_default, all_nsti_updated], positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([0.5], ['Overall'])
xl = plt.xlabel('NSTI per sequence', fontweight='bold')
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tx = ax10.text(0.97, x[0], round(np.median(all_nsti_default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))
tx = ax10.text(0.97, x[1], round(np.median(all_nsti_updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))

plt.sca(ax11)
x = [18.5, 17.5]
xlocs = []
all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_weighted_nsti.tsv', index_col=0, header=0, sep='\t')#['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_weighted_nsti.tsv', index_col=0, header=0, sep='\t')#['weighted_NSTI'].values
  asvs = list(set(list(default.index.values)+list(updated.index.values)))
  for asv in asvs:
    if asv in default.index.values:
      def_nsti = default.loc[asv, 'weighted_NSTI']
      all_nsti_default.append(def_nsti)
    else: def_nsti = 'NA'
    if asv in updated.index.values:
      upd_nsti = updated.loc[asv, 'weighted_NSTI']
      all_nsti_updated.append(upd_nsti)
    else: upd_nsti = 'NA'
    if def_nsti == 'NA' or upd_nsti == 'NA':
      continue
    else:
      nsti_difference.append(upd_nsti-def_nsti)
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  sc = ax11.scatter(default, np.random.normal(x[0], scale=0.1, size=len(default)), color=colours[0], alpha=0.5, s=10)
  sc = ax11.scatter(updated, np.random.normal(x[1], scale=0.1, size=len(updated)), color=colours[1], alpha=0.5, s=10)
  fc, alp = 'w', 0.7
  if np.median(default) < np.median(updated): fc, alp = 'gray', 0.6
  tx = ax11.text(0.68, x[0], round(np.median(default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax11.text(0.68, x[1], round(np.median(updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  box = ax11.boxplot([default, updated], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] -= 2.5
  x[1] -= 2.5

yl = plt.xlim([-0.02, 0.7])
xt = plt.yticks([]), plt.xticks([])

#
plt.sca(ax12)
x = [1, 0]
sc = ax12.scatter(all_nsti_default, np.random.normal(x[0], scale=0.1, size=len(all_nsti_default)), color=colours[0], alpha=0.5, s=10)
sc = ax12.scatter(all_nsti_updated, np.random.normal(x[1], scale=0.1, size=len(all_nsti_updated)), color=colours[1], alpha=0.5, s=10)
yl = plt.xlim([-0.02, 0.7])
box = ax12.boxplot([all_nsti_default, all_nsti_updated], positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([])
xl = plt.xlabel('NSTI per sample', fontweight='bold')
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tx = ax12.text(0.68, x[0], round(np.median(all_nsti_default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))
tx = ax12.text(0.68, x[1], round(np.median(all_nsti_updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))
# 
# #taxonomy
# 
tax = pd.read_csv(overall_folder+'mock/genome_full_tax.csv', index_col=0, header=0)

tax_genus, tax_class, tax_family = {}, {}, {}
for row in tax.index.values:
  this_tax = tax.loc[row, 'Taxonomy'].split(';')
  tax_genus[row] = this_tax[-2]
  tax_family[row] = this_tax[-3]
  tax_class[row] = this_tax[2].replace('c__', '')

combined = []
for ds in datasets:
  ft = pd.read_csv(overall_folder+'mock/'+ds+'_genome_abundance.csv', index_col=0, header=0).drop('Tax', axis=1)
  ft = ft.rename(index=tax_class)
  ft = ft.groupby(by=ft.index).sum()
  if 'Unassigned' in ft.index.values: ft = ft.drop('Unassigned')
  ft = ft.divide(ft.sum(axis=0), axis=1).multiply(100)
  ft[ds] = ft.mean(axis=1)
  combined.append(ft.loc[:, [ds]])

combined_df = pd.concat(combined).fillna(value=0)
combined_df = combined_df.groupby(by=combined_df.index).sum()
combined_df['Mean'] = combined_df.mean(axis=1)
combined_df = combined_df.sort_values(by=['Mean'], ascending=False)
combined_df = combined_df.iloc[:19, :]
combined_df = combined_df.drop('Mean', axis=1)
combined_df = combined_df.loc[sorted(combined_df.index.values), :]
combined_df.loc['Other', :] = 100-combined_df.sum(axis=0).values
plt.sca(ax13)
a1 = combined_df.transpose().plot(ax=ax13, kind='bar', stacked=True, width=0.9, cmap='tab20', edgecolor='k')
lg = plt.legend(loc='upper left', bbox_to_anchor=(1.1, 1.02), fontsize=10)
yl = plt.ylabel('Relative abundance (%)')
yl = plt.ylim([0, 100])
xt = plt.xticks([d for d in range(len(dataset_names))], dataset_names, rotation=90)
# 
plt.sca(ax14)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, "Spearman's correlation", ha='center', va='center', fontsize=12, fontweight='bold')

plt.sca(ax15)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, 'Bray-Curtis dissimilarity', ha='center', va='center', fontsize=12, fontweight='bold')

x = [18.5, 17.5]
xlocs = []
all_correlations, all_distances = [], []
for ds in datasets:
  mock = pd.read_csv(overall_folder+'mock/'+ds+'_KO_metagenome_norm.csv', index_col=0, header=0)
  mock = mock.drop('-', axis=0)
  mock = mock[mock.max(axis=1) > 0]
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_KO_metagenome.tsv', index_col=0, header=0, sep='\t')
  rename = {}
  for row in default.index.values:
    rename[row] = 'ko:'+row
  default = default.rename(index=rename)
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_KO_metagenome.tsv', index_col=0, header=0, sep='\t')
  samples = [s for s in mock.columns if s in default.columns and s in updated.columns]
  comparisons = [default, updated]
  names = ['Default', 'Updated']
  ds_correlations, ds_distances = [], []
  for c in range(len(comparisons)):
    comp = comparisons[c].copy(deep=True)
    mock_comp = mock.copy(deep=True)
    comp_relabun = comp.copy(deep=True)
    comp_relabun = comp_relabun.divide(comp_relabun.sum(axis=0), axis=1).multiply(100)
    mock_comp_relabun = mock_comp.copy(deep=True)
    mock_comp_relabun = mock_comp_relabun.divide(mock_comp_relabun.sum(axis=0), axis=1).multiply(100)
    intersection = [trait for trait in comp.index.values if trait in mock_comp.index.values]
    comp = comp.loc[intersection, :]
    mock_comp = mock_comp.loc[intersection, :]
    correlations, distances = [], []
    for sample in samples:
        combined = pd.concat([comp_relabun.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp_relabun.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mock_vals)
        ap = distances.append(bc_dist), correlations.append(stat)
    ap = ds_correlations.append(correlations), ds_distances.append(distances)
  ap = all_correlations.append(ds_correlations), all_distances.append(ds_distances)

overall_correlations, overall_distances = [[], []], [[], []]
for d in range(len(datasets)):
  sc = ax16.scatter(all_correlations[d][0], np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), color=colours[0], alpha=0.5, s=10)
  sc = ax16.scatter(all_correlations[d][1], np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), color=colours[1], alpha=0.5, s=10)
  fc, alp = 'w', 0.7
  if np.median(all_correlations[d][0]) > np.median(all_correlations[d][1]): fc, alp = 'gray', 0.6
  tx = ax16.text(0.94, x[0], round(np.median(all_correlations[d][0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax16.text(0.94, x[1], round(np.median(all_correlations[d][1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  overall_correlations[0].extend(all_correlations[d][0])
  overall_correlations[1].extend(all_correlations[d][1])
  box = ax16.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  sc = ax18.scatter(all_distances[d][0], np.random.normal(x[0], scale=0.1, size=len(all_distances[d][0])), color=colours[0], alpha=0.5, s=10)
  sc = ax18.scatter(all_distances[d][1], np.random.normal(x[1], scale=0.1, size=len(all_distances[d][1])), color=colours[1], alpha=0.5, s=10)
  fc, alp = 'w', 0.7
  if np.median(all_distances[d][0]) < np.median(all_distances[d][1]): fc, alp = 'gray', 0.6
  tx = ax18.text(0.63, x[0], round(np.median(all_distances[d][0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax18.text(0.63, x[1], round(np.median(all_distances[d][1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  overall_distances[0].extend(all_distances[d][0])
  overall_distances[1].extend(all_distances[d][1])
  box = ax18.boxplot(all_distances[d], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xl = xlocs.append(np.mean(x))
  x[0] += 2.5
  x[1] += 2.5

plt.sca(ax16)
#xl = plt.xlabel('Spearman correlation coefficient', fontweight='bold')
#yt = plt.yticks(xlocs, dataset_names)
yt = plt.yticks([])
xl = plt.xlim([0.63, 0.95])

plt.sca(ax18)
#xl = plt.xlabel('Bray-Curtis dissimilarity index', fontweight='bold')
yt = plt.yticks([])
xl = plt.xlim([0.15, 0.65])

plt.sca(ax17)
x = [1, 0]
sc = ax17.scatter(overall_correlations[0], np.random.normal(x[0], scale=0.1, size=len(overall_correlations[0])), color=colours[0], alpha=0.5, s=10)
sc = ax17.scatter(overall_correlations[1], np.random.normal(x[1], scale=0.1, size=len(overall_correlations[1])), color=colours[1], alpha=0.5, s=10)
xl = plt.xlim([0.63, 0.95])
box = ax17.boxplot(overall_correlations, positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([])
xl = plt.xlabel('Spearman correlation coefficient', fontweight='bold')
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tx = ax17.text(0.94, x[0], round(np.median(overall_correlations[0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))
tx = ax17.text(0.94, x[1], round(np.median(overall_correlations[1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))

plt.sca(ax19)
x = [1, 0]
sc = ax19.scatter(overall_distances[0], np.random.normal(x[0], scale=0.1, size=len(overall_distances[0])), color=colours[0], alpha=0.5, s=10)
sc = ax19.scatter(overall_distances[1], np.random.normal(x[1], scale=0.1, size=len(overall_distances[1])), color=colours[1], alpha=0.5, s=10)
xl = plt.xlim([0.15, 0.65])
box = ax19.boxplot(overall_distances, positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([])
xl = plt.xlabel('Bray-Curtis dissimilarity index', fontweight='bold')
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tx = ax19.text(0.63, x[0], round(np.median(overall_distances[0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))
tx = ax19.text(0.63, x[1], round(np.median(overall_distances[1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))

plt.savefig(overall_folder+'figures/overall_figure.png', dpi=600, bbox_inches='tight')
```

```{python}
fig = plt.figure(figsize=(15,20))

ax_db = plt.subplot2grid((60,20),(0,0),rowspan=5, colspan=5, frameon=False)
ax_annot_def_title = plt.subplot2grid((60,20),(7,0),rowspan=1, colspan=2)
ax_annot_upd_title = plt.subplot2grid((60,20),(7,2),rowspan=1, colspan=2)
ax_annot_def = plt.subplot2grid((60,20),(8,0),rowspan=8, colspan=2)
ax_annot_upd = plt.subplot2grid((60,20),(8,2),rowspan=8, colspan=2)
ax1 = plt.subplot2grid((60,20),(0,6),rowspan=1, colspan=4)
ax2 = plt.subplot2grid((60,20),(1,6),rowspan=5, colspan=4)
ax3 = plt.subplot2grid((60,20),(0,10), rowspan=1, colspan=4)
ax4 = plt.subplot2grid((60,20),(1,10), rowspan=5, colspan=4)
ax5 = plt.subplot2grid((60,20),(6,6), rowspan=9, colspan=4)
ax6 = plt.subplot2grid((60,20),(6,10), rowspan=9, colspan=4)
ax_leg = plt.subplot2grid((60,20),(15,6), rowspan=2, colspan=8, frameon=False)
#tax
ax13 = plt.subplot2grid((60,20),(0,15), rowspan=14, colspan=2)

#NSTI

#ax8 = plt.subplot2grid((60,20),(19,0), rowspan=1, colspan=5)
ax8_2 = plt.subplot2grid((60,21),(19,0), rowspan=1, colspan=7)
#ax9 = plt.subplot2grid((60,20),(20,0), rowspan=14, colspan=5)
#ax10 = plt.subplot2grid((60,20),(34,0), rowspan=2, colspan=5)
ax11 = plt.subplot2grid((60,21),(20,0), rowspan=14, colspan=7)
ax12 = plt.subplot2grid((60,21),(34,0), rowspan=2, colspan=7)

#Spearmans
ax14 = plt.subplot2grid((60,21),(19,7), rowspan=1, colspan=7)
ax15 = plt.subplot2grid((60,21),(19,14), rowspan=1, colspan=7)
ax16 = plt.subplot2grid((60,21),(20,7), rowspan=14, colspan=7)
ax17 = plt.subplot2grid((60,21),(34,7), rowspan=2, colspan=7)
ax18 = plt.subplot2grid((60,21),(20,14), rowspan=14, colspan=7)
ax19 = plt.subplot2grid((60,21),(34,14), rowspan=2, colspan=7)

ti = ax_db.set_title('a) SC construction', fontweight='bold', loc='left', fontsize=16)
ti = ax_annot_def_title.set_title('b) Functions included', fontweight='bold', loc='left', fontsize=16)
ti = ax1.set_title('c) Number of taxa included', fontweight='bold', loc='left', fontsize=16)
ti = ax13.set_title('d) Simulated samples', fontweight='bold', loc='left', fontsize=16)
ti = ax8_2.set_title('e) Performance of PICRUSt2 databases on simulated samples', fontweight='bold', loc='left', fontsize=16)

steps = ['Step 1: Get GTDB genomes\n', 'Step 2: Filter to GTDB reference genomes\n', 'Step 3: Annotate genomes with Eggnog\n', 'Step 4: Identify genomes present in GTDB\n            phylogenetic trees\n', 'Step 5: Identify genomes with 16S\n', 'Step 6: Keep only genomes 10% cont-\n            amination & 90% completion']
string = ''
for step in steps: string += step
tx = ax_db.text(-0.7, 1.03, string, ha='left', va='top', fontsize=10)

plt.sca(ax_db)
li = plt.xlim([0, 1]), plt.ylim([0, 1])

plt.sca(ax_db)
li = plt.xlim([-0.4, 4.5]), plt.ylim([0.9, 1.04])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax1)
bar = ax1.bar(0, 1, width=1, color=colours[-1], alpha=0.5)
tx = ax1.text(0, 0.5, 'Bacteria', fontweight='bold', fontsize=12, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax_annot_def_title)
tx = ax_annot_def_title.text(0, 0.5, 'oldIMG', fontweight='bold', fontsize=12, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax_annot_upd_title)
tx = ax_annot_upd_title.text(0, 0.5, 'SC', fontweight='bold', fontsize=12, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

plt.sca(ax3)
bar = ax3.bar(0, 1, width=1, color=colours[-2], alpha=0.5)
tx = ax3.text(0, 0.5, 'Archaea', fontweight='bold', fontsize=12, ha='center', va='center')
li = plt.xlim([-0.5, 0.5]), plt.ylim([0, 1])
xt = plt.xticks([]), plt.yticks([])

# #number of taxa
data = pd.read_csv(folder+'Number_of_genomes_old_and_new.csv', index_col=0, header=0)
levels = ['Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'Genomes']
locs = [0, 1.5, 3, 4.5, 6, 7.5, 9]
x2 = [locs[l]+0.3 for l in range(len(levels))]
x1 = [locs[l]-0.3 for l in range(len(levels))]
b = ax5.barh(x1, data.loc['PICRUSt2 updated Bacteria', :].values, color=colours[1], edgecolor='k', height=0.5)
b = ax6.barh(x1, data.loc['PICRUSt2 updated Archaea', :].values, color=colours[1], edgecolor='k', height=0.5)
b = ax5.barh(x2, data.loc['PICRUSt2 default Bacteria', :].values, color=colours[0], edgecolor='k', height=0.5)
b = ax6.barh(x2, data.loc['PICRUSt2 default Archaea', :].values, color=colours[0], edgecolor='k', height=0.5)
# 
multiply = 0.1
adding = [10, 10, 100, 100, 100, 1000, 1000]
for l in range(len(levels)):
  upd_bac = data.loc['PICRUSt2 updated Bacteria', levels[l]]
  def_bac = data.loc['PICRUSt2 default Bacteria', levels[l]]
  upd_arc = data.loc['PICRUSt2 updated Archaea', levels[l]]
  def_arc = data.loc['PICRUSt2 default Archaea', levels[l]]
  tx = ax5.text(upd_bac+200, x1[l], str(upd_bac), ha='left', va='center', fontsize=10)
  tx = ax5.text(def_bac+200, x2[l], str(def_bac), ha='left', va='center', fontsize=10)
  tx = ax6.text(upd_arc+10, x1[l], str(upd_arc), ha='left', va='center', fontsize=10)
  tx = ax6.text(def_arc+10, x2[l], str(def_arc), ha='left', va='center', fontsize=10)

genomes_at_each_step = [[394932, 7777], [80789, 4416], [77930, 4184], [31476, 1531], [26868, 1002]]
#ax2, ax4
y = [5, 4, 3, 2, 1]
steps = ['Step 1', 'Step 2', 'Step 4', 'Step 5', 'Step 6']

for a in range(len(genomes_at_each_step)):
  ba = ax2.barh(y[a], genomes_at_each_step[a][0], color=colours[2], edgecolor='k')
  tx = ax2.text(genomes_at_each_step[a][0]+5000, y[a], genomes_at_each_step[a][0], ha='left', va='center')
  ba = ax4.barh(y[a], genomes_at_each_step[a][1], color=colours[2], edgecolor='k')
  tx = ax4.text(genomes_at_each_step[a][1]+100, y[a], genomes_at_each_step[a][1], ha='left', va='center')
  
plt.sca(ax2)
xl = plt.xlim([0, 530000])
yt = plt.yticks(y, steps), plt.xticks([])

plt.sca(ax4)
xl = plt.xlim([0, 9500])
yt = plt.yticks([]), plt.xticks([])

databases = ['PICRUSt2-oldIMG', 'PICRUSt2-SC']
handles = [Line2D([0], [0], marker='s', color='w', label=databases[s], markerfacecolor=colours[s], markersize=12) for s in range(len(databases))]
legend = ax_leg.legend(handles=handles, loc='center', ncol=2)
plt.sca(ax_leg)
xt = plt.xticks([]), plt.yticks([])

functions = ['BiGG reactions', 'CAZy', 'COG', 'EC numbers', 'Gene names', 'GO', 'KO', 'Pfam', 'Phenotypes', 'TIGRFAM']
y = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
default_counts = ['-', '-', 4598, 2913, '-', '-', 10543, 11089, 41, 4287]
updated_counts = []
names = ['bigg_reaction', 'cazy', '', 'ec', 'preferred_name', 'go', 'ko', 'pfam', '', '']

# for name in names:
#   if name == '': 
#     updated_counts.append(0)
#     continue
#   bac = pd.read_csv(filepath_or_buffer='/Users/robynwright/Dropbox/Langille_Lab_postdoc/Github/picrust2/picrust2/default_files/bacteria/'+name+'.txt.gz', index_col=0, header=0, sep='\t')
#   arc = pd.read_csv(filepath_or_buffer='/Users/robynwright/Dropbox/Langille_Lab_postdoc/Github/picrust2/picrust2/default_files/archaea/'+name+'.txt.gz', index_col=0, header=0, sep='\t')
#   names = list(set(list(bac.columns)+list(arc.columns)))
#   updated_counts.append(len(names))
#
# print(updated_counts)

updated_counts = [10452, 151, '-', 3763, 31131, 23273, 14106, 11402, '-', '-']
for f in range(len(functions)):
  if default_counts[f] == '-':
    ba = ax_annot_def.bar(0, 1, bottom=y[f]-0.5, width=1, edgecolor='k', color='gray', alpha=0.2)
  else:
    ba = ax_annot_def.bar(0, 1, bottom=y[f]-0.5, width=1, edgecolor='k', color='w')
  tx = ax_annot_def.text(0, y[f], default_counts[f], ha='center', va='center')
  if updated_counts[f] == '-':
    ba = ax_annot_upd.bar(0, 1, bottom=y[f]-0.5, width=1, edgecolor='k', color='gray', alpha=0.2)
  else:
    ba = ax_annot_upd.bar(0, 1, bottom=y[f]-0.5, width=1, edgecolor='k', color='w')
  tx = ax_annot_upd.text(0, y[f], updated_counts[f], ha='center', va='center')

plt.sca(ax_annot_def)
yt = plt.yticks(y, functions), plt.xticks([])
li = plt.ylim([0.5, 10.5]), plt.xlim([-0.5, 0.5])

plt.sca(ax_annot_upd)
yt = plt.yticks([]), plt.xticks([])
li = plt.ylim([0.5, 10.5]), plt.xlim([-0.5, 0.5])

plt.sca(ax5)
yt = plt.yticks([locs[l] for l in range(len(levels))], [l.replace('Genomes', 'Genome') for l in levels]), plt.xticks([])
xl = plt.xlim([-400, 34500])
# 
plt.sca(ax6)
yt = plt.yticks([]), plt.xticks([])
xl = plt.xlim([-15, 1280])

# plt.sca(ax7)
# yt = plt.yticks([]), plt.xticks([])
# xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
# tx = plt.text(0.5, 0.5, 'Simulated sample composition', ha='center', va='center', fontsize=12, fontweight='bold')

# plt.sca(ax8)
# yt = plt.yticks([]), plt.xticks([])
# xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
# tx = plt.text(0.5, 0.5, 'NSTI per sequence', ha='center', va='center', fontsize=12, fontweight='bold')

plt.sca(ax8_2)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, 'Weighted NSTI', ha='center', va='center', fontsize=12, fontweight='bold')

# plt.sca(ax9)
# x = [18.5, 17.5]
# xlocs = []
# all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
# for ds in datasets:
#   default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
#   updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_nsti.tsv', index_col=0, header=0, sep='\t')#['metadata_NSTI'].values
#   asvs = list(set(list(default.index.values)+list(updated.index.values)))
#   for asv in asvs:
#     if asv in default.index.values:
#       def_nsti = default.loc[asv, 'metadata_NSTI']
#       all_nsti_default.append(def_nsti)
#     else: def_nsti = 'NA'
#     if asv in updated.index.values:
#       upd_nsti = updated.loc[asv, 'metadata_NSTI']
#       all_nsti_updated.append(upd_nsti)
#     else: upd_nsti = 'NA'
#     if def_nsti == 'NA' or upd_nsti == 'NA':
#       continue
#     else:
#       nsti_difference.append(upd_nsti-def_nsti)
#   default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
#   updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_nsti.tsv', index_col=0, header=0, sep='\t')['metadata_NSTI'].values
#   sc = ax9.scatter(default, np.random.normal(x[0], scale=0.1, size=len(default)), color=colours[0], alpha=0.5, s=10)
#   sc = ax9.scatter(updated, np.random.normal(x[1], scale=0.1, size=len(updated)), color=colours[1], alpha=0.5, s=10)
#   fc, alp = 'w', 0.7
#   if np.median(default) < np.median(updated): fc, alp = 'gray', 0.6
#   tx = ax9.text(0.97, x[0], round(np.median(default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
#   tx = ax9.text(0.97, x[1], round(np.median(updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
#   box = ax9.boxplot([default, updated], positions=x, widths=0.8, showfliers=False, vert=False)
#   for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
#   xlocs.append(np.mean(x))
#   x[0] -= 2.5
#   x[1] -= 2.5
# 
# yl = plt.xlim([-0.05, 1])
# xt = plt.yticks(xlocs, dataset_names), plt.xticks([])
# 
# #
# plt.sca(ax10)
# x = [1, 0]
# sc = ax10.scatter(all_nsti_default, np.random.normal(x[0], scale=0.1, size=len(all_nsti_default)), color=colours[0], alpha=0.5, s=10)
# sc = ax10.scatter(all_nsti_updated, np.random.normal(x[1], scale=0.1, size=len(all_nsti_updated)), color=colours[1], alpha=0.5, s=10)
# yl = plt.xlim([-0.05, 1])
# box = ax10.boxplot([all_nsti_default, all_nsti_updated], positions=x, widths=0.8, showfliers=False, vert=False)
# xt = plt.yticks([0.5], ['Overall'])
# xl = plt.xlabel('NSTI per sequence', fontweight='bold')
# for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
# tx = ax10.text(0.97, x[0], round(np.median(all_nsti_default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))
# tx = ax10.text(0.97, x[1], round(np.median(all_nsti_updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor='w', alpha=0.7))

plt.sca(ax11)
x = [18.5, 17.5]
xlocs = []
all_nsti_default, all_nsti_updated, nsti_difference = [], [], []
for ds in datasets:
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_weighted_nsti.tsv', index_col=0, header=0, sep='\t')#['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_weighted_nsti.tsv', index_col=0, header=0, sep='\t')#['weighted_NSTI'].values
  asvs = list(set(list(default.index.values)+list(updated.index.values)))
  for asv in asvs:
    if asv in default.index.values:
      def_nsti = default.loc[asv, 'weighted_NSTI']
      all_nsti_default.append(def_nsti)
    else: def_nsti = 'NA'
    if asv in updated.index.values:
      upd_nsti = updated.loc[asv, 'weighted_NSTI']
      all_nsti_updated.append(upd_nsti)
    else: upd_nsti = 'NA'
    if def_nsti == 'NA' or upd_nsti == 'NA':
      continue
    else:
      nsti_difference.append(upd_nsti-def_nsti)
  default = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_default_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  updated = pd.read_csv(overall_folder+'mgs_picrust2/'+ds+'_updated_weighted_nsti.tsv', index_col=0, header=0, sep='\t')['weighted_NSTI'].values
  sc = ax11.scatter(default, np.random.normal(x[0], scale=0.1, size=len(default)), color=colours[0], alpha=0.5, s=10)
  sc = ax11.scatter(updated, np.random.normal(x[1], scale=0.1, size=len(updated)), color=colours[1], alpha=0.5, s=10)
  tstat, p = ttest_ind(default, updated)
  fc, alp = 'w', 0.7
  if p <= 0.05: fc, alp ='gray', 0.6
  tx = ax11.text(0.68, x[0], round(np.median(default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax11.text(0.68, x[1], round(np.median(updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  box = ax11.boxplot([default, updated], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xlocs.append(np.mean(x))
  x[0] -= 2.5
  x[1] -= 2.5

yl = plt.xlim([-0.02, 0.7])
xt = plt.yticks(xlocs, dataset_names), plt.xticks([])

#
plt.sca(ax12)
x = [1, 0]
sc = ax12.scatter(all_nsti_default, np.random.normal(x[0], scale=0.1, size=len(all_nsti_default)), color=colours[0], alpha=0.5, s=10)
sc = ax12.scatter(all_nsti_updated, np.random.normal(x[1], scale=0.1, size=len(all_nsti_updated)), color=colours[1], alpha=0.5, s=10)
yl = plt.xlim([-0.02, 0.7])
box = ax12.boxplot([all_nsti_default, all_nsti_updated], positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([0.5], ['Overall'])
xl = plt.xlabel('Weighted NSTI', fontweight='bold')
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tstat, p = ttest_ind(all_nsti_default, all_nsti_updated)
fc, alp = 'w', 0.7
if p <= 0.05: fc, alp ='gray', 0.6
tx = ax12.text(0.68, x[0], round(np.median(all_nsti_default), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
tx = ax12.text(0.68, x[1], round(np.median(all_nsti_updated), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
# 
# #taxonomy
# 
tax = pd.read_csv(overall_folder+'mock/genome_full_tax.csv', index_col=0, header=0)

tax_genus, tax_class, tax_family = {}, {}, {}
for row in tax.index.values:
  this_tax = tax.loc[row, 'Taxonomy'].split(';')
  tax_genus[row] = this_tax[-2]
  tax_family[row] = this_tax[-3]
  tax_class[row] = this_tax[2].replace('c__', '')

combined = []
for ds in datasets:
  ft = pd.read_csv(overall_folder+'mock/'+ds+'_genome_abundance.csv', index_col=0, header=0).drop('Tax', axis=1)
  ft = ft.rename(index=tax_class)
  ft = ft.groupby(by=ft.index).sum()
  if 'Unassigned' in ft.index.values: ft = ft.drop('Unassigned')
  ft = ft.divide(ft.sum(axis=0), axis=1).multiply(100)
  ft[ds] = ft.mean(axis=1)
  combined.append(ft.loc[:, [ds]])

combined_df = pd.concat(combined).fillna(value=0)
combined_df = combined_df.groupby(by=combined_df.index).sum()
combined_df['Mean'] = combined_df.mean(axis=1)
combined_df = combined_df.sort_values(by=['Mean'], ascending=False)
combined_df = combined_df.iloc[:19, :]
combined_df = combined_df.drop('Mean', axis=1)
combined_df = combined_df.loc[sorted(combined_df.index.values), :]
combined_df.loc['Other', :] = 100-combined_df.sum(axis=0).values
plt.sca(ax13)
a1 = combined_df.transpose().plot(ax=ax13, kind='bar', stacked=True, width=0.9, cmap='tab20', edgecolor='k')
lg = plt.legend(loc='upper left', bbox_to_anchor=(1.1, 1.02), fontsize=10)
yl = plt.ylabel('Relative abundance (%)')
yl = plt.ylim([0, 100])
xt = plt.xticks([d for d in range(len(dataset_names))], dataset_names, rotation=90)
# 
plt.sca(ax14)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, "Spearman's correlation", ha='center', va='center', fontsize=12, fontweight='bold')

plt.sca(ax15)
yt = plt.yticks([]), plt.xticks([])
xl, yl = plt.xlim([0, 1]), plt.ylim([0, 1])
tx = plt.text(0.5, 0.5, 'Bray-Curtis dissimilarity', ha='center', va='center', fontsize=12, fontweight='bold')

x = [18.5, 17.5]
xlocs = []
all_correlations, all_distances = [], []
for ds in datasets:
  mock = pd.read_csv(overall_folder+'mock/'+ds+'_KO_metagenome_norm.csv', index_col=0, header=0)
  mock = mock.drop('-', axis=0)
  mock = mock[mock.max(axis=1) > 0]
  default = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_default_full_KO_metagenome.tsv', index_col=0, header=0, sep='\t')
  rename = {}
  for row in default.index.values:
    rename[row] = 'ko:'+row
  default = default.rename(index=rename)
  updated = pd.read_csv(overall_folder+'mock_picrust2/'+ds+'_updated_full_KO_metagenome.tsv', index_col=0, header=0, sep='\t')
  samples = [s for s in mock.columns if s in default.columns and s in updated.columns]
  comparisons = [default, updated]
  names = ['Default', 'Updated']
  ds_correlations, ds_distances = [], []
  for c in range(len(comparisons)):
    comp = comparisons[c].copy(deep=True)
    mock_comp = mock.copy(deep=True)
    comp_relabun = comp.copy(deep=True)
    comp_relabun = comp_relabun.divide(comp_relabun.sum(axis=0), axis=1).multiply(100)
    mock_comp_relabun = mock_comp.copy(deep=True)
    mock_comp_relabun = mock_comp_relabun.divide(mock_comp_relabun.sum(axis=0), axis=1).multiply(100)
    intersection = [trait for trait in comp.index.values if trait in mock_comp.index.values]
    comp = comp.loc[intersection, :]
    mock_comp = mock_comp.loc[intersection, :]
    correlations, distances = [], []
    for sample in samples:
        combined = pd.concat([comp_relabun.loc[:, [sample]].rename(columns={sample:'Comp'}), mock_comp_relabun.loc[:, [sample]].rename(columns={sample:'Mock'})]).fillna(value=0)
        combined = combined.groupby(by=combined.index.values).sum()
        X = combined.transpose().iloc[0:].values
        bc_dist = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))[0][1]
        comp_vals = comp.loc[:, sample].values
        mock_vals = mock_comp.loc[:, sample].values
        stat, p = spearmanr(comp_vals, mock_vals)
        ap = distances.append(bc_dist), correlations.append(stat)
    ap = ds_correlations.append(correlations), ds_distances.append(distances)
  ap = all_correlations.append(ds_correlations), all_distances.append(ds_distances)

overall_correlations, overall_distances = [[], []], [[], []]
for d in range(len(datasets)):
  sc = ax16.scatter(all_correlations[d][0], np.random.normal(x[0], scale=0.1, size=len(all_correlations[d][0])), color=colours[0], alpha=0.5, s=10)
  sc = ax16.scatter(all_correlations[d][1], np.random.normal(x[1], scale=0.1, size=len(all_correlations[d][1])), color=colours[1], alpha=0.5, s=10)
  tstat, p = ttest_ind(all_correlations[d][0], all_correlations[d][1])
  fc, alp = 'w', 0.7
  if p <= 0.05: fc, alp ='gray', 0.6
  tx = ax16.text(0.94, x[0], round(np.median(all_correlations[d][0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax16.text(0.94, x[1], round(np.median(all_correlations[d][1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  overall_correlations[0].extend(all_correlations[d][0])
  overall_correlations[1].extend(all_correlations[d][1])
  box = ax16.boxplot(all_correlations[d], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  sc = ax18.scatter(all_distances[d][0], np.random.normal(x[0], scale=0.1, size=len(all_distances[d][0])), color=colours[0], alpha=0.5, s=10)
  sc = ax18.scatter(all_distances[d][1], np.random.normal(x[1], scale=0.1, size=len(all_distances[d][1])), color=colours[1], alpha=0.5, s=10)
  tstat, p = ttest_ind(all_distances[d][0], all_distances[d][1])
  fc, alp = 'w', 0.7
  if p <= 0.05: fc, alp ='gray', 0.6
  tx = ax18.text(0.63, x[0], round(np.median(all_distances[d][0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  tx = ax18.text(0.63, x[1], round(np.median(all_distances[d][1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
  overall_distances[0].extend(all_distances[d][0])
  overall_distances[1].extend(all_distances[d][1])
  box = ax18.boxplot(all_distances[d], positions=x, widths=0.8, showfliers=False, vert=False)
  for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
  xl = xlocs.append(np.mean(x))
  x[0] += 2.5
  x[1] += 2.5

plt.sca(ax16)
#xl = plt.xlabel('Spearman correlation coefficient', fontweight='bold')
#yt = plt.yticks(xlocs, dataset_names)
yt = plt.yticks([])
xl = plt.xlim([0.63, 0.95])

plt.sca(ax18)
#xl = plt.xlabel('Bray-Curtis dissimilarity index', fontweight='bold')
yt = plt.yticks([])
xl = plt.xlim([0.15, 0.65])

plt.sca(ax17)
x = [1, 0]
sc = ax17.scatter(overall_correlations[0], np.random.normal(x[0], scale=0.1, size=len(overall_correlations[0])), color=colours[0], alpha=0.5, s=10)
sc = ax17.scatter(overall_correlations[1], np.random.normal(x[1], scale=0.1, size=len(overall_correlations[1])), color=colours[1], alpha=0.5, s=10)
xl = plt.xlim([0.63, 0.95])
box = ax17.boxplot(overall_correlations, positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([])
xl = plt.xlabel('Spearman correlation coefficient', fontweight='bold')
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tstat, p = ttest_ind(overall_correlations[0], overall_correlations[1])
fc, alp = 'w', 0.7
if p <= 0.05: fc, alp ='gray', 0.6
tx = ax17.text(0.94, x[0], round(np.median(overall_correlations[0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
tx = ax17.text(0.94, x[1], round(np.median(overall_correlations[1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))

plt.sca(ax19)
x = [1, 0]
sc = ax19.scatter(overall_distances[0], np.random.normal(x[0], scale=0.1, size=len(overall_distances[0])), color=colours[0], alpha=0.5, s=10)
sc = ax19.scatter(overall_distances[1], np.random.normal(x[1], scale=0.1, size=len(overall_distances[1])), color=colours[1], alpha=0.5, s=10)
xl = plt.xlim([0.15, 0.65])
box = ax19.boxplot(overall_distances, positions=x, widths=0.8, showfliers=False, vert=False)
xt = plt.yticks([])
xl = plt.xlabel('Bray-Curtis dissimilarity index', fontweight='bold')
for item in ['boxes', 'whiskers', 'fliers', 'medians', 'caps']: bi = plt.setp(box[item], color='k')
tstat, p = ttest_ind(overall_distances[0], overall_distances[1])
fc, alp = 'w', 0.7
if p <= 0.05: fc, alp ='gray', 0.6
tx = ax19.text(0.63, x[0], round(np.median(overall_distances[0]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))
tx = ax19.text(0.63, x[1], round(np.median(overall_distances[1]), 3), ha='right', va='center', bbox=dict(boxstyle='round', facecolor=fc, alpha=alp))

plt.savefig(overall_folder+'figures/overall_figure.png', dpi=600, bbox_inches='tight')
```